{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "6aac95ec-fd31-454b-9310-03db3fe03662",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# Open the fasta file and read sequences\n",
    "salmo_chrome = r\"C:\\Users\\User\\Downloads\\Salmonella_enterica.txt\"  # Replace with your file path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "959fb6e3-23a3-4986-8b5c-77311073fe69",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readGenome(salmo_chrome):\n",
    "    genome = ''\n",
    "    with open(salmo_chrome, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "da6d43d2-1fc3-4019-a443-423f4d07c62d",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = readGenome(salmo_chrome)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "17d94a09-e795-40ca-abf2-06ed48825ffb",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "e9f9c2c5-b55b-474f-a52b-0513fc9a9b56",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pattern_count(text, pattern):\n",
    "    count = 0\n",
    "    text_length = len(text)\n",
    "    pattern_length = len(pattern)\n",
    "\n",
    "    for i in range(text_length - pattern_length + 1):\n",
    "        if text[i:i + pattern_length] == pattern:\n",
    "            count += 1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "bb26de4d-773f-4814-8ab0-0b7da4b5e339",
   "metadata": {},
   "outputs": [],
   "source": [
    "def frequent_words(text, k):\n",
    "    frequent_patterns = set()  # Use a set to automatically remove duplicates\n",
    "    counts = []\n",
    "    text_length = len(text)\n",
    "\n",
    "    for i in range(text_length - k + 1):\n",
    "        pattern = text[i:i + k]\n",
    "        counts.append(pattern_count(text, pattern))  # Use the pattern_count function\n",
    "\n",
    "    max_count = max(counts)\n",
    "\n",
    "    for i in range(text_length - k + 1):\n",
    "        if counts[i] == max_count:\n",
    "            frequent_patterns.add(text[i:i + k])\n",
    "\n",
    "    return frequent_patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "4a243b9a-9279-41a3-a0f2-2cb962b25883",
   "metadata": {},
   "outputs": [],
   "source": [
    "def frequency_table(text, k):\n",
    "    freq_map = {}\n",
    "    n = len(text)\n",
    "    for i in range(n - k + 1):\n",
    "        pattern = text[i:i + k]\n",
    "        if pattern not in freq_map:\n",
    "            freq_map[pattern] = 1\n",
    "        else:\n",
    "            freq_map[pattern] += 1\n",
    "    return freq_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "2d3372df-bb39-4856-88d6-9633e8b97bda",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "text = genome\n",
    "k = 9\n",
    "print(frequency_table(text, k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "4594fce4-cee8-41ab-967c-5a1f32f107bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def max_map(freq_map):\n",
    "\n",
    "    if not freq_map:  # Handle the case of an empty dictionary\n",
    "        return 0  # Or raise an exception, depending on the desired behavior\n",
    "    return max(freq_map.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "9f802436-85cd-4e8f-999b-24e8a9f87e8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def better_frequent_words(text, k):\n",
    "   \n",
    "    frequent_patterns = []\n",
    "    freq_map = frequency_table(text, k)\n",
    "    max_count = max_map(freq_map)\n",
    "\n",
    "    for pattern, count in freq_map.items():\n",
    "        if count == max_count:\n",
    "            frequent_patterns.append(pattern)\n",
    "    return frequent_patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "ce112056-df49-4010-b7b5-9d47f15327d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TGG']\n"
     ]
    }
   ],
   "source": [
    "text = 'TGGACGTTGGCCCAGCTGGTCCCACGTGGT'\n",
    "k = 3\n",
    "print(better_frequent_words(text, k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "b5c6e0e5-2a29-4fca-845a-936fff1385df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TTTT']\n"
     ]
    }
   ],
   "source": [
    "text = 'CGTTTTGAACATTTTCAACAAGTTTTGCAACATTTT'\n",
    "k = 4\n",
    "print(better_frequent_words(text, k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "58a83858-a689-48ad-9db1-fbdef12eb370",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AACAA']\n"
     ]
    }
   ],
   "source": [
    "text = 'GTTGGGTTGGAACAACAACAACAACAAGTTGGGTTGG'\n",
    "k = 5\n",
    "print(better_frequent_words(text, k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "8e88093c-b085-41af-a22f-7b62db88c402",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GCAT', 'CATG']\n"
     ]
    }
   ],
   "source": [
    "text = 'ACGTTGCATGTCGCATGATGCATGAGAGCT'\n",
    "k = 4\n",
    "print(better_frequent_words(text, k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "e00823aa-c4ea-43bb-b461-032c4c013b8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GTAGGTCAAACCGGGTGTAC', 'TAGGTCAAACCGGGTGTACA', 'AGGTCAAACCGGGTGTACAT', 'GGTCAAACCGGGTGTACATA', 'GTCAAACCGGGTGTACATAC', 'TCAAACCGGGTGTACATACC', 'CAAACCGGGTGTACATACCC', 'AAACCGGGTGTACATACCCG', 'AACCGGGTGTACATACCCGC', 'ACCGGGTGTACATACCCGCT', 'CCGGGTGTACATACCCGCTC', 'CGGGTGTACATACCCGCTCA', 'GGGTGTACATACCCGCTCAA', 'GGTGTACATACCCGCTCAAT', 'GTGTACATACCCGCTCAATC', 'TGTACATACCCGCTCAATCG', 'GTACATACCCGCTCAATCGC', 'TACATACCCGCTCAATCGCC', 'ACATACCCGCTCAATCGCCC', 'CATACCCGCTCAATCGCCCA', 'ATACCCGCTCAATCGCCCAG', 'TACCCGCTCAATCGCCCAGC', 'ACCCGCTCAATCGCCCAGCA', 'CCCGCTCAATCGCCCAGCAC', 'CCGCTCAATCGCCCAGCACT', 'CGCTCAATCGCCCAGCACTT', 'GCTCAATCGCCCAGCACTTC', 'CTCAATCGCCCAGCACTTCG', 'TCAATCGCCCAGCACTTCGG', 'CAATCGCCCAGCACTTCGGG', 'AATCGCCCAGCACTTCGGGC', 'ATCGCCCAGCACTTCGGGCA', 'TCGCCCAGCACTTCGGGCAG', 'CGCCCAGCACTTCGGGCAGA', 'GCCCAGCACTTCGGGCAGAT', 'CCCAGCACTTCGGGCAGATC', 'CCAGCACTTCGGGCAGATCA', 'CAGCACTTCGGGCAGATCAC', 'AGCACTTCGGGCAGATCACC', 'GCACTTCGGGCAGATCACCG', 'CACTTCGGGCAGATCACCGG', 'ACTTCGGGCAGATCACCGGG', 'CTTCGGGCAGATCACCGGGT', 'TTCGGGCAGATCACCGGGTT', 'TCGGGCAGATCACCGGGTTT', 'CGGGCAGATCACCGGGTTTC', 'GGGCAGATCACCGGGTTTCC', 'GGCAGATCACCGGGTTTCCC', 'GCAGATCACCGGGTTTCCCC', 'CAGATCACCGGGTTTCCCCG', 'AGATCACCGGGTTTCCCCGG', 'GATCACCGGGTTTCCCCGGT', 'ATCACCGGGTTTCCCCGGTA', 'TCACCGGGTTTCCCCGGTAT', 'CACCGGGTTTCCCCGGTATC', 'ACCGGGTTTCCCCGGTATCA', 'CCGGGTTTCCCCGGTATCAC', 'CGGGTTTCCCCGGTATCACC', 'GGGTTTCCCCGGTATCACCA', 'GGTTTCCCCGGTATCACCAA', 'GTTTCCCCGGTATCACCAAT', 'TTTCCCCGGTATCACCAATA', 'TTCCCCGGTATCACCAATAC', 'TCCCCGGTATCACCAATACT', 'CCCCGGTATCACCAATACTG', 'CCCGGTATCACCAATACTGC', 'CCGGTATCACCAATACTGCC', 'CGGTATCACCAATACTGCCA', 'GGTATCACCAATACTGCCAC', 'GTATCACCAATACTGCCACC', 'TATCACCAATACTGCCACCA', 'ATCACCAATACTGCCACCAA', 'TCACCAATACTGCCACCAAA', 'CACCAATACTGCCACCAAAC', 'ACCAATACTGCCACCAAACA', 'CCAATACTGCCACCAAACAC', 'CAATACTGCCACCAAACACA', 'AATACTGCCACCAAACACAG', 'ATACTGCCACCAAACACAGC', 'TACTGCCACCAAACACAGCA', 'ACTGCCACCAAACACAGCAG', 'CTGCCACCAAACACAGCAGG', 'TGCCACCAAACACAGCAGGC', 'GCCACCAAACACAGCAGGCG', 'CCACCAAACACAGCAGGCGG', 'CACCAAACACAGCAGGCGGG', 'ACCAAACACAGCAGGCGGGA', 'CCAAACACAGCAGGCGGGAA', 'CAAACACAGCAGGCGGGAAG', 'AAACACAGCAGGCGGGAAGG', 'AACACAGCAGGCGGGAAGGG', 'ACACAGCAGGCGGGAAGGGG', 'CACAGCAGGCGGGAAGGGGC', 'ACAGCAGGCGGGAAGGGGCG', 'CAGCAGGCGGGAAGGGGCGA', 'AGCAGGCGGGAAGGGGCGAA', 'GCAGGCGGGAAGGGGCGAAA', 'CAGGCGGGAAGGGGCGAAAG', 'AGGCGGGAAGGGGCGAAAGT', 'GGCGGGAAGGGGCGAAAGTC', 'GCGGGAAGGGGCGAAAGTCC', 'CGGGAAGGGGCGAAAGTCCC', 'GGGAAGGGGCGAAAGTCCCT', 'GGAAGGGGCGAAAGTCCCTT', 'GAAGGGGCGAAAGTCCCTTA', 'AAGGGGCGAAAGTCCCTTAT', 'AGGGGCGAAAGTCCCTTATC', 'GGGGCGAAAGTCCCTTATCC', 'GGGCGAAAGTCCCTTATCCG', 'GGCGAAAGTCCCTTATCCGA', 'GCGAAAGTCCCTTATCCGAC', 'CGAAAGTCCCTTATCCGACA', 'GAAAGTCCCTTATCCGACAA', 'AAAGTCCCTTATCCGACAAT', 'AAGTCCCTTATCCGACAATA', 'AGTCCCTTATCCGACAATAA', 'GTCCCTTATCCGACAATAAA', 'TCCCTTATCCGACAATAAAA', 'CCCTTATCCGACAATAAAAC', 'CCTTATCCGACAATAAAACT', 'CTTATCCGACAATAAAACTT', 'TTATCCGACAATAAAACTTC', 'TATCCGACAATAAAACTTCG', 'ATCCGACAATAAAACTTCGC', 'TCCGACAATAAAACTTCGCT', 'CCGACAATAAAACTTCGCTT', 'CGACAATAAAACTTCGCTTG', 'GACAATAAAACTTCGCTTGT', 'ACAATAAAACTTCGCTTGTT', 'CAATAAAACTTCGCTTGTTC', 'AATAAAACTTCGCTTGTTCG', 'ATAAAACTTCGCTTGTTCGA', 'TAAAACTTCGCTTGTTCGAC', 'AAAACTTCGCTTGTTCGACG', 'AAACTTCGCTTGTTCGACGC', 'AACTTCGCTTGTTCGACGCC', 'ACTTCGCTTGTTCGACGCCC', 'CTTCGCTTGTTCGACGCCCG', 'TTCGCTTGTTCGACGCCCGG', 'TCGCTTGTTCGACGCCCGGT', 'CGCTTGTTCGACGCCCGGTT', 'GCTTGTTCGACGCCCGGTTC', 'CTTGTTCGACGCCCGGTTCA', 'TTGTTCGACGCCCGGTTCAC', 'TGTTCGACGCCCGGTTCACC', 'GTTCGACGCCCGGTTCACCC', 'TTCGACGCCCGGTTCACCCG', 'TCGACGCCCGGTTCACCCGA', 'CGACGCCCGGTTCACCCGAT', 'GACGCCCGGTTCACCCGATA', 'ACGCCCGGTTCACCCGATAT', 'CGCCCGGTTCACCCGATATG', 'GCCCGGTTCACCCGATATGC', 'CCCGGTTCACCCGATATGCA', 'CCGGTTCACCCGATATGCAC', 'CGGTTCACCCGATATGCACG', 'GGTTCACCCGATATGCACGG', 'GTTCACCCGATATGCACGGC', 'TTCACCCGATATGCACGGCG', 'TCACCCGATATGCACGGCGC', 'CACCCGATATGCACGGCGCC', 'ACCCGATATGCACGGCGCCC', 'CCCGATATGCACGGCGCCCA', 'CCGATATGCACGGCGCCCAG', 'CGATATGCACGGCGCCCAGC', 'GATATGCACGGCGCCCAGCC', 'ATATGCACGGCGCCCAGCCA', 'TATGCACGGCGCCCAGCCAT', 'ATGCACGGCGCCCAGCCATT', 'TGCACGGCGCCCAGCCATTC', 'GCACGGCGCCCAGCCATTCG', 'CACGGCGCCCAGCCATTCGT', 'ACGGCGCCCAGCCATTCGTG', 'CGGCGCCCAGCCATTCGTGA', 'GGCGCCCAGCCATTCGTGAC', 'GCGCCCAGCCATTCGTGACC', 'CGCCCAGCCATTCGTGACCG', 'GCCCAGCCATTCGTGACCGA', 'CCCAGCCATTCGTGACCGAC', 'CCAGCCATTCGTGACCGACG', 'CAGCCATTCGTGACCGACGT', 'AGCCATTCGTGACCGACGTC', 'GCCATTCGTGACCGACGTCC', 'CCATTCGTGACCGACGTCCC', 'CATTCGTGACCGACGTCCCC', 'ATTCGTGACCGACGTCCCCA', 'TTCGTGACCGACGTCCCCAG', 'TCGTGACCGACGTCCCCAGC', 'CGTGACCGACGTCCCCAGCC', 'GTGACCGACGTCCCCAGCCC', 'TGACCGACGTCCCCAGCCCC', 'GACCGACGTCCCCAGCCCCA', 'ACCGACGTCCCCAGCCCCAA', 'CCGACGTCCCCAGCCCCAAG', 'CGACGTCCCCAGCCCCAAGG', 'GACGTCCCCAGCCCCAAGGC', 'ACGTCCCCAGCCCCAAGGCC', 'CGTCCCCAGCCCCAAGGCCG', 'GTCCCCAGCCCCAAGGCCGA', 'TCCCCAGCCCCAAGGCCGAA', 'CCCCAGCCCCAAGGCCGAAC', 'CCCAGCCCCAAGGCCGAACG', 'CCAGCCCCAAGGCCGAACGA', 'CAGCCCCAAGGCCGAACGAC', 'AGCCCCAAGGCCGAACGACC', 'GCCCCAAGGCCGAACGACCC', 'CCCCAAGGCCGAACGACCCT', 'CCCAAGGCCGAACGACCCTA', 'CCAAGGCCGAACGACCCTAG', 'CAAGGCCGAACGACCCTAGG', 'AAGGCCGAACGACCCTAGGA', 'AGGCCGAACGACCCTAGGAG', 'GGCCGAACGACCCTAGGAGC', 'GCCGAACGACCCTAGGAGCC', 'CCGAACGACCCTAGGAGCCA', 'CGAACGACCCTAGGAGCCAC', 'GAACGACCCTAGGAGCCACG', 'AACGACCCTAGGAGCCACGA', 'ACGACCCTAGGAGCCACGAG', 'CGACCCTAGGAGCCACGAGC', 'GACCCTAGGAGCCACGAGCA', 'ACCCTAGGAGCCACGAGCAA', 'CCCTAGGAGCCACGAGCAAT', 'CCTAGGAGCCACGAGCAATT', 'CTAGGAGCCACGAGCAATTC', 'TAGGAGCCACGAGCAATTCA', 'AGGAGCCACGAGCAATTCAC', 'GGAGCCACGAGCAATTCACA', 'GAGCCACGAGCAATTCACAG', 'AGCCACGAGCAATTCACAGC', 'GCCACGAGCAATTCACAGCG', 'CCACGAGCAATTCACAGCG ']\n"
     ]
    }
   ],
   "source": [
    "text = 'GTAGGTCAAACCGGGTGTACATACCCGCTCAATCGCCCAGCACTTCGGGCAGATCACCGGGTTTCCCCGGTATCACCAATACTGCCACCAAACACAGCAGGCGGGAAGGGGCGAAAGTCCCTTATCCGACAATAAAACTTCGCTTGTTCGACGCCCGGTTCACCCGATATGCACGGCGCCCAGCCATTCGTGACCGACGTCCCCAGCCCCAAGGCCGAACGACCCTAGGAGCCACGAGCAATTCACAGCG '\n",
    "k = 20\n",
    "print(better_frequent_words(text, k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "07cb98e3-1d1d-473a-a182-e5db36ed38a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_complement(dna):\n",
    "   \n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "    reverse_dna = dna[::-1]  # Efficiently reverse the string\n",
    "    reverse_complement_dna = \"\"\n",
    "\n",
    "    for base in reverse_dna:\n",
    "        if base in complement:\n",
    "            reverse_complement_dna += complement[base]\n",
    "        else:\n",
    "            return \"\"  # Return empty string for invalid input\n",
    "\n",
    "    return reverse_complement_dna\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "bc7cda5f-544d-412d-b522-28e7ea22875d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTATAACATGCACATCAAGCCCGCGGATGAACACAAAACGCTCCGAAAGCGCTGTTGTGGAGAAGCGCGTACTCTTAGCTTTATTGACTATGGAGGAGTGATACGGAGGGAAACGCTCGGGCTCCGTTAGTAAGCGGTCGCAGTAATCAGGCAGACTATGGCCCACCGAAGTTTGACTATCATGTTACGCTCTGGAATAAAAGCTAAATGGCCATCAACCTAAAACTCGGGAAACCCGGTTGATAGTTGATAAAGTCGCGAATCATGGAACAATTAATCGAAATTGTCCCAAAGCACACCAGGAAATGTGACAGCCTCAGGTATTAAGCGTTTATTGTTATGCACCCGTCTCTCGTTTATGGTGCATTGACTAACAGTTCGGCGGGTACTGGTCTAAGGGCCTGCGTGGGTTATCGTGAAGCGCGCATAAAACAATGTATGACTACACTATCTCGAACAGAAATCCGGGGGCCAAACGGGCTTACTATAATAGTCCCTGTTCTTCCAAGTAGCGAACTCGGACTCGCTTTTGCTCCATCACCGCGTATCCTCTTGAGTATCTCGGGCGTTTTAGTACTAATGTTACACCATTGATCGGTCTCTGTCTACTTTCTGCTCCTTAATGTTGAAGTCCAGAATTTGTGTGATGTCCCCACACGGAACCTTCGCATGCTTTTGCAGGTGCTAATGTGCTAAGTCTATCGTCCGGGAGGGTGATCAACGTAACTTGACCATATCCTGGGTAACATACTTGAACGCTTCTACCTTCGTTCGAACTACTTAAAAGCTCCAGTTGCATAGGTCACTACCTCGCGTCTTGAAATTTCTCAGGCACTCGCACCACGCTGGCGCGGCTCAAGCAATGCGACTCTACGATGCATTATCACGTGTCTCCGACCAAGATATCATAGCCGATCAGGTTCTCGGTTATCGTGCTACATTAAAACTGGCTGAAAGTTAAGCGCCAGAGACGGAACACCCGTCCGAGCATAGACTCGTGAGGAGGGTAAAACTGAGTCCGGATCAGCCGTTTATTGGTTTAGATATGCCTCACGGTACTATAGTCTGCCCCCCAGTAATGTTATGCGGACATCGGACGCCGGGCCGAGGTATCTATCACACAAGGCTTCATATCCTTGCTAAATAAGCGTGGCGCGCTGGATGAATTTGGTGGCCAGCAGTCGCCCGTGTTCCGCCGTCTTTTATTTTAACCCAGTATATAGCAAAGTCGACAGTCAAGTCCAGGCGCCATCCATGAAGTCCCTATCCCCAGTACAATGAAACACAGCTGATGGACACCTTCTCGCCAAGTCGCATTCCTGGGATCCGAACCTCGGATCGTGCGCGAACGACAACTGAGCCGACCACAGATGATGGACATAGCTACTCCCGCGAATCTCGTTAACCGGATGATACCCTCCGTGCTGGGAAACGGCCTGGGAGCGTAAACTGACATGTCAGTGAAACTCCTAGCGGCTACCGTGTTCGAGTACCTGAACAACACGTGATACGGAAAAAATAATGTTCCCTATCGTACCGGTGAAGGGCTCTCTTGCTCAGCATTGCCTATTGAAGCGTCATCTAAATTGGGATGTCGCGTTTGTTGGGTGTCAAAGCGGCCTACCAAAGAAATTCCGGACCGAATAACACTTTCTTTTTAATACAGATTCAATAAGTCGATCATATCCTAGACCGCACTTTTCCCGGTTGCCGAGTATGAGCGTACTCGTGGAAAGGGGCGAGTAGGACGCAGGCGCCCCCACATAACTCTTGCTTACCCGCCCGTGCGGTAACCTTTATCAGCGGGGAGATGAGATTTCTCGCCGAGGGCGACAGGAACTATCCGGGGAGCGAACGCGTCGCCCATAGGTCCGGGAAATAGGACTTCGAACTGCCACTTGTCCCCGCTCGTAATTAATGGCGGTACTGTATGGTATAACAAAAGTAAACGGTTAGCGTAGATCTTAAGCGTCCTCCCGTATGAGCACCTGAAGGATGACATGATCACGCCTCATTCCGCTGTTCCATGTTCGTCTTGAATCCTTGTCACTCCCATTAGTGCTCGACTTCCCCATTTTACCGTTACGGTTCCACGCTCCCACCCTCGATTTCGCTCTTTGCGCGCCCCGCGGAGGGATAGAAGCCATCATACTCGCAGACTCGCACTGCGTCCACTCGAATGGCACGGTTCGCCAGTCATCGCTCTACGCTACGCTCCTGAATTAAGCGTTTGCCGAAGGGCACTGAACTTTACTGCTAGACCGAACTAATAGTGCTTGGTGCCGGTTAATCGTTCATCCCCTGCACAAAGCCGGTGCATCGGCCCCCTACCACTCCACCTCTTAGCGGACGTACGAACTTAGAAGGAATTGCGTGCAACTCTTGACTGCAGAGCGCAGCGAGTTCTTCGAGGGCCCAACGCTGATTCGACCGAGCCACTAATCATCAATTGAGCTGGCAGCCGCTAATTGATTTCCGCGACACTGTTACTGGGACTCCCAATGTCGCCTGTAAAACTCTAGTTAGGTCCCGTGACGTATCGTATAATAGCTAAATGTCATGGTGGCTTTCCCTAAGCTATAAGGGTACGCGCGCCAGTAGGTTACCGCGCCACTATACTACGGCTAAGCCGGGCCAATGCCATCCTGTTCCGTTTGTTAAGTATTGCGTTCATTTATATATAATACGCCCGGGTAAGTTTAGCAGGAGATGTCCAGTGCCTATTATTAAGTTTAGACCCCGGAGAAGGTAACCGGAACAGTTATCACACGTTACAAGCCACCGTAGGCTGTCACTGACCCAAACTCAATTATAAGTTATCTAGACAGGCGACTGAAACCCGTTCACCTGGCGCTGTTTCGGATTTACACCATATCAGGTTTAACTAAGTCACCGCAAGCAGACTGTGACCCGCAGGTCTGGGAGCTACAGCTAAATGCGTTTCCACTATTCATGATCTGGCCACGCCTTTTTGTAGGACTTTCCGCTTTCCGCTATTGAAGGTCACGGATCTACCACACTAAGGTTGCAGCGGGCGCGCCGCCGGTCTGGGTTAGCGATATGTTCGAGCGTAAATATAACGAGATGCGGCTGGTACATGGTGATATCGTTCTGATATACGCTATGATTCAATCTCAGATATACGTATGAAACAGGACGTTCGTATACTACTACCAGAGAGCTCGTACGACAGGGACCGAAACACTGGCCCTGGGAGAGCGCGGAGGCGTTCGTGTCCGATCTCATAGGGTGGGCTTCATGACGCGGTCTCACTCTTACCCCAGTACGATCGTGTGATACTGAAGGATAGACCCTGTGCCACTCATGAGAAAGACGTTTTTACTCTTCGTTCCAACATCAACACTACGGCGTAGCTTGATTACCCTGTGACATGGTTATGAGGGCGTCGGGATACAAATCCCGCACAACCTAAGAAATTCCTGACAAAAGTTGGACTCGATTAGGGACAGTTGCGAGACTGAACTCGGCACAGAAGAAGGTTAACTGGACGTTAGTCGAGGTTAAGAAGTGAAGACTAGTCAGACTTAGATCAGGTCGCTGCGCGGTTACGCCCCTTCTAAAGAATTCTTTATCGAGACAAGAGAAGAGAGGTCTGCCATGACCGTTGTAGTCCTATAAGAAGAGCCCAGATGTAGACGTTAGTGTCCTGAACTGGAGGAGACGAGATGGACAATCGGTTAAAGAATTGACGGAAGAGGATAAGGAAAAGTGGCGATAATGGGACCTGTATCCTGGAAGTCCCCCGTCCTATCGACACTTTTCTCGGATCAGCAGATGATTGGCGTATTGGATACAATTAGGTCAGATCAAGGCCGGCTGCAACACGAGTGTACGCTTATACTACACCGCGCACGCAGGGAGGCAATCAGGGTCTGGAAGCCACTATTACCTTGTCGGTTTACCCCCATATGTTACTCCGGGTAGTTGAATCCATAAAAGGTACACAAAGATCTAGCTACGGGAAAAGTAACCAGACCAGTTGACACAGGCAGTAAGAGTGCGGTATTGGTGACCTCTTAAAAGGGAAGAGACGCCAATAGTGGCGGGGTCAGCCGGTAAAACAGGAGAAGACCTACCAGTTTTAGTCCTAGCGCCTAGATTTCCCAGGGTAGCAGGTGCTACGTACCCCCAAATTAATTCAACAATGCCCGTTACGTAACCGTTTCAACCGCCGAATCTCTGACAGTACAAGAAAGGAGGATCTGCAATGAAACTGAAAACTTGAGCAGTTGAAAACTTCAGGGGGGCCTGTTGCCCCATATCAACAAGGCATGCTCGCCACGCATCAGGATATGAGAATTCCGACAAGAAATCGAAATCATCGCCACACGCATACCTTTCTGACTACTCCGGGCGGATTTGAGTTTTTTTTACTTGGCCGTCATTTTTCTTTAGGCTGTTTTCAACTACTCACCTCGGTTTCGTGCGGTACGGAATTCGCTCGACCTGTCATCCGGGGCTTCGGACAGAACGCAATCTCCTTCTTAGACGTACGAAAGACTGTGATAGAGGCTAGCTCGTGCGGCCTCTACCGAGGGCTATTCTGGTAGATAGCGGTCCATCAGAGAGGCCGGCCACAAACGCCGCGCCGCTCTAGGTCTGACTAGATACATACATCCCACCATAACAGGACGTTGAGCGCTCCCCAGTAGAAGCTTAAAATTAGTCCCCGAAACTTATGTCGGACTCGTTAAACGAAATATGGATACAAGTAATGTCAGTTAGAGACGTTATTCGGTTCTTGGGGAAACTCGCTGTCACTTCTGCTTGATAAGTCCACAACTCGGGATCGTGAAAATCAGCCACGAAGTCCTGAATCCCCCTCGTTCTCAGGGTAGTTGATAGATCTCGCGGAACACCTTAAAACCTCAAGCGCTTTCAATGCGCAAAGGGACTACATTGACTAGTAATGCTTGGCTATACGCCTGATCCGCTGAAAAACGTGACCCCATACGGTCACAGGAGGCCGGTCCAATGGCAATCGTCACACTCGAACAGAGAACCCCTAGGAGAACCCACCTATTCAAGGGGGACATACGGACATTGTTATTGTTCTGAATACTGCGCAGCTAATGAATGGCTAGGGGAGGGACTTAAAATGAAATTGCACCGAGCGTTACTTGATGAATACGCAGCGCGTACTATAGGTTTAGTTCGCATGGCTCGCGTTGATCGTCATGAATTTGCCTTTTGTGCAGAAAGTTTTAATGGGAAGCACAAGGCAGATCTACGTCTCAGAACCGCGGGACATTATCCGTCAGATCAAGTATGGGGGTTAAATATAGGACAGCGGAGTGTGAGCCACGCGATCACCTTTTCGTAAATCCGGATCACATTCTAAAGGAACATCTAACGGCTCACGCTCCGGTAGTAAGTGGGGGGGATCATATTCACCAGCTGGCTTTTGGAGTTTTCCCAAGTTGGAGGAAGTCTGGCCGGATTTGAGATGCGATGACCTTTGGATTGTGTGGCGAGCCGACCGTTGTCGCAGCGAGATCTGGGACTGGGGGTAAGGTTCAAATGTGTCGTGGAAGGGCCCAATGGGTAGCTTAGCTTTAAAGTTGTGTGATAGAAGAGGAAGGTTACAACACGTCGCCTGACTATCGTCGCGATGGTAATCAGGCTGATTTTATCGCTGGAGAGCACACATCTAACATGCCAACCATCCAGAGCGAGCAGTGAGTGGTAATGAAAGTAGATTGAGCGGGAAGAACCTTACTATGGATAACGCTGATAGACTCAGCTCGTTTTGCACCACCATACAAGATGCCGCCAATGAGACACCGTGCGCGCCCTGTTGCTTACTGATAGTCGGATGGGTTAATCTCTTCCGCGAGCTTATGACCTTTGTCCAAGGAGCCTTCATGAGTTAAATTCTAATGGTAGCGAGTGAAGCCATTTCACCAAATTAGTTGCCGTTTAGATACCGCGGCATGCAGCCAGGGATTGAAGGGGTGGCGTTCTTCTCTAAGTGGGGCTGCTGCTATGGCCGAAGATCAGTACTATTTTAGGATCCAATCATCATCCGGCGCCTGCCGCCTGCCGCGAGCTGAAACATCCGGGGAAGAACACTCCGGTCACAGTTATCAGAACGGTTTAAGAACGGCGCGACTTTCATAATAGCGCTCGTCTCGAACAGGGTTACGTGGAGCATGGGACTTTTAACACGATGGAGTCCCCTTGTATCCTATGTAAAAACACATACGCCAGAATCGCGGGAGTCCAGAAGTCTATGTAGGAGTATTTCTGTGACTGCCCTGAATATACATTTATCGTTATGAAGATACTAACCACGCCGCATCGGAAAACGATAGACAAAGTCAAGCAGTGTAGCAACTAAGTTTTGCGTCAGCTCTGTGTCCGCATTCTCCATCATTGCCAAGTAGAGGGAGCAAATGGTACCCCCAGAATAGGCACGTGACTGCTAGTTGTTAGGCTTACAGGTACTAGTGCCCAGCCACTAATGCATAGCGACCCAGTTATTACATCCTGGTGAATCGCAGCGGATAATAGAGCAACAACTGGCGCCAGAAGACTCCACCAAACCAATATAGATAGGGGTCCGGCCTCATAAAAGATCAAATTAGAGATGCAGGATGGGGAGGCGAGGGGTGATAGACCTCACGAAGCAAACCAGGGCCACCTGGCTCACGCGTATGAGCATTCTTGAGCTCGAAACGCAGATTAGTTTCTCATTTGGACCGAGTCCCGAATTGGTGTAAAATACCGGGAAAGCGCACACACGCGGAGAACAAGAAATTAGCCGCAAGCCTGTTACTTCCGGTGTTCGTTCAGCATCAAACAATTGTTTGCCGCCCAGTGTTTAGGGGTTAGAATACACCTTTGGGCTTAGCCCGGGAGGTCGTCAGGCGTACTCACCGCGCGTGGACAGCAACAACCATCGTGACTGCCCGGCAGTATCTCCACGTTAACCTTACCCCCAGGTTGATATTTACCCTGACTCGTAAAGTGTCTAGTCAGCGGGAATGTGGTACTTCTGCCACCCACTACAGGATGTCTATAGTCGCAAAAGTTGATTAAGCTCCGAAGGGCCTTAGGTTTGCACCGGACCCTCTATTGATGATTACAGAGGAAGATCAAGTAGCTGCCCTGGGTTAAAAGCCACTTCACGCCATATAGAACATACTGGATCCTGTACTGGGTGCCCGCCAGTAGGCGGCAGTCCAGAAAGTCGCGTTGTCCATGGCTGTGGGGGCTTTAGGGTGGCAAACTAGTTGCGAAATGTAACACAGTTTCAAGTGCATACGGCAAGCCTGCCCCTCGCCTCGCCGCCACCACTCGGCGGCGCACGAGGGAGCTGTCCTATCTGGGAACGACATTAAAGCATACCCATGCTCGTGTTGAAGTACAGGCTATGATGAAGGGGAAACTCGGTAACAGCGACCCCCGTAACTCTAACGCAACCGTCCTTGAGTTAGGGCTGTCTCGGAGACACAGCAGCTCGTATTACCTTATACCTCGCTTTGCAAAAGTGAAGCGATGGGTGAGTAATAGATTGGGGTAAGTAGCTGGCTCGTGGTGATTTACAGGCATCACTAAGTACACTAGCGCACTCGAAGCGCCGTCTCTTGGGGAAGAAAAGTACGAAACGCATCTTGGTCCCCCTTACTCGTGACACGGCGCCATGCCCCCGAATAAATCTTTGCACTCACCCATTAACTTAACACTCCCCTGGCTGTCATAGAACCCGCCTGTAGGTCCGGTATCGAAGAGGCTATTCCCTGACCCTGTACCGTGCGGCACTGAACTCTCCGTTGTCGGTCTGTCACGTGCTACTCGTCCATGCAGCACCATGTACGCTGGTGTCGCGGCTCAGGCATAACCGTGCAGCGTTGCTCTTACATCGAAGGACCTTGCGTTGGCATGACACATCCAAGCGTAATTCTCGCTGCAATTGTATAAGTGTCGGCGAGAATGATTCGACCAAGGTGAATCCCTTGTAACCGTAACCTACACTCCCTAGATGACGTTTGGCAATTGAGGAATATGTACTGCTGGCCGGCACT\n"
     ]
    }
   ],
   "source": [
    "dna = 'AGTGCCGGCCAGCAGTACATATTCCTCAATTGCCAAACGTCATCTAGGGAGTGTAGGTTACGGTTACAAGGGATTCACCTTGGTCGAATCATTCTCGCCGACACTTATACAATTGCAGCGAGAATTACGCTTGGATGTGTCATGCCAACGCAAGGTCCTTCGATGTAAGAGCAACGCTGCACGGTTATGCCTGAGCCGCGACACCAGCGTACATGGTGCTGCATGGACGAGTAGCACGTGACAGACCGACAACGGAGAGTTCAGTGCCGCACGGTACAGGGTCAGGGAATAGCCTCTTCGATACCGGACCTACAGGCGGGTTCTATGACAGCCAGGGGAGTGTTAAGTTAATGGGTGAGTGCAAAGATTTATTCGGGGGCATGGCGCCGTGTCACGAGTAAGGGGGACCAAGATGCGTTTCGTACTTTTCTTCCCCAAGAGACGGCGCTTCGAGTGCGCTAGTGTACTTAGTGATGCCTGTAAATCACCACGAGCCAGCTACTTACCCCAATCTATTACTCACCCATCGCTTCACTTTTGCAAAGCGAGGTATAAGGTAATACGAGCTGCTGTGTCTCCGAGACAGCCCTAACTCAAGGACGGTTGCGTTAGAGTTACGGGGGTCGCTGTTACCGAGTTTCCCCTTCATCATAGCCTGTACTTCAACACGAGCATGGGTATGCTTTAATGTCGTTCCCAGATAGGACAGCTCCCTCGTGCGCCGCCGAGTGGTGGCGGCGAGGCGAGGGGCAGGCTTGCCGTATGCACTTGAAACTGTGTTACATTTCGCAACTAGTTTGCCACCCTAAAGCCCCCACAGCCATGGACAACGCGACTTTCTGGACTGCCGCCTACTGGCGGGCACCCAGTACAGGATCCAGTATGTTCTATATGGCGTGAAGTGGCTTTTAACCCAGGGCAGCTACTTGATCTTCCTCTGTAATCATCAATAGAGGGTCCGGTGCAAACCTAAGGCCCTTCGGAGCTTAATCAACTTTTGCGACTATAGACATCCTGTAGTGGGTGGCAGAAGTACCACATTCCCGCTGACTAGACACTTTACGAGTCAGGGTAAATATCAACCTGGGGGTAAGGTTAACGTGGAGATACTGCCGGGCAGTCACGATGGTTGTTGCTGTCCACGCGCGGTGAGTACGCCTGACGACCTCCCGGGCTAAGCCCAAAGGTGTATTCTAACCCCTAAACACTGGGCGGCAAACAATTGTTTGATGCTGAACGAACACCGGAAGTAACAGGCTTGCGGCTAATTTCTTGTTCTCCGCGTGTGTGCGCTTTCCCGGTATTTTACACCAATTCGGGACTCGGTCCAAATGAGAAACTAATCTGCGTTTCGAGCTCAAGAATGCTCATACGCGTGAGCCAGGTGGCCCTGGTTTGCTTCGTGAGGTCTATCACCCCTCGCCTCCCCATCCTGCATCTCTAATTTGATCTTTTATGAGGCCGGACCCCTATCTATATTGGTTTGGTGGAGTCTTCTGGCGCCAGTTGTTGCTCTATTATCCGCTGCGATTCACCAGGATGTAATAACTGGGTCGCTATGCATTAGTGGCTGGGCACTAGTACCTGTAAGCCTAACAACTAGCAGTCACGTGCCTATTCTGGGGGTACCATTTGCTCCCTCTACTTGGCAATGATGGAGAATGCGGACACAGAGCTGACGCAAAACTTAGTTGCTACACTGCTTGACTTTGTCTATCGTTTTCCGATGCGGCGTGGTTAGTATCTTCATAACGATAAATGTATATTCAGGGCAGTCACAGAAATACTCCTACATAGACTTCTGGACTCCCGCGATTCTGGCGTATGTGTTTTTACATAGGATACAAGGGGACTCCATCGTGTTAAAAGTCCCATGCTCCACGTAACCCTGTTCGAGACGAGCGCTATTATGAAAGTCGCGCCGTTCTTAAACCGTTCTGATAACTGTGACCGGAGTGTTCTTCCCCGGATGTTTCAGCTCGCGGCAGGCGGCAGGCGCCGGATGATGATTGGATCCTAAAATAGTACTGATCTTCGGCCATAGCAGCAGCCCCACTTAGAGAAGAACGCCACCCCTTCAATCCCTGGCTGCATGCCGCGGTATCTAAACGGCAACTAATTTGGTGAAATGGCTTCACTCGCTACCATTAGAATTTAACTCATGAAGGCTCCTTGGACAAAGGTCATAAGCTCGCGGAAGAGATTAACCCATCCGACTATCAGTAAGCAACAGGGCGCGCACGGTGTCTCATTGGCGGCATCTTGTATGGTGGTGCAAAACGAGCTGAGTCTATCAGCGTTATCCATAGTAAGGTTCTTCCCGCTCAATCTACTTTCATTACCACTCACTGCTCGCTCTGGATGGTTGGCATGTTAGATGTGTGCTCTCCAGCGATAAAATCAGCCTGATTACCATCGCGACGATAGTCAGGCGACGTGTTGTAACCTTCCTCTTCTATCACACAACTTTAAAGCTAAGCTACCCATTGGGCCCTTCCACGACACATTTGAACCTTACCCCCAGTCCCAGATCTCGCTGCGACAACGGTCGGCTCGCCACACAATCCAAAGGTCATCGCATCTCAAATCCGGCCAGACTTCCTCCAACTTGGGAAAACTCCAAAAGCCAGCTGGTGAATATGATCCCCCCCACTTACTACCGGAGCGTGAGCCGTTAGATGTTCCTTTAGAATGTGATCCGGATTTACGAAAAGGTGATCGCGTGGCTCACACTCCGCTGTCCTATATTTAACCCCCATACTTGATCTGACGGATAATGTCCCGCGGTTCTGAGACGTAGATCTGCCTTGTGCTTCCCATTAAAACTTTCTGCACAAAAGGCAAATTCATGACGATCAACGCGAGCCATGCGAACTAAACCTATAGTACGCGCTGCGTATTCATCAAGTAACGCTCGGTGCAATTTCATTTTAAGTCCCTCCCCTAGCCATTCATTAGCTGCGCAGTATTCAGAACAATAACAATGTCCGTATGTCCCCCTTGAATAGGTGGGTTCTCCTAGGGGTTCTCTGTTCGAGTGTGACGATTGCCATTGGACCGGCCTCCTGTGACCGTATGGGGTCACGTTTTTCAGCGGATCAGGCGTATAGCCAAGCATTACTAGTCAATGTAGTCCCTTTGCGCATTGAAAGCGCTTGAGGTTTTAAGGTGTTCCGCGAGATCTATCAACTACCCTGAGAACGAGGGGGATTCAGGACTTCGTGGCTGATTTTCACGATCCCGAGTTGTGGACTTATCAAGCAGAAGTGACAGCGAGTTTCCCCAAGAACCGAATAACGTCTCTAACTGACATTACTTGTATCCATATTTCGTTTAACGAGTCCGACATAAGTTTCGGGGACTAATTTTAAGCTTCTACTGGGGAGCGCTCAACGTCCTGTTATGGTGGGATGTATGTATCTAGTCAGACCTAGAGCGGCGCGGCGTTTGTGGCCGGCCTCTCTGATGGACCGCTATCTACCAGAATAGCCCTCGGTAGAGGCCGCACGAGCTAGCCTCTATCACAGTCTTTCGTACGTCTAAGAAGGAGATTGCGTTCTGTCCGAAGCCCCGGATGACAGGTCGAGCGAATTCCGTACCGCACGAAACCGAGGTGAGTAGTTGAAAACAGCCTAAAGAAAAATGACGGCCAAGTAAAAAAAACTCAAATCCGCCCGGAGTAGTCAGAAAGGTATGCGTGTGGCGATGATTTCGATTTCTTGTCGGAATTCTCATATCCTGATGCGTGGCGAGCATGCCTTGTTGATATGGGGCAACAGGCCCCCCTGAAGTTTTCAACTGCTCAAGTTTTCAGTTTCATTGCAGATCCTCCTTTCTTGTACTGTCAGAGATTCGGCGGTTGAAACGGTTACGTAACGGGCATTGTTGAATTAATTTGGGGGTACGTAGCACCTGCTACCCTGGGAAATCTAGGCGCTAGGACTAAAACTGGTAGGTCTTCTCCTGTTTTACCGGCTGACCCCGCCACTATTGGCGTCTCTTCCCTTTTAAGAGGTCACCAATACCGCACTCTTACTGCCTGTGTCAACTGGTCTGGTTACTTTTCCCGTAGCTAGATCTTTGTGTACCTTTTATGGATTCAACTACCCGGAGTAACATATGGGGGTAAACCGACAAGGTAATAGTGGCTTCCAGACCCTGATTGCCTCCCTGCGTGCGCGGTGTAGTATAAGCGTACACTCGTGTTGCAGCCGGCCTTGATCTGACCTAATTGTATCCAATACGCCAATCATCTGCTGATCCGAGAAAAGTGTCGATAGGACGGGGGACTTCCAGGATACAGGTCCCATTATCGCCACTTTTCCTTATCCTCTTCCGTCAATTCTTTAACCGATTGTCCATCTCGTCTCCTCCAGTTCAGGACACTAACGTCTACATCTGGGCTCTTCTTATAGGACTACAACGGTCATGGCAGACCTCTCTTCTCTTGTCTCGATAAAGAATTCTTTAGAAGGGGCGTAACCGCGCAGCGACCTGATCTAAGTCTGACTAGTCTTCACTTCTTAACCTCGACTAACGTCCAGTTAACCTTCTTCTGTGCCGAGTTCAGTCTCGCAACTGTCCCTAATCGAGTCCAACTTTTGTCAGGAATTTCTTAGGTTGTGCGGGATTTGTATCCCGACGCCCTCATAACCATGTCACAGGGTAATCAAGCTACGCCGTAGTGTTGATGTTGGAACGAAGAGTAAAAACGTCTTTCTCATGAGTGGCACAGGGTCTATCCTTCAGTATCACACGATCGTACTGGGGTAAGAGTGAGACCGCGTCATGAAGCCCACCCTATGAGATCGGACACGAACGCCTCCGCGCTCTCCCAGGGCCAGTGTTTCGGTCCCTGTCGTACGAGCTCTCTGGTAGTAGTATACGAACGTCCTGTTTCATACGTATATCTGAGATTGAATCATAGCGTATATCAGAACGATATCACCATGTACCAGCCGCATCTCGTTATATTTACGCTCGAACATATCGCTAACCCAGACCGGCGGCGCGCCCGCTGCAACCTTAGTGTGGTAGATCCGTGACCTTCAATAGCGGAAAGCGGAAAGTCCTACAAAAAGGCGTGGCCAGATCATGAATAGTGGAAACGCATTTAGCTGTAGCTCCCAGACCTGCGGGTCACAGTCTGCTTGCGGTGACTTAGTTAAACCTGATATGGTGTAAATCCGAAACAGCGCCAGGTGAACGGGTTTCAGTCGCCTGTCTAGATAACTTATAATTGAGTTTGGGTCAGTGACAGCCTACGGTGGCTTGTAACGTGTGATAACTGTTCCGGTTACCTTCTCCGGGGTCTAAACTTAATAATAGGCACTGGACATCTCCTGCTAAACTTACCCGGGCGTATTATATATAAATGAACGCAATACTTAACAAACGGAACAGGATGGCATTGGCCCGGCTTAGCCGTAGTATAGTGGCGCGGTAACCTACTGGCGCGCGTACCCTTATAGCTTAGGGAAAGCCACCATGACATTTAGCTATTATACGATACGTCACGGGACCTAACTAGAGTTTTACAGGCGACATTGGGAGTCCCAGTAACAGTGTCGCGGAAATCAATTAGCGGCTGCCAGCTCAATTGATGATTAGTGGCTCGGTCGAATCAGCGTTGGGCCCTCGAAGAACTCGCTGCGCTCTGCAGTCAAGAGTTGCACGCAATTCCTTCTAAGTTCGTACGTCCGCTAAGAGGTGGAGTGGTAGGGGGCCGATGCACCGGCTTTGTGCAGGGGATGAACGATTAACCGGCACCAAGCACTATTAGTTCGGTCTAGCAGTAAAGTTCAGTGCCCTTCGGCAAACGCTTAATTCAGGAGCGTAGCGTAGAGCGATGACTGGCGAACCGTGCCATTCGAGTGGACGCAGTGCGAGTCTGCGAGTATGATGGCTTCTATCCCTCCGCGGGGCGCGCAAAGAGCGAAATCGAGGGTGGGAGCGTGGAACCGTAACGGTAAAATGGGGAAGTCGAGCACTAATGGGAGTGACAAGGATTCAAGACGAACATGGAACAGCGGAATGAGGCGTGATCATGTCATCCTTCAGGTGCTCATACGGGAGGACGCTTAAGATCTACGCTAACCGTTTACTTTTGTTATACCATACAGTACCGCCATTAATTACGAGCGGGGACAAGTGGCAGTTCGAAGTCCTATTTCCCGGACCTATGGGCGACGCGTTCGCTCCCCGGATAGTTCCTGTCGCCCTCGGCGAGAAATCTCATCTCCCCGCTGATAAAGGTTACCGCACGGGCGGGTAAGCAAGAGTTATGTGGGGGCGCCTGCGTCCTACTCGCCCCTTTCCACGAGTACGCTCATACTCGGCAACCGGGAAAAGTGCGGTCTAGGATATGATCGACTTATTGAATCTGTATTAAAAAGAAAGTGTTATTCGGTCCGGAATTTCTTTGGTAGGCCGCTTTGACACCCAACAAACGCGACATCCCAATTTAGATGACGCTTCAATAGGCAATGCTGAGCAAGAGAGCCCTTCACCGGTACGATAGGGAACATTATTTTTTCCGTATCACGTGTTGTTCAGGTACTCGAACACGGTAGCCGCTAGGAGTTTCACTGACATGTCAGTTTACGCTCCCAGGCCGTTTCCCAGCACGGAGGGTATCATCCGGTTAACGAGATTCGCGGGAGTAGCTATGTCCATCATCTGTGGTCGGCTCAGTTGTCGTTCGCGCACGATCCGAGGTTCGGATCCCAGGAATGCGACTTGGCGAGAAGGTGTCCATCAGCTGTGTTTCATTGTACTGGGGATAGGGACTTCATGGATGGCGCCTGGACTTGACTGTCGACTTTGCTATATACTGGGTTAAAATAAAAGACGGCGGAACACGGGCGACTGCTGGCCACCAAATTCATCCAGCGCGCCACGCTTATTTAGCAAGGATATGAAGCCTTGTGTGATAGATACCTCGGCCCGGCGTCCGATGTCCGCATAACATTACTGGGGGGCAGACTATAGTACCGTGAGGCATATCTAAACCAATAAACGGCTGATCCGGACTCAGTTTTACCCTCCTCACGAGTCTATGCTCGGACGGGTGTTCCGTCTCTGGCGCTTAACTTTCAGCCAGTTTTAATGTAGCACGATAACCGAGAACCTGATCGGCTATGATATCTTGGTCGGAGACACGTGATAATGCATCGTAGAGTCGCATTGCTTGAGCCGCGCCAGCGTGGTGCGAGTGCCTGAGAAATTTCAAGACGCGAGGTAGTGACCTATGCAACTGGAGCTTTTAAGTAGTTCGAACGAAGGTAGAAGCGTTCAAGTATGTTACCCAGGATATGGTCAAGTTACGTTGATCACCCTCCCGGACGATAGACTTAGCACATTAGCACCTGCAAAAGCATGCGAAGGTTCCGTGTGGGGACATCACACAAATTCTGGACTTCAACATTAAGGAGCAGAAAGTAGACAGAGACCGATCAATGGTGTAACATTAGTACTAAAACGCCCGAGATACTCAAGAGGATACGCGGTGATGGAGCAAAAGCGAGTCCGAGTTCGCTACTTGGAAGAACAGGGACTATTATAGTAAGCCCGTTTGGCCCCCGGATTTCTGTTCGAGATAGTGTAGTCATACATTGTTTTATGCGCGCTTCACGATAACCCACGCAGGCCCTTAGACCAGTACCCGCCGAACTGTTAGTCAATGCACCATAAACGAGAGACGGGTGCATAACAATAAACGCTTAATACCTGAGGCTGTCACATTTCCTGGTGTGCTTTGGGACAATTTCGATTAATTGTTCCATGATTCGCGACTTTATCAACTATCAACCGGGTTTCCCGAGTTTTAGGTTGATGGCCATTTAGCTTTTATTCCAGAGCGTAACATGATAGTCAAACTTCGGTGGGCCATAGTCTGCCTGATTACTGCGACCGCTTACTAACGGAGCCCGAGCGTTTCCCTCCGTATCACTCCTCCATAGTCAATAAAGCTAAGAGTACGCGCTTCTCCACAACAGCGCTTTCGGAGCGTTTTGTGTTCATCCGCGGGCTTGATGTGCATGTTATAC'\n",
    "print(reverse_complement(dna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "210f9b44-1968-4576-9c99-2211431b87ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clump_finder(text, k, L, t):\n",
    "    \n",
    "    frequent_k_mers = set()\n",
    "    n = len(text)\n",
    "    if n < L:\n",
    "        return frequent_k_mers\n",
    "\n",
    "    # 1. Compute the frequency table for the first window\n",
    "    first_window = text[:L]\n",
    "    freq_map = frequency_table(first_window, k)\n",
    "\n",
    "    # 2. Check for clumps in the first window\n",
    "    for kmer, count in freq_map.items():\n",
    "        if count >= t:\n",
    "            frequent_k_mers.add(kmer)\n",
    "\n",
    "    # 3. Slide the window and update the frequency table\n",
    "    for i in range(1, n - L + 1):\n",
    "        # Remove the k-mer that is leaving the window\n",
    "        old_kmer = text[i - 1:i - 1 + k]\n",
    "        freq_map[old_kmer] -= 1\n",
    "        if freq_map[old_kmer] == 0:\n",
    "            del freq_map[old_kmer]  # Remove k-mer if its count reaches 0\n",
    "\n",
    "        # Add the k-mer that is entering the window\n",
    "        new_kmer = text[i + L - k:i + L]\n",
    "        if new_kmer in freq_map:\n",
    "            freq_map[new_kmer] += 1\n",
    "        else:\n",
    "            freq_map[new_kmer] = 1\n",
    "\n",
    "        # Check for clumps in the current window\n",
    "        if freq_map[new_kmer] >= t:\n",
    "            frequent_k_mers.add(new_kmer)\n",
    "\n",
    "    return frequent_k_mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "3ff3d82e-3c06-4078-bf21-1b440a247c03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k-mers forming clumps: set()\n"
     ]
    }
   ],
   "source": [
    "text = 'GTAGGTCAAACCGGGTGTACATACCCGCTCAATCGCCCAGCACTTCGGGCAGATCACCGGGTTTCCCCGGTATCACCAATACTGCCACCAAACACAGCAGGCGGGAAGGGGCGAAAGTCCCTTATCCGACAATAAAACTTCGCTTGTTCGACGCCCGGTTCACCCGATATGCACGGCGCCCAGCCATTCGTGACCGACGTCCCCAGCCCCAAGGCCGAACGACCCTAGGAGCCACGAGCAATTCACAGCG '\n",
    "k = 20\n",
    "L = 250\n",
    "t = 3\n",
    "clumps = clump_finder(text, k, L, t)\n",
    "print(f\"k-mers forming clumps: {clumps}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "7afa1a6d-6b17-4902-b78f-c39246484d76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "175\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "result = {'CCAT', 'GACT', 'AAAG', 'GATT', 'GGAT', 'GAGG', 'CATT', 'GCTC', 'GCGT', 'TCGC', 'AAGG', 'GTCT', 'CGCT', 'AGTA', 'CCTT', 'CACT', 'GCAC', 'TTGA', 'AAAC', 'TCCC', 'GAGT', 'GTAT', 'TAGT', 'GCTG', 'ATTC', 'ACTC', 'GCCG', 'AGAC', 'GAGC', 'TCTT', 'CAGC', 'TCTA', 'AGAA', 'CTCT', 'AATA', 'ATGT', 'AACC', 'GTGA', 'GCTT', 'CAGT', 'GAAT', 'ACGA', 'AACT', 'CCGG', 'ATAA', 'GGGA', 'GCGG', 'TGGT', 'TGTT', 'CTGA', 'ACTA', 'TCAA', 'GCGA', 'AGAG', 'AACG', 'ACGC', 'TGCT', 'ATAT', 'CCCG', 'ACAC', 'GGTT', 'ACCT', 'GGGG', 'TCCG', 'TAGG', 'CCGT', 'CCAA', 'CTGG', 'GTGG', 'CCCA', 'TTGT', 'GCCA', 'AATT', 'CGGT', 'TAGC', 'TATT', 'CGGA', 'TCAT', 'AAAT', 'CCAG', 'GAGA', 'CTGC', 'ATCT', 'CCAC', 'GCTA', 'CTGT', 'AGCT', 'AGCG', 'CTTT', 'ATCA', 'TCTC', 'CGAT', 'AAGC', 'TCCT', 'ATCC', 'ACGT', 'TAAT', 'TGAT', 'AGCA', 'GCGC', 'GTGC', 'CGTT', 'TCGG', 'AGGC', 'CCTG', 'ACAA', 'AGGG', 'CCTC', 'AACA', 'GTGT', 'CGGG', 'GTTT', 'ACCG', 'TCTG', 'CTAT', 'CAAT', 'TACT', 'ATTA', 'GGCT', 'ATCG', 'TTAT', 'TGGC', 'ATGG', 'TGGG', 'AAGT', 'GCCC', 'ATAC', 'TTCT', 'ACCA', 'AGCC', 'ATTT', 'CCCC', 'CCTA', 'CCCT', 'ATGA', 'ACCC', 'CCGC', 'CAGA', 'TCAG', 'ATTG', 'AGAT', 'TTGG', 'AAAA', 'ACTG', 'TCGT', 'TCCA', 'AGTC', 'ACGG', 'AATG', 'TCAC', 'AGGT', 'GCAG', 'AATC', 'ACAT', 'GGGT', 'AAGA', 'AGTG', 'TAGA', 'CAGG', 'TGGA', 'ATAG', 'TCGA', 'GCAA', 'CGGC', 'GGGC', 'GCAT', 'ATGC', 'AGTT', 'TTGC', 'AGGA', 'TTTT', 'ACTT', 'CCGA', 'ACAG', 'GCCT'}\n",
    "print(len(result))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "e75a14fb-9140-4e4a-b17b-c50d3cd0accc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "escherichia_cl = r\"C:\\Users\\User\\Downloads\\E_coli.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5e7aad4-86c1-4082-a725-30ec4a84dcf7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "8fe2904f-38f5-4d00-9e73-b76b2af64a4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readGenome(escherichia_cl):\n",
    "    e_coli = ''\n",
    "    with open(escherichia_cl, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                e_coli += line.rstrip()\n",
    "    return e_coli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "e61487c9-45a3-4e18-bba9-b3da322c8a02",
   "metadata": {},
   "outputs": [],
   "source": [
    "e_coli = readGenome(escherichia_cl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "928b8d4b-09de-41e0-bdc2-c471f36fd5ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clump_finder(text, k, L, t):\n",
    "    \"\"\"\n",
    "    Finds k-mers that form clumps in a DNA string.\n",
    "\n",
    "    Args:\n",
    "        text: The DNA string to search within.\n",
    "        k: The length of the k-mers to find.\n",
    "        L: The size of the clump window.\n",
    "        t: The minimum number of times a k-mer must appear in a clump to be considered frequent.\n",
    "\n",
    "    Returns:\n",
    "        A set of k-mers that form clumps in the DNA string.\n",
    "    \"\"\"\n",
    "    frequent_k_mers = set()\n",
    "    text_length = len(text)\n",
    "\n",
    "    for i in range(text_length - L + 1):\n",
    "        window = text[i:i + L]\n",
    "        kmer_counts = frequency_table(window, k)  # Use frequency_table to count k-mers in the window\n",
    "\n",
    "        for kmer, count in kmer_counts.items():\n",
    "            if count >= t:\n",
    "                frequent_k_mers.add(kmer)\n",
    "\n",
    "    return frequent_k_mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "0d752cb5-1141-4132-b2c7-31a4fe92c2bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def dna_skewer_efficient(dna_string):\n",
    "    \"\"\"\n",
    "    Computes the skew of a DNA string efficiently for large strings.\n",
    "\n",
    "    Args:\n",
    "        dna_string: The input DNA string.\n",
    "\n",
    "    Returns:\n",
    "        A space-separated string of skew values at each position.\n",
    "    \"\"\"\n",
    "    skew_values = [0] * (len(dna_string) + 1)\n",
    "    current_skew = 0\n",
    "    for i, nucleotide in enumerate(dna_string):\n",
    "        if nucleotide == 'G':\n",
    "            current_skew += 1\n",
    "        elif nucleotide == 'C':\n",
    "            current_skew -= 1\n",
    "        skew_values[i + 1] = current_skew\n",
    "    return \" \".join(map(str, skew_values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "68c3cad6-956e-4dcc-9d7e-1dcd0d362cc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 -1 -1 -1 0 1 2\n"
     ]
    }
   ],
   "source": [
    "dna_string = 'CATGGG'\n",
    "print(dna_skewer_efficient(dna_string))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "4e758708-1189-4ffd-88f7-d647d47fdf4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def dna_ori(dna_string):\n",
    "    \"\"\"\n",
    "    Finds the minimum skew value and the positions of bases with that skew in a DNA string.\n",
    "\n",
    "    Args:\n",
    "        dna_string: The input DNA string.\n",
    "\n",
    "    Returns:\n",
    "        A dictionary where the key is the minimum skew value and the value is a list\n",
    "        of the 0-based indices of the bases where that minimum skew occurs.\n",
    "    \"\"\"\n",
    "    skew_values = [0]\n",
    "    current_skew = 0\n",
    "    for nucleotide in dna_string:\n",
    "        if nucleotide == 'G':\n",
    "            current_skew += 1\n",
    "        elif nucleotide == 'C':\n",
    "            current_skew -= 1\n",
    "        skew_values.append(current_skew)\n",
    "\n",
    "    min_skew = min(skew_values)\n",
    "    min_skew_indices = [i for i, skew in enumerate(skew_values) if skew == min_skew]\n",
    "\n",
    "    return {min_skew: min_skew_indices}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "3f1c2f63-7a6e-495b-9e70-ce1cf75bcc14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{-5: [13, 14]}\n"
     ]
    }
   ],
   "source": [
    "dna_string = 'GCATACACTTCCCAGTAGGTACTG'\n",
    "print(dna_ori(dna_string))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "033146a9-6139-409e-b3e4-327116474154",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Hamming_dist(p, q):\n",
    "    \"\"\"\n",
    "    Computes the Hamming distance between two approximately matched DNA strings.\n",
    "\n",
    "    Args:\n",
    "        p: The first DNA string.\n",
    "        q: The second DNA string.\n",
    "\n",
    "    Returns:\n",
    "        The Hamming distance between the two strings (the number of positions\n",
    "        at which the corresponding symbols are different).\n",
    "        Returns -1 if the strings are not of equal length.\n",
    "    \"\"\"\n",
    "    if len(p) != len(q):\n",
    "        return -1  # Hamming distance is only defined for strings of equal length\n",
    "\n",
    "    distance = 0\n",
    "    for i in range(len(p)):\n",
    "        if p[i] != q[i]:\n",
    "            distance += 1\n",
    "    return distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26afd8f5-abfa-4ca5-8ac6-46c00660d8d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "748bffa3-f150-4295-9a61-554eda10639b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23\n"
     ]
    }
   ],
   "source": [
    "p = 'CAGAAAGGAAGGTCCCCATACACCGACGCACCAGTTTA'\n",
    "q = 'CACGCCGTATGCATAAACGAGCCGCACGAACCAGAGAG'\n",
    "print(Hamming_dist(p,q))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "0faba92c-8c55-46e1-989b-29bf5a896633",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_complement(text):\n",
    "    \"\"\"\n",
    "    Computes the reverse complement of a DNA string.\n",
    "\n",
    "    Args:\n",
    "        dna_string: The DNA string to be reversed and complemented.\n",
    "\n",
    "    Returns:\n",
    "        The reverse complement of the DNA string.\n",
    "    \"\"\"\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "    reversed_string = text[::-1]  # Reverse the string\n",
    "    reverse_complement_string = ''.join(complement[base] for base in reversed_string)\n",
    "    return reverse_complement_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "7c3b6ecd-a612-42a9-8a3e-e5cce2c7f2e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def neighbors(pattern, d):\n",
    "    \"\"\"\n",
    "    Generates the neighborhood of a pattern with up to d mismatches.\n",
    "\n",
    "    Args:\n",
    "        pattern: The input string.\n",
    "        d: The maximum number of mismatches.\n",
    "\n",
    "    Returns:\n",
    "        A set of all strings within Hamming distance d of pattern.\n",
    "    \"\"\"\n",
    "    if d == 0:\n",
    "        return {pattern}\n",
    "    if len(pattern) == 1:\n",
    "        return {'A', 'C', 'G', 'T'}\n",
    "    neighborhood = set()\n",
    "    suffix_neighbors = neighbors(pattern[1:], d)\n",
    "    for suffix_neighbor in suffix_neighbors:\n",
    "        for nucleotide in ['A', 'C', 'G', 'T']:\n",
    "            if nucleotide == pattern[0]:\n",
    "                neighborhood.add(nucleotide + suffix_neighbor)\n",
    "            else:\n",
    "                if d > 0:\n",
    "                    neighborhood.add(nucleotide + suffix_neighbor)\n",
    "    return neighborhood"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "ae0806de-a97f-4db0-98cb-0b395e7b751a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def hamming_distance(str1, str2):\n",
    "    \n",
    "    if len(str1) != len(str2):\n",
    "        return -1  # Hamming distance is only defined for equal-length strings\n",
    "    distance = 0\n",
    "    for i in range(len(str1)):\n",
    "        if str1[i] != str2[i]:\n",
    "            distance += 1\n",
    "    return distance\n",
    "\n",
    "def approximate_pattern_count(text, pattern, d):\n",
    "    \"\"\"\n",
    "    Counts the number of approximate occurrences of a pattern in a text,\n",
    "    allowing up to d mismatches.\n",
    "\n",
    "    Args:\n",
    "        text: The text string to search within.\n",
    "        pattern: The pattern string to search for.\n",
    "        d: The maximum number of mismatches allowed.\n",
    "\n",
    "    Returns:\n",
    "        The number of times the pattern appears in the text with at most d mismatches.\n",
    "    \"\"\"\n",
    "    count = 0\n",
    "    for i in range(len(text) - len(pattern) + 1):\n",
    "        pattern_prime = text[i:i + len(pattern)]  # Extract substring of text\n",
    "        if hamming_distance(pattern, pattern_prime) <= d:\n",
    "            count += 1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "b114ace2-c9bd-43bc-9938-2d70427c300f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def frequent_words_with_mismatches(text, k, d):\n",
    "    \n",
    "    freq_map = {}\n",
    "    combined_text = text + reverse_complement(text)\n",
    "    n = len(combined_text)\n",
    "    for i in range(n - k + 1):\n",
    "        pattern = combined_text[i:i + k]\n",
    "        neighborhood = neighbors(pattern, d)\n",
    "        for neighbor in neighborhood:\n",
    "            if neighbor not in freq_map:\n",
    "                freq_map[neighbor] = 1\n",
    "            else:\n",
    "                freq_map[neighbor] += 1\n",
    "    m = max(freq_map.values()) if freq_map else 0  # Handle empty freq_map\n",
    "\n",
    "    frequent_patterns = []\n",
    "    for pattern, frequency in freq_map.items():\n",
    "        if frequency == m:\n",
    "            frequent_patterns.append(pattern)\n",
    "\n",
    "    # Find the pattern(s) with the highest approximate_pattern_count\n",
    "    max_approx_count = -1  # Initialize to a value that will be less than any count\n",
    "    most_frequent_patterns = []\n",
    "\n",
    "    for pattern in frequent_patterns:\n",
    "        approx_count = approximate_pattern_count(combined_text, pattern, d)\n",
    "        if approx_count > max_approx_count:\n",
    "            max_approx_count = approx_count\n",
    "            most_frequent_patterns = [pattern]  # Start a new list with the new most frequent pattern\n",
    "        elif approx_count == max_approx_count:\n",
    "            most_frequent_patterns.append(pattern)  # Add to the list if it's as frequent as the current most frequent\n",
    "\n",
    "    return most_frequent_patterns # Return the list of most frequent patterns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "5be638be-50f9-4c35-aeba-7c06d6473b9f",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "' '",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[88], line 4\u001b[0m\n\u001b[0;32m      2\u001b[0m k \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m20\u001b[39m\n\u001b[0;32m      3\u001b[0m d \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m4\u001b[39m\n\u001b[1;32m----> 4\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mfrequent_words_with_mismatches\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtext\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43md\u001b[49m\u001b[43m)\u001b[49m)\n",
      "Cell \u001b[1;32mIn[87], line 4\u001b[0m, in \u001b[0;36mfrequent_words_with_mismatches\u001b[1;34m(text, k, d)\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mfrequent_words_with_mismatches\u001b[39m(text, k, d):\n\u001b[0;32m      3\u001b[0m     freq_map \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m----> 4\u001b[0m     combined_text \u001b[38;5;241m=\u001b[39m text \u001b[38;5;241m+\u001b[39m \u001b[43mreverse_complement\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtext\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m      5\u001b[0m     n \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(combined_text)\n\u001b[0;32m      6\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(n \u001b[38;5;241m-\u001b[39m k \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m):\n",
      "Cell \u001b[1;32mIn[84], line 13\u001b[0m, in \u001b[0;36mreverse_complement\u001b[1;34m(text)\u001b[0m\n\u001b[0;32m     11\u001b[0m complement \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mT\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mT\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mC\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mG\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mG\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mC\u001b[39m\u001b[38;5;124m'\u001b[39m}\n\u001b[0;32m     12\u001b[0m reversed_string \u001b[38;5;241m=\u001b[39m text[::\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]  \u001b[38;5;66;03m# Reverse the string\u001b[39;00m\n\u001b[1;32m---> 13\u001b[0m reverse_complement_string \u001b[38;5;241m=\u001b[39m \u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcomplement\u001b[49m\u001b[43m[\u001b[49m\u001b[43mbase\u001b[49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mbase\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mreversed_string\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     14\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m reverse_complement_string\n",
      "Cell \u001b[1;32mIn[84], line 13\u001b[0m, in \u001b[0;36m<genexpr>\u001b[1;34m(.0)\u001b[0m\n\u001b[0;32m     11\u001b[0m complement \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mT\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mT\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mC\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mG\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mG\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mC\u001b[39m\u001b[38;5;124m'\u001b[39m}\n\u001b[0;32m     12\u001b[0m reversed_string \u001b[38;5;241m=\u001b[39m text[::\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]  \u001b[38;5;66;03m# Reverse the string\u001b[39;00m\n\u001b[1;32m---> 13\u001b[0m reverse_complement_string \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(\u001b[43mcomplement\u001b[49m\u001b[43m[\u001b[49m\u001b[43mbase\u001b[49m\u001b[43m]\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m base \u001b[38;5;129;01min\u001b[39;00m reversed_string)\n\u001b[0;32m     14\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m reverse_complement_string\n",
      "\u001b[1;31mKeyError\u001b[0m: ' '"
     ]
    }
   ],
   "source": [
    "text = 'GTAGGTCAAACCGGGTGTACATACCCGCTCAATCGCCCAGCACTTCGGGCAGATCACCGGGTTTCCCCGGTATCACCAATACTGCCACCAAACACAGCAGGCGGGAAGGGGCGAAAGTCCCTTATCCGACAATAAAACTTCGCTTGTTCGACGCCCGGTTCACCCGATATGCACGGCGCCCAGCCATTCGTGACCGACGTCCCCAGCCCCAAGGCCGAACGACCCTAGGAGCCACGAGCAATTCACAGCG '\n",
    "k = 20\n",
    "d = 4\n",
    "print(frequent_words_with_mismatches(text, k, d))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c844a95-8019-4b10-b522-ca4c842e989d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def immediate_neighbors(pattern, d):\n",
    "    \"\"\"\n",
    "    Generates the immediate neighbors of a pattern with up to d mismatches.\n",
    "\n",
    "    Args:\n",
    "        pattern: The input string.\n",
    "        d: The maximum number of mismatches allowed.\n",
    "\n",
    "    Returns:\n",
    "        A set of strings that are immediate neighbors of pattern.\n",
    "    \"\"\"\n",
    "    neighborhood = {pattern}\n",
    "    if d == 0:\n",
    "        return neighborhood\n",
    "    for i in range(len(pattern)):\n",
    "        symbol = pattern[i]\n",
    "        for nucleotide in ['A', 'C', 'G', 'T']:\n",
    "            if nucleotide != symbol:\n",
    "                neighbor = pattern[:i] + nucleotide + pattern[i+1:]\n",
    "                neighborhood.add(neighbor)\n",
    "    return neighborhood\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a94430e-e803-473d-a36c-47593b67da9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern = 'ACGT'\n",
    "d = 3\n",
    "print(immediate_neighbors(pattern, d))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1c25e34-4d38-4ad1-a95f-48a325c17a30",
   "metadata": {},
   "outputs": [],
   "source": [
    "def neighbors(pattern, d):\n",
    "    \"\"\"\n",
    "    Generates the neighborhood of a pattern with up to d mismatches.\n",
    "\n",
    "    Args:\n",
    "        pattern: The input string.\n",
    "        d: The maximum number of mismatches.\n",
    "\n",
    "    Returns:\n",
    "        A set of all strings within Hamming distance d of pattern.\n",
    "    \"\"\"\n",
    "    if d == 0:\n",
    "        return {pattern}\n",
    "    if len(pattern) == 1:\n",
    "        return {'A', 'C', 'G', 'T'}\n",
    "    neighborhood = set()\n",
    "    suffix_neighbors = neighbors(pattern[1:], d)\n",
    "    for suffix_neighbor in suffix_neighbors:\n",
    "        if hamming_distance(pattern[1:], suffix_neighbor) < d:\n",
    "            for nucleotide in ['A', 'C', 'G', 'T']:\n",
    "                neighborhood.add(nucleotide + suffix_neighbor)\n",
    "        else:\n",
    "            neighborhood.add(pattern[0] + suffix_neighbor)\n",
    "    return neighborhood\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ca15dd6-fef5-438c-9523-c6367e21e542",
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern = 'ACGT'\n",
    "d = 3\n",
    "print(neighbors(pattern, d))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49ba33e0-59a1-40ee-a676-0d1309bc5612",
   "metadata": {},
   "outputs": [],
   "source": [
    "def motif_enumeration(dna, k, d):\n",
    "    \"\"\"\n",
    "    Finds all (k, d)-motifs in a collection of DNA strings.\n",
    "\n",
    "    Args:\n",
    "        dna: A list of DNA strings.\n",
    "        k: The length of the motif (k-mer).\n",
    "        d: The maximum number of mismatches allowed.\n",
    "\n",
    "    Returns:\n",
    "        A set of strings representing the (k, d)-motifs.\n",
    "    \"\"\"\n",
    "\n",
    "    def hamming_distance(str1, str2):\n",
    "        \"\"\"\n",
    "        Calculates the Hamming distance between two strings.\n",
    "\n",
    "        Args:\n",
    "            str1: The first string.\n",
    "            str2: The second string.\n",
    "\n",
    "        Returns:\n",
    "            The Hamming distance between the two strings.\n",
    "        \"\"\"\n",
    "        if len(str1) != len(str2):\n",
    "            return float('inf')  # Or raise an exception, depending on desired behavior\n",
    "        distance = 0\n",
    "        for i in range(len(str1)):\n",
    "            if str1[i] != str2[i]:\n",
    "                distance += 1\n",
    "        return distance\n",
    "\n",
    "    def neighbors(pattern, d):\n",
    "        \"\"\"\n",
    "        Generates all neighbors of a pattern with at most d mismatches.\n",
    "\n",
    "        Args:\n",
    "            pattern: The input string (k-mer).\n",
    "            d: The maximum number of mismatches.\n",
    "\n",
    "        Returns:\n",
    "            A list of strings representing the neighbors.\n",
    "        \"\"\"\n",
    "        if d == 0:\n",
    "            return [pattern]\n",
    "        if len(pattern) == 1:\n",
    "            return ['A', 'C', 'G', 'T']\n",
    "        neighborhood = []\n",
    "        suffix_neighbors = neighbors(pattern[1:], d)\n",
    "        for text in suffix_neighbors:\n",
    "            if hamming_distance(pattern[1:], text) < d:\n",
    "                for nucleotide in ['A', 'C', 'G', 'T']:\n",
    "                    neighborhood.append(nucleotide + text)\n",
    "            else:\n",
    "                neighborhood.append(pattern[0] + text)\n",
    "        return neighborhood\n",
    "\n",
    "    patterns = set()  # Use a set to automatically handle duplicates\n",
    "    first_dna = dna[0]\n",
    "    for i in range(len(first_dna) - k + 1):\n",
    "        pattern = first_dna[i:i + k]\n",
    "        neighborhood = neighbors(pattern, d)\n",
    "        for neighbor in neighborhood:\n",
    "            found_in_all = True\n",
    "            for j in range(len(dna)):\n",
    "                found_in_current = False\n",
    "                for l in range(len(dna[j]) - k + 1):\n",
    "                    if hamming_distance(neighbor, dna[j][l:l + k]) <= d:\n",
    "                        found_in_current = True\n",
    "                        break  # Found in current DNA string, move to next\n",
    "                if not found_in_current:\n",
    "                    found_in_all = False\n",
    "                    break  # Not found in all DNA strings, no need to check others\n",
    "            if found_in_all:\n",
    "                patterns.add(neighbor)\n",
    "    return patterns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "704c2104-5788-437a-8d35-257eba7a3dba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def motif_enumeration(dna_input, k, d):\n",
    "    \"\"\"\n",
    "    Finds all (k, d)-motifs in a collection of DNA strings.\n",
    "\n",
    "    Args:\n",
    "        dna_input: A space-separated string of DNA strings.\n",
    "        k: The length of the motif (k-mer).\n",
    "        d: The maximum number of mismatches allowed.\n",
    "\n",
    "    Returns:\n",
    "        A set of strings representing the (k, d)-motifs.\n",
    "    \"\"\"\n",
    "\n",
    "    def hamming_distance(str1, str2):\n",
    "        \"\"\"\n",
    "        Calculates the Hamming distance between two strings.\n",
    "\n",
    "        Args:\n",
    "            str1: The first string.\n",
    "            str2: The second string.\n",
    "\n",
    "        Returns:\n",
    "            The Hamming distance between the two strings.\n",
    "        \"\"\"\n",
    "        if len(str1) != len(str2):\n",
    "            return float('inf')\n",
    "        distance = 0\n",
    "        for i in range(len(str1)):\n",
    "            if str1[i] != str2[i]:\n",
    "                distance += 1\n",
    "        return distance\n",
    "\n",
    "    def neighbors(pattern, d):\n",
    "        \"\"\"\n",
    "        Generates all neighbors of a pattern with at most d mismatches.\n",
    "\n",
    "        Args:\n",
    "            pattern: The input string (k-mer).\n",
    "            d: The maximum number of mismatches.\n",
    "\n",
    "        Returns:\n",
    "            A list of strings representing the neighbors.\n",
    "        \"\"\"\n",
    "        if d == 0:\n",
    "            return [pattern]\n",
    "        if len(pattern) == 1:\n",
    "            return ['A', 'C', 'G', 'T']\n",
    "        neighborhood = []\n",
    "        suffix_neighbors = neighbors(pattern[1:], d)\n",
    "        for text in suffix_neighbors:\n",
    "            if hamming_distance(pattern[1:], text) < d:\n",
    "                for nucleotide in ['A', 'C', 'G', 'T']:\n",
    "                    neighborhood.append(nucleotide + text)\n",
    "            else:\n",
    "                neighborhood.append(pattern[0] + text)\n",
    "        return neighborhood\n",
    "\n",
    "    patterns = set()\n",
    "    dna_list = dna_input.split()  # Split the input string into a list of DNA strings\n",
    "    #print(dna_list) #for testing\n",
    "    first_dna = dna_list[0] # Changed this\n",
    "    for i in range(len(first_dna) - k + 1):\n",
    "        pattern = first_dna[i:i + k]\n",
    "        neighborhood = neighbors(pattern, d)\n",
    "        for neighbor in neighborhood:\n",
    "            found_in_all = True\n",
    "            for j in range(len(dna_list)): # Changed this\n",
    "                found_in_current = False\n",
    "                for l in range(len(dna_list[j]) - k + 1): # Changed this\n",
    "                    if hamming_distance(neighbor, dna_list[j][l:l + k]) <= d: # Changed this\n",
    "                        found_in_current = True\n",
    "                        break\n",
    "                if not found_in_current:\n",
    "                    found_in_all = False\n",
    "                    break\n",
    "            if found_in_all:\n",
    "                patterns.add(neighbor)\n",
    "    return patterns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "c656c819-805e-492b-a7a1-0dae5ce6c61c",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[91], line 4\u001b[0m\n\u001b[0;32m      2\u001b[0m k \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m20\u001b[39m\n\u001b[0;32m      3\u001b[0m d \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m4\u001b[39m\n\u001b[1;32m----> 4\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mmotif_enumeration\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdna\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43md\u001b[49m\u001b[43m)\u001b[49m)\n",
      "Cell \u001b[1;32mIn[89], line 70\u001b[0m, in \u001b[0;36mmotif_enumeration\u001b[1;34m(dna_input, k, d)\u001b[0m\n\u001b[0;32m     68\u001b[0m found_in_current \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[0;32m     69\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m l \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(dna_list[j]) \u001b[38;5;241m-\u001b[39m k \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m): \u001b[38;5;66;03m# Changed this\u001b[39;00m\n\u001b[1;32m---> 70\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mhamming_distance\u001b[49m\u001b[43m(\u001b[49m\u001b[43mneighbor\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdna_list\u001b[49m\u001b[43m[\u001b[49m\u001b[43mj\u001b[49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[43ml\u001b[49m\u001b[43m:\u001b[49m\u001b[43ml\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m d: \u001b[38;5;66;03m# Changed this\u001b[39;00m\n\u001b[0;32m     71\u001b[0m         found_in_current \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m     72\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "Cell \u001b[1;32mIn[89], line 30\u001b[0m, in \u001b[0;36mmotif_enumeration.<locals>.hamming_distance\u001b[1;34m(str1, str2)\u001b[0m\n\u001b[0;32m     28\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(str1)):\n\u001b[0;32m     29\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m str1[i] \u001b[38;5;241m!=\u001b[39m str2[i]:\n\u001b[1;32m---> 30\u001b[0m         distance \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[0;32m     31\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m distance\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "dna = 'GTAGGTCAAACCGGGTGTACATACCCGCTCAATCGCCCAGCACTTCGGGCAGATCACCGGGTTTCCCCGGTATCACCAATACTGCCACCAAACACAGCAGGCGGGAAGGGGCGAAAGTCCCTTATCCGACAATAAAACTTCGCTTGTTCGACGCCCGGTTCACCCGATATGCACGGCGCCCAGCCATTCGTGACCGACGTCCCCAGCCCCAAGGCCGAACGACCCTAGGAGCCACGAGCAATTCACAGCG CTCGCTGGTCATAATGCATCCTTGCGATTGGGGTGTAACCTCCACAGACAAATCCGACCACCGCGCCGGAGCCCTCCGATAGGGCCTGAAGTCCCGGCCACTGGGGTCATTCGTCCGCCAGGGGCGAAAGGTCTGCCACATTGGGGACTTCCGGCCCTAACTGAGCCGGCGCCATCGAAGCCAGGGTTAGCGCATGGCTGTCCGGGCGGGGCGCGGAGTTCGCCGGAGTCCGAGACCCCGGATCGTGTCG GCGCCCCGCCCGGACAGCCATGCGCTAACCCTGGCTTCGATGGCGCCGGCTCAGTTAGGGCCGGAAGTCCCCAATGTGGCAGACCTTTCGCCCCTGGCGGACGAATGACCCCAGTGGCCGGGACTTCAGGCCCTATCGGAGGGCTCCGGCGCGGTGGTCGGATTTGTCTGTGGAGGTTACACCCCAATCGCAAGGATGCATTATGACCAGCGAGCTGAGCCTGGTCGCCACTGGAAAGGGGAGCAACATC GTACATGTCCAGAGCGAGCCTCAGCTTCTGCGCAGCGACGGAAACTGCCACACTCAAAGCCTACTGGGCGCACGTGTGGCAACGAGTCGATCCACACGAAATGCCGCCGTTGGGCCGCGGACTAGCCGAATTTTCCGGGTGGTGACACAGCCCACATTTGGCATGGGACTTTCGGCCCTGTCCGCGTCCGTGTCGGCCAGACAAGCTTTGGGCATTGGCCACAATCGGGCCACAATCGAAAGCCGAGCAG ACGGGACTGGCGACCGAGGTCAGCGATCCCGAGCAGGTTGCCCGCTACCAGCGGCTGCTACACCCGTGGGTGAACATGGCGATGGACACCGTGGTCGCGATCGAACCCGAGATCGTCACCGGCATCCGCATCGTTGCTGACTCGCGTACGCCGTAGCCGATTGGCCGCGGGCGGCCCGCACGCATCCGCACTATCTGATAAATTCTTCAACTCGTCAACCGATGTAACGCTGAAGCTCTCAGGAGACGCG GCATCCGCGTCCGGGCAACCAGGGACGTTTCGGCCACGTCGCAAGCGACAGCCCAGGGTGCCGGATGCCCAGCCGCTCGCGGTGGCGGGCCGTCCAGAGCCGTCGGCAGCCAGTGACCAGCTGGGGTATCCGCATGGGGTCGCCCAGCGGGTCCCGAGGGGACTTTTGGCCACCGGCGCTGGTGGCCTACTGCCCTCCCGCCGTTGCGCCGGGTGCGTGCACGATTGAAGTCCCCAAGGAAGGGACGCTC ACGATAGATCAACCGGACCACCGAGGGTTGATTATTGAGGTGCGCTCATCCGATGGTTCGCCGCCGTATGTGGTGCGCTGGCTCGAGACCGACCATGTGGCGACGGTGATTCCGGGTCCGGATGCGGTCGTGGTCACTGCGGAGGAGCAGAATGCGGCCGACGAGCGGGCGCAGCATCGGTTCGGCGCGGTTCAGTCGGCGATCCTCCATGCCAGGGGAACGTAGGCGATTCGCTCAAGCGACGAAGTCG CGATCACCTCAATCCGATGACCTATGAAGCGCGCGGGCCCGGCCGCCATCGGCCCGTCGATCCGAGTGCGCACGGCCGAAGTGAGCCACCACCGTAGCGCCGCCGAGTTCGCTTCCGCGGACGCAAGCCCGGGATTTGCGGAGTAGCGTACAAGGCGGGATGCGCAATCCGGTCGCTGAGGGACTGTCGTCAGTCGGCCCAGGGACCTAATTCCATATTTGGAGCCATGAGATGGCGACACACTCGGCAG GCAAACCGTGAGGCTAGGGAAGCGAGGAGCACATGGCCGCCGACCCGCAATGTACACGCTGCAAGCAAACCATCGAACCCGGATGGCTATACATCACCGCCCATCGCCGCGGTCAAGCCGGGATCGTCGATGACGGCGCAGTACTGATTCACGTGCCCGGTGAATGCCCGCACCCCGGGGAGCACGTTCCGCGAAGCTAGCCGGTTGGCATGCTAGGCAACGCAGCGTCAAGTGAGAGGGCAAAACCAAA CCGCTGGCGACGCTGTTCGCCGGCAGCGTGCGTGACGACTTCGAGCTGCCCGACTACACCTGGTGACCACCGCCGACGGGCACCTCTCCGCCAGGTAGGCACGGTTTGTCGCCGGCAATGTGACCTTTGGGCGCGGTCTTGAGGACCTTCGGCCCCACCCACGAGGCCGCCGCCGGCCGATCGTATGACGTGCAATGTACGCCATAGGGTGCGTGTTACGGCGATTACCTGAAGGCGGCGGTGGTCCGGA TCAGCACCATGACCGCCTGGCCACCAATCGCCCGTAACAAGCGGGACGTCCGCGACGACGCGTGCGCTAGCGCCGTGGCGGTGACAACGACCAGATATGGTCCGAGCACGCGGGCGAACCTCGTGTTCTGGCCTCGGCCAGTTGTGTAGAGCTCATCGCTGTCATCGAGCGATATCCGACCACTGATCCAAGTCGGGGGCTCTGGGGACCGAAGTCCCCGGGCTCGGAGCTATCGGACCTCACGATCACC ATGGCGATCCGCTCGATACTCCGCTGGCAGCCCGAGCGGGGGGTGTCCCCTACCCTGGCTGCCCCGGCTGACGGTATTCACCGGTATCCCATGATGCACTCAATCCCGATGCCTGCACCTTCTGTAAGCTCGCGACGGCTAACGGGGTTATCGGATGGCCCGGGCTCGAAGGAGGTTGGCACGTGTGGATGTTCCTGCCCGGTCGCCACGCGGCGGCGGCGATACCCCAGAGTCGTCAGGTGTTTAGCCG GGGTCAGGTATATTTATCGCACACTTGGGCACATGACACACAAGCGCCAGAATCCCGGACCGAACCGAGCACCGTGGGTGGGCAGCCTCCATACAGCGATGACCTGATCGATCATCGGCCAGGGCGCCGGGCTTCCAACCGTGGCCGTCTCAGTACCCAGCCTCATTGACCCTTCGACGCATCCACTGCGCGTAAGTCGGCTCAACCCTTTCAAACCGCTGGATTACCGACCGCAGAAAGGGGGCAGGAC TGTCCGCTGTCGAGTTTCGCGACATGGCCGAGACCTTTGAAGACGAGGAGCACCGGCGCTTTGGCGAGGCCGGTTTTCAATCGGTGGTCGACAAGGTCGCCGATATCGAAAAAAGCCTTGGCATCTACGACCTGAGCCAGTTCACCCCCAGCTAAAGACACTAATGCCCTTGGGTTAGGGACCATCGCCTCCTGACGCGATCGCGACAGCTGGCTAACGTCGGTAGTACACCCATGCAGAGGGGACGCCA TCCGGTTATCGTCGCCCGAATCCCCCAAGATCCGGCAGTGCCGGCCTGAGGGCCTGTGCGATCTGCTCGGGTGGTGCCCACCCGCGCGGAAAGCCCCGTCCGAACCGTGATTGGGCAACGTCGGGCCGGGCCAGCAGCGCTGGACCGTAGGTCCCTGCAGTGGATGACTTACGGCCCTGATCCACACCGGCGACCGTTAGGCAGGGTTGAGCCAACCGTCGGTTGAGCGTCTGGCTGCGAGGTGAGGTGA TTCGGCGACGGTTGACGTGATGGCCGACAGAATCGTTGCTAGAGCCGGGGGCAACTCCGACGCCACCGCCGAGATCGCGGCGGCCTTGGCCGCCCGGCAAGCGGACTGGGACACCGGGCACCGGATCGACACGGCGGGGCCGCGTGAGCGCTCCGTGGGACAGGCCTACCACATCTGGCGCAGCGCGATCTGAGAACGCCGAAAGGAAAACCGATGCCAACCATCACTGTCAGCAGCACATCGTCGCTGT CGTCCATGCCGCGGCGTGCACTCGCGTGGCCTTGTCGACCACGTTGTCGAGGCCGACGATGACAGCGTCGTGGTAGCGCCGGTCGATGGTGGGCGGCGGTCCGGTCACATTGACGACGATGCCACAGCGCACGCTTGGATTCGGGCGTCGAATCCAATCATGGTCGGTTGGCCGTCCGATTGGGGACTAAAGCCTCATGACCGGTGACTGTCCCGGTTCATAGCGTTGCTCGAGGCGAAGGGAGAATGAG CACCACGTGGACCACGGTCAGCGGAATGTTCCTCATCGCCGCATCGGTGGCACCCCAACAGGCGGCGGCATCCGATTCGAGCGAACCATCTACCCCGACGACAACTCCGTGCTGCTTGCGGGGTTTAGACATCTCATTCTCCCTTCGCCTCGAGCAACGCTATGAACCGGGACAGTCACCGGTCATGAGGCTTTAGTCCCCAATCGGACGGCCAACCGACCATGATTGGATTCGACGCCCGAATCCAAGC TTCCCGCGGATCAGATCTTGACCACCGGGAGTGTCGATGAACTTCTCGCGCTCTTGAAATGACGGGCTATCGTAAGTTTATGGCCTGGGGGAGCGTGAATCCCGCTGGCGGTCGGGTGAACCGCCCCGGTTTTCTTGCACCCCGCGTCGACGTGCCAGTGACGAACTTGACGAATAAGGCCTTTGGTCCTTTCCGGTAGGGGTCTTTGGATAGGCGCGATCCTCGGCATCGGGCCGGTAGCTTGCCGTTT GGCCAACTGCACCGCGCTCTTGATGACATCGGTGGTCACCATGGTGTCCGGCATGATCAACCTCCGCTGTTCGATATCACCCCGATCTTTCTGAACGGCGGTTGGCAGACAACAGGGTCAATGGTCCCCAAGTGGATCACCGACGGGCGCGGACAAATGGCCCGCGCTTCGGGGACTTCTGTCCCTAGCCCTGGCCACGATGGGCTGGTCGGATCAAAGGCATCCGTTTCCATCGATTAGGAGGCATCAA AAACTCGGGGAAGAGGGACCGCGGGTGGCGCTGAACGGGAAGGGTGGTGGCCATTTGATGCCTCCTAATCGATGGAAACGGATGCCTTTGATCCGACCAGCCCATCGTGGCCAGGGCTAGGGACAGAAGTCCCCGAAGCGCGGGCCATTTGTCCGCGCCCGTCGGTGATCCACTTGGGGACCATTGACCCTGTTGTCTGCCAACCGCCGTTCAGAAAGATCGGGGTGATATCGAACAGCGGAGGTTGATC CATCCATCGTGCCGCGCGTCGGCGAGTCCTGTTGATGCGCCACACATTCCGCAGGCATCGTGAACGCTTGACAGCCGTCGCCATTGTCGCGCACAAACCGCACGTCGATTCGTGATCCATTGAGGACCTAAGCCCGTTGGGCTAGTGACAAACGCCTCCTGAGCAAAACCCTCCTCCCCCGTTACCGTCGTGCGGTAGGGACAAGCCACATCGGCCGAGCGGGCGATCAGCCAACGACAGGAGGACCGCG GTCGATGCGAACGCAAGCATCCAGGAGATGCTCAACGTCATGGAAGAACATCAGGTCCGCCGTGTTCCGGTCATCTCAGAGCACCGCTTGGTCGGAATCGTCACCGAAGCCGACATCGCCCGACACCTGCCCGAGCACGCCATTGTGCAGTTCGTCAAGGCAATCTGCTCGCCCATGGCCCTCGCCAGCTAGCGACCTCGGATCATCCGCCGGCACGAAAGGCTTTATCCGCCAGCATCGGAGGTACCCA CCGATCGGCATCACTATCGGTCCTGCGGCCGCCCATAGCGCTATATCCGGCTGGTGAAATCAATTGACAACCTTCGACTTTGAGGTGGCCTACGGCGAGGACAAGCCAGGCAAGCCAGCTGCCTCAACGCGCGCCAGTACGGGTCCATCGACCCGCGGCCCACGGGTCAAACGACCCTAGTGTTCGCTACGACGTGGTCGTACCTTCGGCAGCAGATCAGCAATAGCACCCCGACTCGAGGAGGATCCCG ACCGTCGATGTGCCCGGTCGCGCCGCGTCCACCTCGGTCATCGACCCCACGATGAGGACGCCATCGGCCGCGACCAAGCCCCGTGAAACTCTGACGGCGTGCTGGCCGGGCTGCGGCACCTGATCACCTTAGGGCACTTGGGCCACCACAACGGGCCGCCGGTCTCGACAGTGGCCACCACCACACAGGTGACTTCCGGCGGGACGTAAGTCCCTAACGCGTCGTTCCGCACGCGGTTAGCTTTGCTGCC CACCTGTGTGGTGGTGGCCACTGTCGAGACCGGCGGCCCGTTGTGGTGGCCCAAGTGCCCTAAGGTGATCAGGTGCCGCAGCCCGGCCAGCACGCCGTCAGAGTTTCACGGGGCTTGGTCGCGGCCGATGGCGTCCTCATCGTGGGGTCGATGACCGAGGTGGACGCGGCGCGACCGGGCACATCGACGGTCCCCGGGGCTTTGTGGGCCAGTGAAGTGACGAAAGACCCCAGTGGACACGGACTTCGGC GTGCCACGACGCATAGAAGACATGGCCATGCCACACCCTGATAGCATTGCAGCAAGCTACATGTACTGCTCTACCAGGATCCTTATGGGCAACAGTGGGTTTGAGTTATGAAACCCGTGGGCACATACCCTTCCGCGTCGTACTGGTCAGTCTCGACAGCGAAGAGATCACCGGTTGATCCACCAAGCATGCATTGGCGGGCATCTGCATAAACGGTGACGTATCAGCACAAAACAGCGGAGAGAACAAC GCGAGTACGCTAGTTCAGGTGGGTGCCGTGCCGAAGGCGGTGTCACTCAACGAACTTCGGTTCTCGCAGGGTCGCCACGGCTGGCGATGTGCGGTAACGCTCGATGTGTGAATTGAGACCTGATTCATGAAAATCGTCGAGGAGACCCCATACCGGTTCCGGATCGAACAAGAGGGCGCGATGCGGGTGCCCGGGATCGTGTTCGCGTCCAGGTCGTTGCTGCCTCGTGACGAAGGCGACATGGCCCTTG CTTGAGGTCCAGTCCGGCGCAGAACACCGGATCGGCGCCGGTGAGGATGACGACGTCGATGTCGTCGTCGGCCTCGGCGTCGGCCAACGCCGCGAAAAACCGATCCCGTAGCGCCGCCGAGAGCGCGTTGCGGGACTGCGGCCGGTTGAGGGTGAGGGTTCGCACCCGTTCGTCGGTGTCGATCAGCAGGATGTCGTCGGTCATTCGATCACCGTAACAGGACCGTTAGACTGTCCTAATGACCAGAAAA CGCGCGAGATCTCATCGACAACCCACTTCCCATGCCTCACGACGGTCACCATGTCGCGGGCATATTTACGTGAGGCACCGAGGGTGTTTCGCGGGCATTCTTGGTGAGTCAAGTCGAACGGTTGAGCCATGATCGACGATTCCGTTACCGTGCTGTCAGAAGACGAAAGTTGGCACCGGCTGGGCAGCGTTGCACTCGGTCGGCTAGTTACCACCTTTGCTGATGAGCCTGGGATCTTCCAGTCAATTTC GGGGGCCCGGACGGCCAGGGTGAGAACCGTTCGCACGGTTTCGGCGTCCGGGAAATGGGTGTTCATGGCTGCCGGGCCTTTCCCATCAACGGCTGCGGTCATCTATTAAGGATCGCGCGTCGACAGCCGCGGTGGCAGAGCACGAAGGCTCGCCAGCGGAGGACCTTTGGCCCTGCCTCAACGGCTCCTCGCAGCGGAGAGTGGTACTGACGGTTGCAGATAATCCGGATCACCGGGGAAGGCGCTGACC ACTCACGTGCCGATCCACGTCTTCTGCCTTGAGAAACCCGGCGTCAAGTGTCGTTAGGTGATTCATGGTCAGCGCCTTCCCCGGTGATCCGGATTATCTGCAACCGTCAGTACCACTCTCCGCTGCGAGGAGCCGTTGAGGCAGGGCCAAAGGTCCTCCGCTGGCGAGCCTTCGTGCTCTGCCACCGCGGCTGTCGACGCGCGATCCTTAATAGATGACCGCAGCCGTTGATGGGAAAGGCCCGGCAGCC GTTGGCGCATGTACACCTGAGCCGTCGGCTCGCCCACTGGACCCGGCTCTACCCCGAGGTGCGGGTGGATCGGGCCATCGCCGGCGGCAGTGCGTGCCGTCATCTGGCCGCCAACGCAAAGCCGGGTCAGCTGTTCGTCGCGGACTCACACTCCGCGCACGAATTGTGCGGTGCATACCAGCCCGGATGCGCCGTACTTACGGTACGCAGTGCCAACTTGTAGGGAGCGGATCTTGGGAGTGGTGCCCTG GGCAGCTGTCGGCAACTGTAAGCCATTTCTGGGACTTTGCTGTGAAAAGCTGGGCGATGGTTGTGGACCTGGACGAGCCACCCGTGCGATAGGTGAGATTCATTCTCGCCCTGACGGGTTGCGTCTGTCATCGGTCGATAAGGACTAACGGCCCTCAGGTGGGGACCAACGCCCCTGGGAGATAGCGGTCCCCGCCAGTAACGTACCGCTGAACCGACGGGATGTATCCGCCCCAGCGAAGGAGACGGCG GTTGTCGATGAGATCTCGCGCGGAGGTCACCAGCAGGTACGCCAAGGCGTATGTGCAGGCTTTGAAGAAGAGCCGGGGCCGGATTTTCGACCAGGTGGTTGACCTGACGGGCTAGTCACGTGATAACGCGCGGCGCCGGCTTGTCGCAGCGGCCAAGCTATCGCCGGGGCTGGGCCGCAGTGTTGCCAAGCGGCGGCGCAAACCGCGTTCGCTGAAGTACTCCTATGACGCGCTGAAGGTGTTGCAGAGG TAGAGGGCTCCGACGTGCCGGTGCCAGCCGCCGCGTTCGAAACACAGCCCTAACGACACGCTGCCGAATATGACCCGTGTCGGAAATTAGGGCGACAAGAGTAATGCGGCTCAACATAGCCTTGCTTTACTTAGGCAAACCTGCCTTCAACCAGGAGGTTATTATCATCCTGTGGTAACTAGGAAAGCCTTTCCTGAGTAAGTATTGCCTTCGTTGCATACCGCCCTTTACCTGCGTTAATCTGCATTTT'\n",
    "k = 20\n",
    "d = 4\n",
    "print(motif_enumeration(dna, k, d))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7cb54c1-98c7-4d43-b76c-4b0b5286a3df",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "def median_string(dna_input, pattern):\n",
    "    \"\"\"\n",
    "    Calculates the sum of Hamming distances between a given k-mer pattern\n",
    "    and a set of DNA strings.\n",
    "\n",
    "    Args:\n",
    "        dna_input: A space-separated string of DNA strings.\n",
    "        pattern: The k-mer pattern string.\n",
    "\n",
    "    Returns:\n",
    "        The sum of Hamming distances between the pattern and the DNA strings.\n",
    "    \"\"\"\n",
    "\n",
    "    def hamming_distance(str1, str2):\n",
    "        \"\"\"\n",
    "        Calculates the Hamming distance between two strings.\n",
    "\n",
    "        Args:\n",
    "            str1: The first string.\n",
    "            str2: The second string.\n",
    "\n",
    "        Returns:\n",
    "            The Hamming distance between the two strings.\n",
    "        \"\"\"\n",
    "        if len(str1) != len(str2):\n",
    "            raise ValueError(\"Strings must have equal length to calculate Hamming distance.\")\n",
    "        distance = 0\n",
    "        for i in range(len(str1)):\n",
    "            if str1[i] != str2[i]:\n",
    "                distance += 1\n",
    "        return distance\n",
    "\n",
    "    def d(pattern, dna):\n",
    "        \"\"\"\n",
    "        Calculates the total Hamming distance between a k-mer pattern and a list of DNA strings.\n",
    "        d(Pattern, Dna)  The output is the sum of distances between Pattern and each string in Dna = {Dna1, ..., Dnat}.\n",
    "\n",
    "        Args:\n",
    "            pattern: The k-mer pattern.\n",
    "            dna: A list of DNA strings.\n",
    "\n",
    "        Returns:\n",
    "            The total Hamming distance.\n",
    "        \"\"\"\n",
    "        k = len(pattern)\n",
    "        distance = 0\n",
    "        for text in dna:\n",
    "            min_dist = float('inf')\n",
    "            for i in range(len(text) - k + 1):\n",
    "                kmer = text[i:i + k]\n",
    "                dist = hamming_distance(pattern, kmer)\n",
    "                if dist < min_dist:\n",
    "                    min_dist = dist\n",
    "            distance += min_dist\n",
    "        return distance\n",
    "\n",
    "    dna_list = dna_input.split()\n",
    "    return d(pattern, dna_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "add25a3f-43d5-48ac-b901-e12091bb0ff9",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_input = 'TACGAGGAGATACAAACCTCAACAAACAGAGAGTCCATGTGCCATGGCGCCTAGGTGTAACTTGGCTTCCCCCAACCTTGGAAGAGGACCTAAGAGAAAAGAAGCAAGGT AGCGTATGTGGTGTTGCGTGCTCTCCAGTGAGAGGTCGAGTTCATCATATCCCATGCCTCGCCTATGATCGGAATTATTGTTAGGCGTTATACGGGGTAAATCGAGTTGA GCATACGCGCTGAATCATAGTAGCAGCTTCAAGGTCTATGCGTCCCTACAAGAGTCGTCCATGTCGTATGTGCGGGGTTGACGCCCTTACATTGTCATACTAGGCGGGCA CTAGAACTAGACTAGAAGTACAGAACATCGCAGCAGATTGACTCAGCGTCGGAAGTAATGTCTTATTGCTTATAGAATTCCGCGCGACCCGTAGTGTATTCGTGGGATAC CACACTACATGTTTGAGTAGCCAGTAGCGGTACTATCACACGGGTGAACATGGGTAGCTAGTGCTAGTTTATTCAGACTGACCATCGTCGCGATTGATTCTCGCAACTGA GGTGGAGACCGACCGCGATGCGGTGATCGTCAGGACCCGACTATGGTTAGTTCCATTGACTTCGGACGAGTCACACTTCCTGTCAGCCGCATATTGCAAAGAATGCAACC TTGGCTGACGTGCCTTGCGCTCACCCGAATACTCCCTTGAGAGAAGAAGCGGCATTGAGGGGTCTAGGGCCCATACATTAAAGGTCAACCCCCATATAGACCAGAACGTG TCATCGGTGTATCTCGTTAGGACTCCGGGCGCCGAAGTTAATTTTCCACTATGGTAAGTATGACGAGCCCTTTTAAAATAAAGGAACCTTAATGTTGAACACGGGTACCA CTAGGAGCCGTGAGCTACCCCAAAGCCAAGGTGGTCTCAAGCTAATATGACACCGCCAGCCCATGGCGCGGACATTTCCACCCACGACAAAAGGTACCGGGTTTGCTAAC GTCTCAATTTTTGGAGGAATAAAGAACTACGGACCCTAAGAGCTTGAACAATTCCGCCTATAGGCTGGTGGATCCGCCATTCTTGCTATTGCTGAGGGAACCTCTACCAG GCACGTACCTCCCATCACGCCCTTACGGACGAGCGCATCCTACGTAGTTCCTAGCCAGTAGATTTTTTGTACTACCCGGACAACCGAACGGAGGGAGCAAGGACCTAGCC CGAAGGGTACGAAAGCTGGGAGCGTAGGACTAGACGCCGTACACGAATGAAAGTTTCTGTGAGTTATTATATCCGAGAGCAATTAGTTAAAGACAAGTGCCCACGGAGAC ATGCTCCCAACAAGGTTAATTCTGTGCTTAGTTTATATGTGATTGGAAGGATGTGCTGGTAAGGTTTTCAATTAAACACCCTAAAAAGTCTGGGCATGCGACGCCTCTAA CGAAGGACCTATGTAGCCGTCTCCGGTCTCAAAGTCGGGTTTGCCTGTTGCTAATGGTAACTCTCATCGCTCCCGGTGTTTGGAGCGCTTGCGAACAGACGAGTCGCGTG CCGTAGGACCAAAATCTTTAAGCTTAGGACATTGACTAGGGTAACCTGGGTTAGGTCCCCCAGGCCTTAAAGATTTTAGCACCTCCAACAGTAGAGGCAGGGGAACCACT TACGAAGACAGTGAAGGGCTATCAGCGCATCATCAGTCCAAGTGCCTCACCTATGTTAACGTTGTTCGGCCTTTTTACCGTAGTCTTAATGTACACCTTATTAACCGGAT CTCAAACGACACACTATAGTCGCTCTTGTTTTTGGTACTTCGTCCCAGCCCGATCAGAGGAGTAAATACAAATCTTGATACTTAGCGTGTTATTTCATATCTGCGGGGGT CTCCAGCTCATTTGCTTCCATGCAATACACAATCAGATACGCACGTGCTATATATTAGGCGCGTCTCTAAGCTCGAGAAAAAGTATCGCGGCCCATAGGTAATAGTAGAG ATCAATGGTGAGTCGAACTGCGAGAACGTGCCGCACATAGGTAGTATGGAAAGTGGCCAATGAACTCAGACTTTTGTCTTCCAAGGTCACCGGGAGGTTGTGGCCGGAAA CCTCAAGACCGTCAGGGTGCGCAATGAGGCCTACTAGTCGAACCGCACAATATGGGAATGTACCGCTGGCATACCCATTAACATCGTGATCAGAAATCAACCTTGACTGC CCGGAATAGTGTCGTGCCAACCTCTCTAGCGAGAGCGAAGGGACTCTTTCTAGTCTTTTCTACATTGCGCATCCGGGATTAAATTTGGCTCGTTGTATACTTCGGATACT GATGATTATCCCAGAAGGCAGAGAAACTCGTCTGAGAAGTTGACAGCTAGAAACGGGTACAACCGGCATCTGGCGATCCAATCGCACAAATTAACACGCAACTGGGGTTA TTTCCTTTCTCCATGATTCATGACTCTGCCGTTATCTACGTAGTAATATGGAAACGACGCGGATTTCCTTGTGGCGGAATTAAACTTTAATTCAATGGATATCCTGGTAT GATGGTTGTATGGGCAAGATTCTAGGTGGTTGAAGGGGCGGAGCTTATATTCCTTTGTGAAAAGTCAATTAAGTCGGAAGAAGCGCCTAATATCTGACATTGAGCGCTTG GCATAAAGAGGTGATAAGTCGAACGTTACGTGGCCCTATTTATCTAGTCAGACGTGCCAGTCCCTAATATAGCGCAGCCCACTGCGGGTAAAGAGCCTGTCGTTACACTG CCACTACAGATCAAGCTCAAGTATACCCCGACTACATAAGAGGACCCGAGCTCCCTTCTAATTGTTCTTCAACTACCAGATCGGACCGGATCTCACTACGGCGTAGTTAG AAGCCAGTGATGCGATAAACAACAGAGGGTAGATGTTCCAATCTTAGAGGCGAATAAACGAGCAAGAGAGCAATAGGACGTGCACCACGTTGATGCGGCAGTACTCGACT TCAAGAGGTTAGCTAATCGGTGGTGTGCAGTGGCTTAATCCTTCCCTAGCAGCTAAAGTTTAGGCTGGTAAGAGCAGTCTACAATTTATAAGCCAACCGTGTCAGGGTTT CGTAGTATCATGTTCTGACGTTTTTGTAATATAGTATTCAGAGCTTTTGCGGCGCGACTGTCAGCCGCTGCCGTGGCCATCTATGATCAAGAACACGAACTTTCGATCAC CATCCATTAACGCTGCTTCGTTCTGCTTTCGACACTGATGCATTTCTAAAGATTCACTCATCGCGGAGCTTGGAGGGCACAACTTGCGGCAATTTGGTAAGGGACCATAC AGCCCAGTCTTAGGCACAGCAGGTGGAGTGCACCAGGGCAGTCTATCGATAGGTTTGGAGATACCAATACGCGAACGGAAGAGAACAGGAATGTCTCTGCAGACGGAACA GTCGCTTTTCTCGCAAGAATGATCTCGCGCAGATGAGCAACGTTCCGCTCCGAGTGGTATCCCCATCCTCAGCGTGAGCACTATCCGCGAAGCACATTCTTGGCATCGGT CGAAAGTAGGGATCATCATCTACAACACAGCGGCCTGCGGTTGGCCACGTACACAGTACTGTATCGTGGGCTCAGAATGAACATTGATTAAGGCTGGGTGACGGCGGCTG AATGCACTTGGAACTCTATGTATGTCGTCTGACTGTTTTCTTAAGACGGCGGACGTAGGCCACGACTCGGCAAAGCGTATGTCTACGCCATTCGGGTTGAATGAATACCC'\n",
    "pattern = 'GAGAT'\n",
    "print(median_string(dna_input, pattern))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2934ca9c-e580-4f50-9a06-44c42bcb3e3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "def median_string(dna_input, k):\n",
    "    \"\"\"\n",
    "    Finds a k-mer pattern that minimizes the Hamming distance between the k-mers\n",
    "    of the input DNA strings.  This is an implementation of the Median String problem.\n",
    "\n",
    "    Args:\n",
    "        dna_input: A space-separated string of DNA strings.\n",
    "        k: The length of the k-mer to find.\n",
    "\n",
    "    Returns:\n",
    "        The k-mer pattern that minimizes the Hamming distance.\n",
    "    \"\"\"\n",
    "\n",
    "    def hamming_distance(str1, str2):\n",
    "        \"\"\"\n",
    "        Calculates the Hamming distance between two strings.\n",
    "\n",
    "        Args:\n",
    "            str1: The first string.\n",
    "            str2: The second string.\n",
    "\n",
    "        Returns:\n",
    "            The Hamming distance between the two strings.\n",
    "        \"\"\"\n",
    "        if len(str1) != len(str2):\n",
    "            raise ValueError(\"Strings must have equal length to calculate Hamming distance.\")\n",
    "        distance = 0\n",
    "        for i in range(len(str1)):\n",
    "            if str1[i] != str2[i]:\n",
    "                distance += 1\n",
    "        return distance\n",
    "\n",
    "    def d(pattern, dna):\n",
    "        \"\"\"\n",
    "        Calculates the total Hamming distance between a k-mer pattern and a list of DNA strings.\n",
    "        d(Pattern, Dna)\n",
    "\n",
    "        Args:\n",
    "            pattern: The k-mer pattern.\n",
    "            dna: A list of DNA strings.\n",
    "\n",
    "        Returns:\n",
    "            The total Hamming distance.\n",
    "        \"\"\"\n",
    "        k = len(pattern)\n",
    "        distance = 0\n",
    "        for text in dna:\n",
    "            min_dist = float('inf')\n",
    "            for i in range(len(text) - k + 1):\n",
    "                kmer = text[i:i + k]\n",
    "                dist = hamming_distance(pattern, kmer)\n",
    "                if dist < min_dist:\n",
    "                    min_dist = dist\n",
    "            distance += min_dist\n",
    "        return distance\n",
    "\n",
    "    dna_list = dna_input.split()\n",
    "    distance = float('inf')\n",
    "    median = \"\"\n",
    "\n",
    "    # Generate all possible k-mers: AA...AA to TT...TT\n",
    "    nucleotides = ['A', 'C', 'G', 'T']\n",
    "    all_kmers = [''.join(p) for p in itertools.product(nucleotides, repeat=k)]\n",
    "\n",
    "    for pattern in all_kmers:\n",
    "        current_distance = d(pattern, dna_list)\n",
    "        if distance > current_distance:\n",
    "            distance = current_distance\n",
    "            median = pattern\n",
    "    return median\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80979625-6ac5-4176-8102-c323e1b28406",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_input = 'TAGTTTATTGAGGAGTGATTCTCTGCTAAGCTGAATTAGGCG GTAACCTAAGAATAGTTTCCCGTAGCCTTTTGAAGCATTTAG TTTGTCACGCAACATGTTACCACAGTGGCCGCGCATTAGTAT TAGTCTAATTGGGTCGACCCGCCAAACATGTGTGCGCATCAG CCAACAGCGCACCCTTAATAAATCTAGTTTGGTAATGTCCTC TAGTTTGAGTGCGAGAAGGCACAGCGGAGACACGTAGCACCG TAGTTTGTCATGTTGAAATTACCGCCGGCACATCGCTTTCCG CGCCTGTTTGCCGAAGCGACGATAGGGGGGTAGTGTGGATAC AGGTCCTCCTGCATCAAAGACAATGTGGAATAGTGTGGCCCT CAAGATTACCACTGCCGATAGTGTTATTCTTTTTATTTTTAA'\n",
    "k = 7\n",
    "print(median_string(dna_input, k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac0ca5ae-bd8e-4d53-8612-3b670404e3da",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "def median_subroutine_string(dna, k):\n",
    "    \"\"\"\n",
    "    Finds the median string from all possible k-mers.\n",
    "\n",
    "    Args:\n",
    "        dna: A space-separated string of DNA strings.\n",
    "        k: The length of the k-mer.\n",
    "\n",
    "    Returns:\n",
    "        The median string.\n",
    "    \"\"\"\n",
    "\n",
    "    def hamming_distance(str1, str2):\n",
    "        \"\"\"\n",
    "        Calculates the Hamming distance between two strings.\n",
    "\n",
    "        Args:\n",
    "            str1: The first string.\n",
    "            str2: The second string.\n",
    "\n",
    "        Returns:\n",
    "            The Hamming distance between the two strings.\n",
    "        \"\"\"\n",
    "        if len(str1) != len(str2):\n",
    "            raise ValueError(\"Strings must have equal length to calculate Hamming distance.\")\n",
    "        distance = 0\n",
    "        for i in range(len(str1)):\n",
    "            if str1[i] != str2[i]:\n",
    "                distance += 1\n",
    "        return distance\n",
    "\n",
    "    def distance_between_pattern_and_strings(pattern, dna_list):\n",
    "        \"\"\"\n",
    "        Calculates the total distance between a pattern and a list of DNA strings.\n",
    "        This is equivalent to the d(Pattern, Dna) function.\n",
    "\n",
    "        Args:\n",
    "            pattern: The k-mer pattern.\n",
    "            dna_list: A list of DNA strings.\n",
    "\n",
    "        Returns:\n",
    "            The sum of the minimum Hamming distances between the pattern\n",
    "            and each string in dna_list.\n",
    "        \"\"\"\n",
    "        k = len(pattern)\n",
    "        total_distance = 0\n",
    "        for text in dna_list:\n",
    "            min_distance = float('inf')\n",
    "            for i in range(len(text) - k + 1):\n",
    "                kmer = text[i:i + k]\n",
    "                dist = hamming_distance(pattern, kmer)\n",
    "                if dist < min_distance:\n",
    "                    min_distance = dist\n",
    "            total_distance += min_distance\n",
    "        return total_distance\n",
    "\n",
    "    def all_strings(k):\n",
    "        \"\"\"\n",
    "        Generates all possible strings of length k from the DNA alphabet (A, C, G, T).\n",
    "\n",
    "        Args:\n",
    "            k: The length of the strings to generate.\n",
    "\n",
    "        Returns:\n",
    "            A list of all possible k-mers.\n",
    "        \"\"\"\n",
    "        nucleotides = ['A', 'C', 'G', 'T']\n",
    "        return [''.join(p) for p in itertools.product(nucleotides, repeat=k)]\n",
    "\n",
    "    dna_list = dna.split()\n",
    "    distance = float('inf')\n",
    "    median = \"\"\n",
    "    patterns = all_strings(k)\n",
    "\n",
    "    for pattern in patterns:\n",
    "        current_distance = distance_between_pattern_and_strings(pattern, dna_list)\n",
    "        if distance > current_distance:\n",
    "            distance = current_distance\n",
    "            median = pattern\n",
    "    return median\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "474999cf-7272-409a-956c-3f4f44a8cc14",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna = 'CTCGATGAGTAGGAAAGTAGTTTCACTGGGCGAACCACCCCGGCGCTAATCCTAGTGCCC','GCAATCCTACCCGAGGCCACATATCAGTAGGAACTAGAACCACCACGGGTGGCTAGTTTC','GGTGTTGAACCACGGGGTTAGTTTCATCTATTGTAGGAATCGGCTTCAAATCCTACACAG'\n",
    "k = 7\n",
    "print(median_subroutine_string(dna, k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e9a8b1c-102f-4368-a58b-16eda35bd82d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def profile_most_probable_kmer(text, k, profile):\n",
    "    \"\"\"\n",
    "    Finds the profile-most probable k-mer in a given text.\n",
    "\n",
    "    Args:\n",
    "        text: A DNA string.\n",
    "        k: The length of the k-mer.\n",
    "        profile: A 4xK matrix representing the probability profile for each nucleotide (A, C, G, T) at each position.\n",
    "                 profile[0] corresponds to A, profile[1] to C, profile[2] to G, and profile[3] to T.\n",
    "\n",
    "    Returns:\n",
    "        The profile-most probable k-mer in text.\n",
    "    \"\"\"\n",
    "\n",
    "    def calculate_kmer_probability(kmer, profile):\n",
    "        \"\"\"\n",
    "        Calculates the probability of a k-mer given a profile.\n",
    "\n",
    "        Args:\n",
    "            kmer: A DNA string of length k.\n",
    "            profile: A 4xK matrix representing the probability profile.\n",
    "\n",
    "        Returns:\n",
    "            The probability of the k-mer.\n",
    "        \"\"\"\n",
    "        probability = 1\n",
    "        for i, nucleotide in enumerate(kmer):\n",
    "            if nucleotide == 'A':\n",
    "                probability *= profile[0][i]\n",
    "            elif nucleotide == 'C':\n",
    "                probability *= profile[1][i]\n",
    "            elif nucleotide == 'G':\n",
    "                probability *= profile[2][i]\n",
    "            elif nucleotide == 'T':\n",
    "                probability *= profile[3][i]\n",
    "        return probability\n",
    "\n",
    "    most_probable_kmer = \"\"\n",
    "    max_probability = -1  # Initialize with a value that will be less than any valid probability\n",
    "\n",
    "    for i in range(len(text) - k + 1):\n",
    "        kmer = text[i:i + k]\n",
    "        probability = calculate_kmer_probability(kmer, profile)\n",
    "        if probability > max_probability:\n",
    "            max_probability = probability\n",
    "            most_probable_kmer = kmer\n",
    "\n",
    "    return most_probable_kmer\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c758b66-c72c-454b-a7fd-2b4ca455fb38",
   "metadata": {},
   "outputs": [],
   "source": [
    "text = 'CGCCAGCGTAGAGAATTAGGCGTAAGGCTACATCGGATCAGAGCATGCACGTGTGCTCAAGGACCGCAATGTTCCGTGAACTCAACCGAAGCACATTGCGTTAGATGTCGAATGTAGCATTAGGCACATTTACACCGTGCGCCGTAGCGTCTTTCATCTACTAATACCTAAGCAACCTCCTACCACTAATCCACTTTCGAGTGAGCTCACGCGACTTTGTGTTGGAAGCTAACGGCTATAGCAACCATGAGGCCGAATATTAAGGCTAAACCGTTGGGACAACCGGTGGGGTGTCGCTGGGAGAGGCGATCGCCGGAGTTCCATCCATGACAGCCCCGCGCGTTTCAGCTTTAGAACCTTGAAAGTTTAGCGAGATTGGGCTCTAGTGCCAGAAAACTCGTAGCCAGCTCCACGACAGATGAAGCTTAGCGGAAACAATGAGATTCAATCCAACCTGATAAGACGACTGGGACTCTCCTTAAGAATCCAGTGGGCCGCAGATTCGGGCCAGTGTATGCTTATTGGTCGTACTTCCCCGGTTATTTCAAATGCGGTATCCCAGGCCTCCTGGCGTTGGGCGGGCTCATACAGTCAGGAATCACTGGCTTGGGGTGCACTCTCGCACTTGAAGGACGCAAGGGAAGTGTTTGACCTATAGGAATAGGAGTAGAGATGCGCAATATAAGGGGCTAGTTGTGCCGTAGTCGCGAATGTGGTATGATTGTAAGGACTCAGGAACGTAATTCTTTCTGCCACGTCCCAGACCAGTGAGCGAACTATTTCATAACTGTACAACCGGTCGTGTGGGATAACATCGCTCGTTCCGCCAATAGGGATCCCATATCGAATCCGGAATCGACAAGAGCACTTTGGCGGGAGTAGTGTGCTAAGTATCCTGCGATGTCCCCTAAGACCCAGAGATTCTGACCGACAGTAAATGGACTTTGTGGTTGCGGGCCCTACCGACACGTTGGAGCCCAGCCGTACGTTGATCGGAATT' \n",
    "k = 15\n",
    "profile =[\n",
    "    [0.273, 0.273, 0.227, 0.212, 0.212, 0.212, 0.182, 0.258, 0.182, 0.288, 0.182, 0.348, 0.288, 0.288, 0.273],\n",
    "    [0.258, 0.258, 0.242, 0.273, 0.258, 0.197, 0.242, 0.288, 0.242, 0.273, 0.303, 0.258, 0.318, 0.242, 0.152],\n",
    "    [0.227, 0.288, 0.258, 0.273, 0.288, 0.348, 0.318, 0.197, 0.333, 0.212, 0.273, 0.242, 0.182, 0.288, 0.303],\n",
    "    [0.242, 0.182, 0.273, 0.242, 0.242, 0.242, 0.258, 0.258, 0.242, 0.227, 0.242, 0.152, 0.212, 0.182, 0.273]\n",
    "]\n",
    "print(profile_most_probable_kmer(text, k, profile))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02108562-e628-425f-ba27-4a2a0d4b04ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def greedy_motif_search(dna_input, k, t):\n",
    "    \"\"\"\n",
    "    Finds a set of k-mers that score optimally according to their consensus,\n",
    "    using Laplace's Rule of Succession to construct the profile.\n",
    "\n",
    "    Args:\n",
    "        dna_input: A space-separated string of DNA strings.\n",
    "        k: The length of the motif (k-mer).\n",
    "        t: The number of DNA strings in Dna.\n",
    "\n",
    "    Returns:\n",
    "        A space-separated string representing the best motifs found.\n",
    "    \"\"\"\n",
    "\n",
    "    def calculate_profile(motifs):\n",
    "        \"\"\"\n",
    "        Calculates the profile matrix from a list of motifs, using Laplace's\n",
    "        Rule of Succession.\n",
    "\n",
    "        Args:\n",
    "            motifs: A list of DNA strings, each of length k.\n",
    "\n",
    "        Returns:\n",
    "            A 4xK numpy array representing the probability profile.\n",
    "            Rows are A, C, G, T.\n",
    "        \"\"\"\n",
    "        k = len(motifs[0])\n",
    "        profile = np.zeros((4, k))\n",
    "        for i in range(k):\n",
    "            counts = {'A': 0, 'C': 0, 'G': 0, 'T': 0}\n",
    "            for motif in motifs:\n",
    "                counts[motif[i]] += 1\n",
    "            # Apply Laplace's Rule:\n",
    "            profile[0][i] = (counts['A'] + 1) / (len(motifs) + 4)\n",
    "            profile[1][i] = (counts['C'] + 1) / (len(motifs) + 4)\n",
    "            profile[2][i] = (counts['G'] + 1) / (len(motifs) + 4)\n",
    "            profile[3][i] = (counts['T'] + 1) / (len(motifs) + 4)\n",
    "        return profile\n",
    "\n",
    "    def profile_most_probable_kmer(text, k, profile):\n",
    "        \"\"\"\n",
    "        Finds the profile-most probable k-mer in a given text.\n",
    "\n",
    "        Args:\n",
    "            text: A DNA string.\n",
    "            k: The length of the k-mer.\n",
    "            profile: A 4xK numpy array representing the probability profile.\n",
    "\n",
    "        Returns:\n",
    "            The profile-most probable k-mer in text.\n",
    "        \"\"\"\n",
    "        most_probable_kmer = \"\"\n",
    "        max_probability = -1\n",
    "\n",
    "        for i in range(len(text) - k + 1):\n",
    "            kmer = text[i:i + k]\n",
    "            probability = 1\n",
    "            for j, nucleotide in enumerate(kmer):\n",
    "                if nucleotide == 'A':\n",
    "                    probability *= profile[0][j]\n",
    "                elif nucleotide == 'C':\n",
    "                    probability *= profile[1][j]\n",
    "                elif nucleotide == 'G':\n",
    "                    probability *= profile[2][j]\n",
    "                elif nucleotide == 'T':\n",
    "                    probability *= profile[3][j]\n",
    "            if probability > max_probability:\n",
    "                max_probability = probability\n",
    "                most_probable_kmer = kmer\n",
    "        return most_probable_kmer\n",
    "\n",
    "    def score(motifs):\n",
    "        \"\"\"\n",
    "        Calculates the consensus score of a list of motifs.\n",
    "\n",
    "        Args:\n",
    "            motifs: A list of DNA strings, each of length k.\n",
    "\n",
    "        Returns:\n",
    "            The score (sum of the most frequent nucleotides at each position).\n",
    "        \"\"\"\n",
    "        k = len(motifs[0])\n",
    "        score = 0\n",
    "        for i in range(k):\n",
    "            counts = {'A': 0, 'C': 0, 'G': 0, 'T': 0}\n",
    "            for motif in motifs:\n",
    "                counts[motif[i]] += 1\n",
    "            score += max(counts.values())\n",
    "        return score\n",
    "\n",
    "    dna_list = dna_input.split()\n",
    "    best_motifs = [seq[:k] for seq in dna_list]\n",
    "    best_score = score(best_motifs)\n",
    "\n",
    "    for i in range(len(dna_list[0]) - k + 1):\n",
    "        motif1 = dna_list[0][i:i + k]\n",
    "        motifs = [motif1]\n",
    "        for j in range(1, t):\n",
    "            profile = calculate_profile(motifs[:j])\n",
    "            motif_j = profile_most_probable_kmer(dna_list[j], k, profile)\n",
    "            motifs.append(motif_j)\n",
    "        current_score = score(motifs)\n",
    "        if current_score < best_score:\n",
    "            best_score = current_score\n",
    "            best_motifs = motifs\n",
    "    return ' '.join(best_motifs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35d76d63-d7f5-41b0-a94a-268951dcd559",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_input = 'TAACAATGGGCTACCTACATTTCCGGCCATGGAGACACGCCCCGCGCACCCTTGCACGTAAAATTATACAACGCGTGAAGGGGCGCGATACTGATGGTGTGCGGACAACCTCTGATACCTTGCCAGGTAAAGCTAAGATTCCAATAGGTATACGTC TTAGCTGACAAAAGTTACATCTTCGGTCCTCCCCGATAATTGTCCAGATATTTATCCACTGAGTCCCTCGCGCTGCTTATGCGGCCAATGCGCACAATAGAACTATTTCGCGGTTCATTGTGATTGGCGTTCACGTTCCGAGGGTACCACCTTGGT ACCTACATATCCCGTTTGCGGAAGTCTTCGCTTAGCACTAACGCTCGGTACGTCTCCATGATGAACTTACGCGTCTTGGTCCTTGGCAGAATCATACAGGCGCACTCCCCGGCGCCGGGCTGGGGGATCACCCTTGGTAAATTGTTGGTGTGTCCT GTCCCCCATGAGCTCCCACTTGATCAGGAGACTAGTATATACATCTCCACAGCATTGGTAGTGCCGAAGATAGATTGTAGTGGCCGATCTCGCTAGCCGGCAACTCCCCGTCTTGCAACATGTGCGTCGCAGTGACGTGACCCGCAACATTTACGT GAACGGGCATGTAAGTACATTTCCAATAGTCACACGAGTCCTCCAGCACCATGTGTCTAGAACGGATTTGTATTGATGTTTTGGGGGTATCACGCCTCGTTTCGGGTTGTGCGCTAACGTACGGAACCACCACGGCGAAACCGAACCCTACTGCCT AGGCTTTCTGCGGCCGGTCTGTCGAAGTACATTTTCCGTTAGGCAGCCCACAAGGGCAATTACTTAGACGGGCTTAAAGTGCCACAAGTCAGTATAATTGCGATGAGAAGGCCGGAAAGGAAATGAGGTTCTGATGGCCGGCGGTCACGGGCCGTA GAACCATAACGGAGGCTTATGAAGACTAGTGTAGACGTGAAACGCTTTTAAACCTGGGGCTGGCTGATAATCGGGCGAATTTGGCCTAGCGATAGCAAGTACATTTTCTCCTTTAGTCATGTGATGGACAGTCGCTCTACTTGAACTGAGACTGAG TCGTCAGATACGACATACATATGCGTCCCTACCGTACAGTTCCCGGTACACCCCGAGTCTATGGGCCATTCTTACTTGGATAAAGCTGGCGGCCCACCAGCCCGAGCTGTAAAAATTAGTCGCACTGTCGAATATTAATCTGCGTAAATTAATAAG CACGCTTGTTCACCACTCTTCCGATGCCCTAAACCATACGACGGTGGTCGTAGCCACAACTGCGGTCCCGTAACCTACATATCCGAGTGCGGTCTACTGGCGGTCAGACGCCAACAATTTGGAGCACTAGAACGATCGGTATTCGGTGCATTAGAT AGAACATTGCGTAACACAGGTCTATGCGTTGACGGTTGTCTTTTGGGCATCTACATATTCCAAGCGAGTTTTGGGGTACGTGACGCCCGTTTTACTAGGTTGATTTTCTGTCCATGACAACACTCACTCATCGATTGTTAAAGAATGATTGCTTAC GTCCGTTGGGCCCAGTGTGCGCTGCCTCGTAGTAGAATATGAGCAATACTGCCCTTAAGCGGAACAGAGTTTAAGGCGGAAGTGTCCTCGCCTGACCCGTAGGTACACTGGTGCCAACCTATGTATCCGGAAAATTACATATACAGTCGGTGCCGT GAAATGTGCTATACTTACATATACGCGAGAGTGAACTCAAGTTTGAGGAAGGTTGAAGCCTCGTTGCCGTGTGTTTACATAGCCAGGTTGTTATGTTTCGGGTTGAGACTGGCTTGTGGATGGGAAGGCCAGCTTCTTTGACTCGGCTCCATCATT TCCGCACGAGTAATTGTGCACGTAAGTGCAAGGACAGTTTGAATCGCTGTCGGAAGTTGAGACCGTCATCTGTTGATAGGCAGGGACGGAACCAAAGTTTCCGCGATAAATAGGTTTCGCGTACGTCTTCGCAAAGAGCCGTCCACCTACATGTCC TTGTCGTCGATACGTGTGTGATGGATTTACATCTCCAGATTGAAGAGAAGGGGGAGTGTGCCTAATCTTAACAACAAAAACACTAGATATCGAGCAACATACCCGTTACCTGCCAAGACCTCGACGTGCACCCATTCACGAGCCTTATAGTTTATT GCTTGATAGTTTCACGCCTTATGCACAGCGATACTTAAGACGCGGGTAGGATACCTGTACTGTGATGGTTCTACACTACCGATGACACAATCATGGAGTTACATCTTCAGGCACAGTATAAACGTCGGTCCCCCAATCGAGAGAGTTTTAGTAAAA ATGTATGATTTCAGCTACATTTTCACCTCAGGAGTACATCATCGCCAGGGGCCATCCCGAGCCCCTGTCTTAGCAAGTAACACGACATACTTGTGCGATATGGAGGTTATTCGATTCAGTACATGTTAGGGGCAAGAGCCGTCAGCGTCGCTCACG GCTTGTGTTGGTATGTACATGTTCATCCCCACACTTCAGAGTTACGATGAAGGAGATTGGGTACGCATTCTCTGTCTGCATATACAGACTTCCTCACCCGTCCGGGTTGATCCTCTGGCACTCCGTTATTTAGTGACTAAGCTCGAACTTCCGCGT ACGTACATATGCCAAGCCTACGCCGAAGTTTTACCCCATGGCATAGACCGGCCATATGTTATACGATTTACGCGACCACTCGAGATCGTGCGGCATTTATAAGAGGGGGTTGCTCATATCGCAGTATTACGACCCTTCAATTTCCACAGGTCGTCA TGGAGTCGTACGCAAGAATAACGCAAGTACATGTTCTTGATTGCAACGTCGGAATTCCCGGAGCACATCCCGTTGGGTGCCCGATTCCTAGGTCACTATGCGATCCGACGTTTCACGAAAAGGATCGGGCGGAGCGTATCCCCACCTTATGCACTC CCTCCGGTCTTGAAGGGTTGCATCTTTTCTATTCACGAGTCTATAGGTACCGCGAGGTGGAAGTACATCTTCAACCGAGGGATTCGGAGCGAAGTCAGATGTAACAGCTCTTTTTGCAAGCACGCCAACCCATAACTGAGTCGACGCGCAAAAATG TGTACGGCCCTTCTATTCCTAGCCCAACTTCGGACGAACTACATGTACTCAGACCAGCCTCGCCATGCGAATGGAGTGGATTACCGTCGCATAGCGCCTGTGCCCCTGAGTCAAGGGCGGAAAGACGACAGCCGTGTCCTTGAGTTTATGACGGGT AGACAGGGTGCTGTTGAATGACGCTGATGGAACAATGCGGGTCGCATACCGACAACTCCGTTCCCATTCTCAACGTCCATAAAGTGGTAGGGGCGGCGTTCCGTGGCCTACCAGGGAGCCGTCGTGCTCCCAAAATACATCTTCGAAGCTCAGCAG AATCCAATGGGGCGGTGCAGCCATGAAAGAGCTACCGTATTTTCAGGCGACCCTGTGGTTCCAACCCCATCGATCTTGGGGGTGGATTCATGTCGTACCCAAACTAGAATCTACATTTGCGCAACTTCAGAAGCCTGTTAAAATCGTTGACGGATC ACCTACATGTCCTCCCGGCATACCATTCGCCTGTTCGACGCGTTCCTCTGCGGGCCCAAGTTGGGGGATAAATAAGACAGCTTGCACCGTGCAACCATCCGGGACCGGCTTCAATTCCATAATTCTCATTCGTATGACTGCATTCCAGCTAGGTTG AGCTCAAGGTGGTTGCATGAATCAAGCTGCGCATAGCCATTACAGTCCTTATGAGCGCACGCTTTACTGTCCTCTGGATCACAGATGTACATGTACCATCCCATACCTACAGTGTGTCAAGGTGCCGAGTACTATCCATCTGGAACTACGGCCTTC'\n",
    "k = 12\n",
    "t = 25\n",
    "print(greedy_motif_search(dna_input, k, t))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bef48c3-7318-48d4-b905-0c3ef469877c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def profile_matrix_with_pseudocounts(motifs):\n",
    "    k = len(motifs[0])\n",
    "    profile = {'A': [1] * k, 'C': [1] * k, 'G': [1] * k, 'T': [1] * k}\n",
    "    total_motifs = len(motifs) + 4  # +4 for the pseudocounts\n",
    "\n",
    "    for i in range(k):\n",
    "        for motif in motifs:\n",
    "            profile[motif[i]][i] += 1\n",
    "\n",
    "    for nucleotide in profile:\n",
    "        for i in range(k):\n",
    "            profile[nucleotide][i] /= total_motifs\n",
    "\n",
    "    return profile\n",
    "\n",
    "def profile_most_probable_kmer(text, k, profile):\n",
    "    max_prob = -1\n",
    "    most_probable_kmer = text[:k]\n",
    "\n",
    "    for i in range(len(text) - k + 1):\n",
    "        kmer = text[i:i + k]\n",
    "        prob = 1\n",
    "        for j in range(k):\n",
    "            prob *= profile[kmer[j]][j]\n",
    "        if prob > max_prob:\n",
    "            max_prob = prob\n",
    "            most_probable_kmer = kmer\n",
    "\n",
    "    return most_probable_kmer\n",
    "\n",
    "def score_motifs(motifs):\n",
    "    score = 0\n",
    "    k = len(motifs[0])\n",
    "\n",
    "    for i in range(k):\n",
    "        column = [motif[i] for motif in motifs]\n",
    "        most_common = max(set(column), key=column.count)\n",
    "        score += sum(1 for nucleotide in column if nucleotide != most_common)\n",
    "\n",
    "    return score\n",
    "\n",
    "def greedy_motif_search_with_pseudocounts(dna_strings, k, t):\n",
    "    dna = dna_strings.strip().split()  # Split input string by spaces\n",
    "    best_motifs = [string[:k] for string in dna]\n",
    "\n",
    "    for i in range(len(dna[0]) - k + 1):\n",
    "        motifs = [dna[0][i:i + k]]\n",
    "\n",
    "        for j in range(1, t):\n",
    "            profile = profile_matrix_with_pseudocounts(motifs)\n",
    "            motifs.append(profile_most_probable_kmer(dna[j], k, profile))\n",
    "\n",
    "        if score_motifs(motifs) < score_motifs(best_motifs):\n",
    "            best_motifs = motifs\n",
    "\n",
    "    return ' '.join(best_motifs)  # Return space-separated motifs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71b6a322-bcc8-408d-ba3e-2cb6db3715eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def profile_matrix(motifs):\n",
    "    k = len(motifs[0])\n",
    "    profile = {'A': [0] * k, 'C': [0] * k, 'G': [0] * k, 'T': [0] * k}\n",
    "    total_motifs = len(motifs)\n",
    "\n",
    "    for i in range(k):\n",
    "        for motif in motifs:\n",
    "            profile[motif[i]][i] += 1\n",
    "\n",
    "    for nucleotide in profile:\n",
    "        for i in range(k):\n",
    "            profile[nucleotide][i] /= total_motifs\n",
    "\n",
    "    return profile\n",
    "\n",
    "def profile_most_probable_kmer(text, k, profile):\n",
    "    max_prob = -1\n",
    "    most_probable_kmer = text[:k]\n",
    "\n",
    "    for i in range(len(text) - k + 1):\n",
    "        kmer = text[i:i + k]\n",
    "        prob = 1\n",
    "        for j in range(k):\n",
    "            prob *= profile[kmer[j]][j]\n",
    "        if prob > max_prob:\n",
    "            max_prob = prob\n",
    "            most_probable_kmer = kmer\n",
    "\n",
    "    return most_probable_kmer\n",
    "\n",
    "def score_motifs(motifs):\n",
    "    score = 0\n",
    "    k = len(motifs[0])\n",
    "\n",
    "    for i in range(k):\n",
    "        column = [motif[i] for motif in motifs]\n",
    "        most_common = max(set(column), key=column.count)\n",
    "        score += sum(1 for nucleotide in column if nucleotide != most_common)\n",
    "\n",
    "    return score\n",
    "\n",
    "def greedy_motif_search(dna_strings, k, t):\n",
    "    dna = dna_strings.strip().split()\n",
    "    best_motifs = [string[:k] for string in dna]\n",
    "\n",
    "    for i in range(len(dna[0]) - k + 1):\n",
    "        motifs = [dna[0][i:i + k]]\n",
    "\n",
    "        for j in range(1, t):\n",
    "            profile = profile_matrix(motifs)\n",
    "            motifs.append(profile_most_probable_kmer(dna[j], k, profile))\n",
    "\n",
    "        if score_motifs(motifs_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f348e7a-9b22-427b-b5fe-40f3974e6326",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_strings = 'TAAACGGTCAAGTAACAGATCGTATGCGCCTAAACCTTTCCCGATAGTAAGTTAGTCGATCCAACACTACCGTACGGTCGTCCCCAAGTGCGACAATTCAACTGTAGCTAGCCGTTGACACGTAGTCTCTTGTAGGTGGGCCGCGTATTTAGACAG CAGTTACTCTCTAGCACAGCTACTCCCTACATTTTGGGGTTCCACAAAAGTGCTTATGAGCAGGTGGAGCGCAACATCTCCCTTCTGCATACAACGTCCCACTGATCTACCAATGTGGTTACGCGTTCACCAACGCTGGGGCATACCTAGGAGCTG TGACTGTGTGCTTTTGACAGCAATCCTGTGTGGTGTCAGGTGGAACGATACCCCCAGCAAGTTCCGGGCCATTGATTCTGGGGCGTATTGTCTGCTCGGGGAGTTAGCTCCATTTCCACCCTGATTCCGGATTGACTTAACGGACATGGGCATGGA CGGGCTTTTATCTCCATCTATTTTAATCAACGATTTTTTGCCCAAATTCGTGTAAATAAAATAAACCCAGATTAGCTTACGGTCACTCCCTTTCATAGAATGGGGTGAGAGGTGGTACGTGACGCCTGCCATAATGTCCATAAACGGCAGCGCGGC CCCTGACCTCTACAATTGAGTCACCGCTGCCGTCACCCCCACAAAGCCGTTTGGCTATCATAGGTGGAACGACGGCCAGTGAGCGTTACGTGGAAAGCCCCAGTCACCGGGCAAGTGGTTGATTTGATTCCTGGAGAAGGGTGATCCCTATCCAGA TAAAACCCGAGCAGTGTGGGGCAGGGCAGTAAGTATTGTTGTTAGGAGTACCTACGGGCTCCCCCTGGAGTTTTTGGTATAAGGGTAGTCGGGACCTAGGTGGCCCGACATTACCCTCTAAGATACTTCCATCCCACGGTCAGACTATAAACAGTT GGGTACCTCGCCTGCTAAGCTACTTCTGTAAGGAGTGATTGCTGCAGATAGCGCCGGTCCCAGGTGGAACGCAGTTCCTTTTGTATTCTCTCCTACTACCAGGTTGTTGTCATTCGCGTTTCTTAAAGCTAATACGACTCCCACGACTGGTCCCCG GACATAGGGCATGTGGAAGCGTGCCGCGACGACATACTCTTGGAACACTAGGTGGACCGTTCGCCTTCACTGTGCCATAATGAACGGAATGATTGGGCATACTGGAATCGTATCATGTAACATACACCAAGTGTGAGCTGATGCGAACAGATTATC TGAAAGGAGGTCGACAGGAGCCGGCTGTTTCGATCCAAAATTGCCCTAGCACGCAGACACTCCCTGGTGAGCATCCGCAGCAAGCAGGTGGTGCGCTCACCGCCGCAAAACCAAGGGTATGCTAGTTTCGATCAAGGCTAGAAGATCGCAGCTATT GCCGCGTGTCAATGATGTATTCTCAAGGGCTCAGTCACGGGCAGAGTCACTCACCCCATTAACCAGAGGATAACATTTCACAAGCTGTGTAGACCGCTTGATTCGTAGCAGGTGGGCCGTAGTCCGCTACTTATGTTCCTACCTTAAAAATTTCTA AGATGATGCAAAACTCCTATGCACTGATACCCGCGTTTACGGCCGACAGTTCCTACATTTTAGGGTTGGTGGTAAGGCAGAATAGGGTTCGCGGCCGTGGGACGATGCCGTTAGGTCCAAACTAGACGTAATTAGGTGGCTCGCGCAGGTGGTACG TCCTACCATGTTAGCTCGTTATTATTTGGTAGTTGGTTTGGTGCGCTCGGTTCACGAGGCGCGAAAGGGCGCATTGCAGGGCACACGTACTATTAAACTCGACACTGCGGTCTATTAGAATTTCTTCATCGCGAGGTGGTGCGTTCTTCGGCCTAC AAGCACGATACGCTTATCGAGCCTGGCGGTCAGCACTAGTAGTGTTATAAGGTGGACCGCCCTTTCTATAATCCTGGATCTAGTCCGCTCTTTAGAGCGGGCTCGCTAGTAAGACCTACTTCCCCTATTGAAGTAACTACGTATAGGACATCCGCT GTCCTACTGGCATGACCGAGCAAAGTAAATGCAATGCGGGCTCGTGCAGTTCCCCCAACTTACAGGCATTGCGAGGTGGTACGAGAATAGAGTTGTACTATGGCATAATGAGCCAATTTGGGCAAACACTGGAACTCGAGTGCAGTATCCAGATCG ATGGCGAATCGGATCAAAGGTGAAGAGGTGGCGCGGGTCTAATAATTAAACCCGAGCAAGTGTCTCTTGTGTGATGTGATAGCCTCCATCGGATGACGTTTGCCAATGCCCCTTCTTAAGACTGGCTTAACAGGCCCTTACCATCTGGCTCTGGTC TAAGTAACGGACTCTAGTGCTAGCCGGAGGCTTGAAGAGGTGGTTCGGGCTAATCTCTTGGATCCATGTCATTATATTAAGTTGGCTCTCAGTGGCGTAACCGTGGATTTTCACACCATATAATTATCTACCGACTTATAGCAATAATCGTGCCGT GCACCCGCCCTGAGTCCCACAAAGAGATCCCCTCTATCACAACCTAGCGTGCTATTTCTAGTTCTTTGATGTTCGTAAGACACATAGGTGGTACGTTCAAGCCCACCTCTGTACGCGTAGGATAGATCTTTGGACTGACTCCACCTCTTGCGATTA GCTCAACTTGCAACAAAAGGAGGGTCGCGGCAGAAACTTGACGGGAGCAGGCTAGTATGCAAAACCTCCGGGTAGAGGTTATGCGTTGATCGTACCAAGGTGGGCCGTTGTGGTTATCTGCTAGGGCGAAGTACAAATCTGGTGGGCCTCAGATGC CAGGTGGCACGTCTGGGATTTAAAAAGTAACCGATCAGAGCTCTCAGATTGATCCTCCAGAGCTTCGGAGAACACGTATCCTCCCACCTAGATGACTTTTGGGGGTTGGCGCTAAATTTAGCAACGGTGGCTGCTACCTGCACTCTCAGGGCTTTC CAGATGGCCAGTATTCGAAGTCTATTAACTGGTATACACGAAAATAGTCTGCCCCGCTCATCTAACCATTGTGGAAATAGTCGCTAGGGAGTCCAAGTGACTGAACTCCATTGTCTGATGAAGGTGGGGCGTAGAAACGGATGACAAAGTACGGGT AATTCAGGGTGCTCTGTTCAGTGTACGCCACACCGTGAGATTACACCATAGGTGGGGCGGCACTAAGACAATCCGGGAACTATCTGTGTCTACCTTCTTGTACGCATAATCCTCGGTGACCTCTAAGTTAGACTATAGCAGAGCGTAGCCCACGAG ACCGTTGAGGGAACTCGCCGTCGTTTGACCTAGAAGCGACACGACTGCCAGGTGAGCCCGTAGGTGGCCCGACGGGGGTCAGTCCCAGAAAGTATTTGGGAGGAGTATCTGAACTCTGTAGGAGGATCTTTGCTGAGAACTCAGGCTGAAGTAAGT AAGGTGGCACGGACGCATCTATTCGACGCAAATGGGCTGAACTCCAGAACTATATGTTACCGAACGACAGGAGACAGGGAACCACGCGGCGTGCTTGGTCCGCAGCTAGTGCACTTAGAACCACCTCCACTCGTCGACCGAGAGGGAAGTGCTCGA GCACCGTGGGGAAGACCCAAAACTCTGCAGTTTGAGCAGTGGACCGGCGATGCGTAACGTGCAATGTCGTTATCATATATCAGGTTAGGGAGCTTGTCGACGTCACAGTATCCAAAAGACTTGAATTTCCCCGAGGTGGACCGTTGGTGCTCTGGT TTAGGCGAAGTCAGCGAATTGGCCCATGTCTCTTCCAGAGAATCTATGGAGGTGGGTCGGAATCCTCGGTATCGCTGTTACGATGAAAAGCGGATGGTTCCTCGCAGTGAGTTGGCGAATGTTTCTAACTCCAGGCTCGTAGGAGCCCCCATTTGA'\n",
    "k = 12\n",
    "t = 25\n",
    "print(greedy_motif_search(dna_strings, k, t))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4b985bad-3611-4197-a7fb-07b63a495208",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "random.seed(0)  # <- force reproducible random choices\n",
    "\n",
    "def random_kmer(text, k):\n",
    "    start = random.randint(0, len(text) - k)\n",
    "    return text[start:start + k]\n",
    "\n",
    "def profile_matrix_with_pseudocounts(motifs):\n",
    "    k = len(motifs[0])\n",
    "    # Initialize with pseudocounts (Laplace rule: start with 1 everywhere)\n",
    "    profile = {'A': [1] * k, 'C': [1] * k, 'G': [1] * k, 'T': [1] * k}\n",
    "\n",
    "    for motif in motifs:\n",
    "        for idx, nucleotide in enumerate(motif):\n",
    "            profile[nucleotide][idx] += 1  # Add actual counts\n",
    "\n",
    "    total_counts = len(motifs) + 4  # +4 because we added 1 pseudocount per nucleotide\n",
    "    for nucleotide in profile:\n",
    "        for i in range(k):\n",
    "            profile[nucleotide][i] /= total_counts  # Normalize by total\n",
    "\n",
    "    return profile\n",
    "\n",
    "def profile_most_probable_kmer(text, k, profile):\n",
    "    max_prob = -1\n",
    "    best_kmer = text[0:k]\n",
    "\n",
    "    for i in range(len(text) - k + 1):\n",
    "        kmer = text[i:i+k]\n",
    "        prob = 1\n",
    "        for j, nucleotide in enumerate(kmer):\n",
    "            prob *= profile[nucleotide][j]\n",
    "        if prob > max_prob:\n",
    "            max_prob = prob\n",
    "            best_kmer = kmer\n",
    "\n",
    "    return best_kmer\n",
    "\n",
    "def score_motifs(motifs):\n",
    "    score = 0\n",
    "    k = len(motifs[0])\n",
    "\n",
    "    for i in range(k):\n",
    "        column = [motif[i] for motif in motifs]\n",
    "        most_common = max(set(column), key=column.count)\n",
    "        score += sum(1 for nucleotide in column if nucleotide != most_common)\n",
    "\n",
    "    return score\n",
    "\n",
    "def randomized_motif_search_one_iteration(dna_strings, k, t):\n",
    "    dna = dna_strings.strip().split()\n",
    "    # Step 1: Initialize motifs by selecting a random k-mer from each DNA string.\n",
    "    motifs = [random_kmer(seq, k) for seq in dna]\n",
    "    print(f\"Initial Motifs: {motifs}\")\n",
    "\n",
    "    # Step 2: Construct a profile matrix from these initial motifs.\n",
    "    profile = profile_matrix_with_pseudocounts(motifs)\n",
    "    print(\"Profile Matrix (with pseudocounts):\")\n",
    "    for nucleotide, probabilities in profile.items():\n",
    "        print(f\"{nucleotide}: {probabilities}\")\n",
    "\n",
    "    # Step 3: For each DNA string, find the profile-most probable k-mer.\n",
    "    new_motifs = [profile_most_probable_kmer(seq, k, profile) for seq in dna]\n",
    "    print(f\"Motifs after one iteration: {new_motifs}\")\n",
    "\n",
    "    return new_motifs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa90383b-37b7-42a2-a488-00305548c07a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "random.seed(0)  # <- force reproducible random choices\n",
    "\n",
    "def random_kmer(text, k):\n",
    "    start = random.randint(0, len(text) - k)\n",
    "    return text[start:start + k]\n",
    "\n",
    "def profile_matrix_with_pseudocounts(motifs):\n",
    "    k = len(motifs[0])\n",
    "    # Initialize with pseudocounts (Laplace rule: start with 1 everywhere)\n",
    "    profile = {'A': [1] * k, 'C': [1] * k, 'G': [1] * k, 'T': [1] * k}\n",
    "\n",
    "    for motif in motifs:\n",
    "        for idx, nucleotide in enumerate(motif):\n",
    "            profile[nucleotide][idx] += 1  # Add actual counts\n",
    "\n",
    "    total_counts = len(motifs) + 4  # +4 because we added 1 pseudocount per nucleotide\n",
    "    for nucleotide in profile:\n",
    "        for i in range(k):\n",
    "            profile[nucleotide][i] /= total_counts  # Normalize by total\n",
    "\n",
    "    return profile\n",
    "\n",
    "def profile_most_probable_kmer(text, k, profile):\n",
    "    max_prob = -1\n",
    "    best_kmer = text[0:k]\n",
    "\n",
    "    for i in range(len(text) - k + 1):\n",
    "        kmer = text[i:i+k]\n",
    "        prob = 1\n",
    "        for j, nucleotide in enumerate(kmer):\n",
    "            prob *= profile[nucleotide][j]\n",
    "        if prob > max_prob:\n",
    "            max_prob = prob\n",
    "            best_kmer = kmer\n",
    "\n",
    "    return best_kmer\n",
    "\n",
    "def score_motifs(motifs):\n",
    "    score = 0\n",
    "    k = len(motifs[0])\n",
    "\n",
    "    for i in range(k):\n",
    "        column = [motif[i] for motif in motifs]\n",
    "        most_common = max(set(column), key=column.count)\n",
    "        score += sum(1 for nucleotide in column if nucleotide != most_common)\n",
    "\n",
    "    return score\n",
    "\n",
    "def randomized_motif_search_one_iteration_with_initial_motifs(dna_strings, k, t, initial_motifs_str):\n",
    "    dna = dna_strings.strip().split()\n",
    "    initial_motifs = initial_motifs_str.strip().split()\n",
    "\n",
    "    if len(initial_motifs) != t:\n",
    "        raise ValueError(f\"Number of initial motifs ({len(initial_motifs)}) must match the number of DNA strings ({t}).\")\n",
    "    if any(len(motif) != k for motif in initial_motifs):\n",
    "        raise ValueError(f\"All initial motifs must have length k ({k}).\")\n",
    "\n",
    "    print(f\"Initial Motifs: {initial_motifs}\")\n",
    "\n",
    "    # Step 2: Construct a profile matrix from these initial motifs.\n",
    "    profile = profile_matrix_with_pseudocounts(initial_motifs)\n",
    "    print(\"Profile Matrix (with pseudocounts):\")\n",
    "    for nucleotide, probabilities in profile.items():\n",
    "        print(f\"{nucleotide}: {probabilities}\")\n",
    "\n",
    "    # Step 3: For each DNA string, find the profile-most probable k-mer.\n",
    "    new_motifs = [profile_most_probable_kmer(seq, k, profile) for seq in dna]\n",
    "    print(f\"Motifs after one iteration: {new_motifs}\")\n",
    "\n",
    "    return new_motifs\n",
    "\n",
    "# Example usage with predetermined initial motifs:\n",
    "dna_sequences = 'TGACGTTC TAAGAGTT GGACGAAA CTGTTCGC'\n",
    "k_mer_length = 3\n",
    "number_of_sequences = 4\n",
    "initial_motifs_str = 'TGA GTT GAA TGT'\n",
    "\n",
    "final_motifs = randomized_motif_search_one_iteration_with_initial_motifs(\n",
    "    dna_sequences, k_mer_length, number_of_sequences, initial_motifs_str\n",
    ")\n",
    "print(f\"\\nFinal Motifs after one iteration: {final_motifs}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b9e1744a-179e-4b9d-9678-30f8c55860be",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from collections import Counter\n",
    "\n",
    "def random_kmer(sequence, k, profile):\n",
    "    n = len(sequence)\n",
    "    probabilities = []\n",
    "    kmers = []\n",
    "    for i in range(n - k + 1):\n",
    "        kmer = sequence[i:i + k]\n",
    "        prob = 1\n",
    "        for j, nucleotide in enumerate(kmer):\n",
    "            prob *= profile[nucleotide][j]\n",
    "        kmers.append(kmer)\n",
    "        probabilities.append(prob)\n",
    "\n",
    "    total = sum(probabilities)\n",
    "    probabilities = [p / total for p in probabilities]\n",
    "    return random.choices(kmers, weights=probabilities, k=1)[0]\n",
    "\n",
    "def build_profile(motifs, pseudocount=1):\n",
    "    k = len(motifs[0])\n",
    "    profile = {nuc: [pseudocount] * k for nuc in 'ACGT'}\n",
    "\n",
    "    for motif in motifs:\n",
    "        for idx, nucleotide in enumerate(motif):\n",
    "            profile[nucleotide][idx] += 1\n",
    "\n",
    "    for i in range(k):\n",
    "        total = sum(profile[nuc][i] for nuc in 'ACGT')\n",
    "        for nuc in 'ACGT':\n",
    "            profile[nuc][i] /= total\n",
    "\n",
    "    return profile\n",
    "\n",
    "def score(motifs):\n",
    "    k = len(motifs[0])\n",
    "    score = 0\n",
    "    for i in range(k):\n",
    "        column = [motif[i] for motif in motifs]\n",
    "        most_common = Counter(column).most_common(1)[0][1]\n",
    "        score += len(motifs) - most_common\n",
    "    return score\n",
    "\n",
    "def gibbs_sampler_once(Dna, k, t, N):\n",
    "    motifs = [random.choice([s[i:i+k] for i in range(len(s) - k + 1)]) for s in Dna]\n",
    "    best_motifs = motifs[:]\n",
    "\n",
    "    for j in range(N):\n",
    "        i = random.randint(0, t - 1)\n",
    "        motifs_except_i = motifs[:i] + motifs[i+1:]\n",
    "        profile = build_profile(motifs_except_i, pseudocount=1)\n",
    "        motifs[i] = random_kmer(Dna[i], k, profile)\n",
    "\n",
    "        if score(motifs) < score(best_motifs):\n",
    "            best_motifs = motifs[:]\n",
    "\n",
    "    return best_motifs\n",
    "\n",
    "def run_multiple_starts(dna_string, k, t, N, runs=20):\n",
    "    Dna = dna_string.strip().split()\n",
    "    best_overall = gibbs_sampler_once(Dna, k, t, N)\n",
    "    best_score = score(best_overall)\n",
    "\n",
    "    for _ in range(runs - 1):\n",
    "        motifs = gibbs_sampler_once(Dna, k, t, N)\n",
    "        current_score = score(motifs)\n",
    "        if current_score < best_score:\n",
    "            best_score = current_score\n",
    "            best_overall = motifs\n",
    "\n",
    "    return ' '.join(best_overall)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f3208733-28e8-4dfa-8880-76c4de2532fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "random.seed(0)  # <- force reproducible random choices\n",
    "\n",
    "def random_kmer(text, k):\n",
    "    start = random.randint(0, len(text) - k)\n",
    "    return text[start:start + k]\n",
    "\n",
    "def profile_matrix_with_pseudocounts(motifs):\n",
    "    k = len(motifs[0])\n",
    "    # Initialize with pseudocounts (Laplace rule: start with 1 everywhere)\n",
    "    profile = {'A': [1] * k, 'C': [1] * k, 'G': [1] * k, 'T': [1] * k}\n",
    "\n",
    "    for motif in motifs:\n",
    "        for idx, nucleotide in enumerate(motif):\n",
    "            profile[nucleotide][idx] += 1  # Add actual counts\n",
    "\n",
    "    total_counts = len(motifs) + 4  # +4 because we added 1 pseudocount per nucleotide\n",
    "    for nucleotide in profile:\n",
    "        for i in range(k):\n",
    "            profile[nucleotide][i] /= total_counts  # Normalize by total\n",
    "\n",
    "    return profile\n",
    "\n",
    "def profile_most_probable_kmer(text, k, profile):\n",
    "    max_prob = -1\n",
    "    best_kmer = text[0:k]\n",
    "\n",
    "    for i in range(len(text) - k + 1):\n",
    "        kmer = text[i:i+k]\n",
    "        prob = 1\n",
    "        for j, nucleotide in enumerate(kmer):\n",
    "            prob *= profile[nucleotide][j]\n",
    "        if prob > max_prob:\n",
    "            max_prob = prob\n",
    "            best_kmer = kmer\n",
    "\n",
    "    return best_kmer\n",
    "\n",
    "def score_motifs(motifs):\n",
    "    score = 0\n",
    "    k = len(motifs[0])\n",
    "\n",
    "    for i in range(k):\n",
    "        column = [motif[i] for motif in motifs]\n",
    "        most_common = max(set(column), key=column.count)\n",
    "        score += sum(1 for nucleotide in column if nucleotide != most_common)\n",
    "\n",
    "    return score\n",
    "\n",
    "def randomized_motif_search_one_iteration_with_initial_motifs(dna_strings, k, t, initial_motifs_str):\n",
    "    dna = dna_strings.strip().split()\n",
    "    initial_motifs = initial_motifs_str.strip().split()\n",
    "\n",
    "    if len(initial_motifs) != t:\n",
    "        raise ValueError(f\"Number of initial motifs ({len(initial_motifs)}) must match the number of DNA strings ({t}).\")\n",
    "    if any(len(motif) != k for motif in initial_motifs):\n",
    "        raise ValueError(f\"All initial motifs must have length k ({k}).\")\n",
    "\n",
    "    print(f\"Initial Motifs: {initial_motifs}\")\n",
    "\n",
    "    # Step 2: Construct a profile matrix from these initial motifs.\n",
    "    profile = profile_matrix_with_pseudocounts(initial_motifs)\n",
    "    print(\"Profile Matrix (with pseudocounts):\")\n",
    "    for nucleotide, probabilities in profile.items():\n",
    "        print(f\"{nucleotide}: {probabilities}\")\n",
    "\n",
    "    # Step 3: For each DNA string, find the profile-most probable k-mer.\n",
    "    new_motifs = [profile_most_probable_kmer(seq, k, profile) for seq in dna]\n",
    "    print(f\"Motifs after one iteration: {new_motifs}\")\n",
    "\n",
    "    return new_motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6c30c194-6910-4a41-9af5-3e3b55b58776",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial Motifs: ['TGA', 'GTT', 'GAA', 'TGT']\n",
      "Profile Matrix (with pseudocounts):\n",
      "A: [0.125, 0.25, 0.375]\n",
      "C: [0.125, 0.125, 0.125]\n",
      "G: [0.375, 0.375, 0.125]\n",
      "T: [0.375, 0.25, 0.375]\n",
      "Motifs after one iteration: ['TGA', 'TAA', 'GGA', 'TGT']\n",
      "['TGA', 'TAA', 'GGA', 'TGT']\n"
     ]
    }
   ],
   "source": [
    "dna_strings = 'TGACGTTC TAAGAGTT GGACGAAA CTGTTCGC'\n",
    "k = 3\n",
    "t = 4\n",
    "initial_motifs_str = 'TGA GTT GAA TGT'\n",
    "print(randomized_motif_search_one_iteration_with_initial_motifs(dna_strings, k, t, initial_motifs_str))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "0cfe3d25-95c4-4890-9499-bdbb3b9b5c02",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAGATCAAGTTGACG\n"
     ]
    }
   ],
   "source": [
    "motifs = 'AAGTGCTAATTGAGC AAGTGAAAGTTGACG AAGATTTTGTTGACG AAGATTGGGTTGACG AAGACGTAGTTGACG AAGATCAAGTTACGG CAGATCAAGTTGAGC AAGATCAAGCACACG ACACTCAAGTTGACG AAGGCGAAGTTGACG AAGATCAAGTGCTCG AAGATCAGTGTGACG AAGATCGGTTTGACG AAGATCAATAGGACG AAGATCAAGTTGCTA CGGATCAAGTTGACA AATGCCAAGTTGACG AAGATCCGCTTGACG AAGACATAGTTGACG GCAATCAAGTTGACG'\n",
    "print(consensus_motif(motifs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "3c6b5762-8e85-4bec-9d3d-87750b3c1847",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAACTTGAGGGCGGT CAAAGCGTATGCGTT CAACTTGTCAACGTT CAACTTGTATGATAT CATTGTGTATGCGTT GTACTTGTATGCGTC TAACTTGTATGCGCA CAACTATAATGCGTT CAACTTGTACAAGTT CACCATGCATGCGTT CAACTTGACCGCGTT CAACTTTGGTGCGTT CAACGGTTATGCGTT CAACTTGTATCAATT CAACGCCTATGCGTT CAACTTGTATGCACC CGTATTGTATGCGTT CAACTTCCCTGCGTT CAACTGAAATGCGTT CAAGGGGTATGCGTT\n"
     ]
    }
   ],
   "source": [
    "k = 15\n",
    "t = 20\n",
    "N = 2000\n",
    "dna_string = 'CTTAAACTTCATCACAAGAGAGCTCACGCCTTTCCTGACGGTTCTTCCTGCTATCGTGACTCGGAGCAAGAGTGCGTTTTTGACAGTAAAACAACTTGAGGGCGGTTCTAAATGTTCATAAAGAGGTGAAACATTCGGAGCTGGCTAAAGGTAACTAGGGGCGATAGATAATCTTTGGCCGTGATATACAAAGAACAAACGATGCTTGAACAGTTCCTTTGGCTGCGACGGCTCATGGGAGCCAGGGCTGGACATGATTAGGATCGGGCCCGATACCTACTAAATAGTAGCGAAGCGTATCTTAAACTTCATCAC AAGAGAGCTCACGCCTTTCCTGACGGTTCTTCCTGCTATCGTGACTCGGAGCAAGAGTGCGTTTTTGACAGTAAAACAACTTGAGGGCGGTTCTAAATGTTCATAAAGAGGTGAAACATTCCAAAGCGTATGCGTTGGAGCTGGCTAAAGGTAACTAGGGGCGATAGATAATCTTTGGCCGTGATATACAAAGAACAAACGATGCTTGAACAGTTCCTTTGGCTGCGACGGCTCATGGGAGCCAGGGCTGGACATGATTAGGATCGGGCCCGATACCTACTAAATAGTAGCGAAGCGTATCTTAAACTTCATCAC ATTGCAAAGTCCACTCGCCTGCCGGTGCACTTACAAAGTGAGATACTCCCTTCGGTAACCGGCTAACTGTCGGAAGAGCTAGAACGGGTAGACGAGTGAGATAATCGGTTGGTTCAAGTGCGGTTTCCTTTCTGCAAGGGCGTTGTGTTAGTGGACAAGCGAGACCGGCAGGTGGATCTTAAGCTCCTTAGAGCGTAAGGCATTCGTGTTTTGCAGTGTGTTAACACGAATATGTTCGTGTTTTATAATGGCCTACTTCAGAACGGCCAGCACAACTTGTCAACGTTTCAGATTAACGGGGAGGCAGACTCTCTT AAGCTTGAAAAGCTGTTCCTGGCTTCTCGGAAGAAAAAAGAATTAGGGTCTAAAATCCATGTGACAACGTAAAACGGACAGCCCTACTTGCGTACTTGTCAACTTGTATGATATCGGAAGGAGATATGTTGTGGATGGGGGGCGCCGTTTTGACCTGTACGGAACGCCGCTAAATTTACCGGGGCGAGGCCGCGCGGTCCAACTCTTCAGTCAACTATAAATCACTGTGCACCTGCGTAGCGGTTAAGCAACGATATCCCAGTGCATTCCCCACAACAAGATTACTCCTCACAGCCTTATTACGAATCAGCTGAA CACTGGTTATGTAATTTCCGCCACTTATTTAAAAAAGCGTTCAGCCAACCCATTCGTTGAGTCGCATCACATGAATAAAGCGAAGTATAATGAGCAGCCTTTACTGTATAGACCGCAAGCCATTGTGTATGCGTTTGTCAAGTCAAAGCAGCGCTATAATTATCGGCATGAAGGCTAAATAATTGTACGGTTCTGTTCCATCAAAAACCGAATGTGAGGTGTCATCTCGCCCAAGAGTTGTCTTGTTTAAAACGGTCGGATTAGGACTGAGTTCAAAATATCAAACGCAATAGTCGGACCCTTAAAACCTCCCTA TCCTAAAGCGACACGAAGCAACAGTCTCAACCTAGAACGGTTTTCTTTTTTTGAGCCTTATGGCTTGCTTCTCGGTGTATCGTAAGGGCTAGCCTGTAAACCCGTCACCTAAATACCGACGGTGAGCTCGACTACCACACAATTCTAGAATAGCTGACTTACTTGTGCAACCAACAGTCGTCCGAAAAGCGCGGGCTCGCTTCTAATTCTTCTTTCTCGGTATAAAGTTAATCTATGCATGAGAGACCCTCACCACACGCAATTGAGGATGGTACTTGTATGCGTCATGGTCGTGCCCTATTCCTCAGATACCGA CAAGACTACCTCAGCTCTCAGGGGCCCTTTCACTCCCCAGGTAACTTGTATGCGCATGGACTCAATTATTATCGCCCTTTGATGGATACAGCAGTGTGATCCCCTCACCGAGATGTGGAGCTCGGGAACGCTATGACTCCCCTGTAACCTATTTGGCGATGGCCGCGGCCGCTAGCATGGTCGCCGTGCGTCTGACTGCCGTTTCGGTGACCCATTCTTGAAGTGGTGGTATTGTTTCGATTTACTCTGTGTTGATCGTTAATTCGTAGGCAACAAGGAGTCGTCCATCCGTTTCGCTTTGACTAGTTGAGATTC TCCTATGAATTTGCAGCCATGTCATGGACAATACTCCGGTAGTAGAAATGAACGGTGATATTCGGAAACAATCAGGTTGTCTGCACCCTCGGCACAAAGTGGGCTTCTACCCATTAGCACCTAGGGTCCCCGAGCTTGCGCAACTATAATGCGTTATTGTAGGGAAATCGGTCAAGTATGATGCCTCTTTTTCTGAGTTCAGCCTGAGGGTGTGTAATACGACCCTGACCAGAGCGCACAATCGGCAGGCATGCGAGGACGCCGGAGAATATATTGCACGACAAAATACTTTCGGTAATAGGCCATGACCCTTTC TCGAGACCACGAGCAAGGCTGAAGCGAATAATACACCGCGAGCATGTTGGCAACGGAAAGGTTCTGATCTAGACCTCGCGACTGGACAACTTGTACAAGTTCTAGGGCGTTATTTAACAGACTAATCGGAGCACTGCTTACATCGAGAAGCAACGTGGAAGTTTTATAGATCCAAGTCTTATCCTTGCCTAAGCCTGAACGGCAACTGCTGAGGCACAGAAGGTAACAGCGTACACATTTGGGATGTCCTCCCATCTTCGTCTTTACCAGGTCTCGGTATAAACACGTTTTGAGACCTTGGCTTTGTGTCCGAAG AAAACGCTCTCCACCTCAGGCCCACCATGCATGCGTTTAACGCAACAGCTTGCCACGTAGTCATACTCCGTGAGCCCACGCGACCGCTAGACCAATTAACTCGGACCTTCGAATTCAATAGTCTTCGCTCTTGTATGCGTTGACGAGGGCCGTCCCCTAGGCTTGTCGGCGCCGGGAAATCCTGTCATACATTGGTCCTGTTGTTAAGAAGAGGACGGGGCCTTCTACACGTCCAGCCAGCGATGCATGGGTTCGGCGGGGCCCGACACTCAGGCAGACAGGTGACGAACGATGTGCTCATATCATCCTTCTGTT GCCGGTATGGATCTCGCCCCCTCCTATCTACTACGCGTTGACAATTCTGACAATTACCGTGCACGCAACTTGACCGCGTTGTGCCGCTGATCCCGGTGCACCAAAACCACTCCAACGGTAATAAGCAAGACCTAATTCCCTGCTAGTATTACTTGAATCGGCCTTTTTACTTGCTGCATTTGAGGCCGGCCCTGTACCTGAAAGCCCGGGCATCGCTAAATGAGATGAGCTTGATCCGGCCCCGCAGGGTTTTTCATCCACGTACAACAGCTAACAAGAAGGTGGCCGTGATTACTCGTTTTGCGTCCGGCTCTC CGTCCGGTTTCGCTCCTCGATCTCGGTGTATTAACGGGTCGATGTCCAGATTATTGACCTGTCCTCATGATCAGGTGTAGACAGATTAATAATGACCCCTTAACCTCCTCTCAGAATGGGGCGAAGGTGGCCCGATGCAACTAGTGCTCAACTTTGGTGCGTTGTACCCGGTGAGACTTACACGATGCTTATGACGAATTGTCCATATGTCAAGACAGCACGTTAGGGGATCTACTGTAGGCCACGAATGGGCACAGTCGAGACTTATACAAGATGGTTCCTACATTTATATAAAGGGTGAGCACCACTAGTCAT TATTATAGCGATTTAAATACCTTGACAATCTGGTATGCGTAGTGGAACTCTGTCGGGGGATGTCGTGTCTTGCGGACATAATTGCTAGTTTGAGATCAACGGTTATGCGTTCGTCACTAGTGGGCAAAGGTGGCGAGTTACGTTGCATCTTCCAGGTACCCCAATTGAGCTGGAATTCAGGCTACATTGCCAGGAGGTTGAGGTCATCCCATACACAGTAACGAACGGCTATGGCCGCATCGAACTGAATATTAAGATTTGAGATACCAGTGTATGCATTTTGCCGTAGAATTTAGAACCCTATAATCACTGGGG GGCCCATTATCATAATCGCCTCTTTACATTATTAACGGGCTTTTAGGTGAAATGTAGGTTACTGCACAGAAACTATCCTAGATGTGTCCAAATCCCTGAATCCCGGTATGTCCTGACTAGGAGATACTACGTAGCAACTTGTATCAATTTACCCACTACTAACTTCAGTTTGCTAGAGGGAAACCGCTAGATCTCTCCCGCATGCCATATCAACGCAGCTGACGTACCACTTATATAGGAGTGTGGTGGGAACATGGGCACAAGACACCTTCCATCTGTGTTGCCAATACCGAATGCATGCATATACCTTTCGAT CGTTGAGAAATACTATCGCTTGTCCGCCGCTTTATGGAGTCCTTAGAACTTACCCTGGAGGTGATGCACCACTGTCTACCGTCGGAATCTAGCGGTCGACGGATAATATTCGTCTTCTAACATGTGGAATATGCTTTCAATAGCGGGGTCGCGGCTACGGCTATCAATGCAGACCTTATCTTTTTATGTAAATGACTTAATACTCACTTAGCGCCACAGTTAGGGTATATCTTAGCAAGTTCCCCCCAACGCCTATGCGTTCTACCTTCGTAGTGCAACGTGACTACGAATAAGTACTCCGTCCACCCTTACGGA GTTAAGCCGAGCATAAATGGTCTAATCGGGTGAATGAGAAAGGGAAACCTTTTAGCAAGGACTGCGGCCTCTCTACGTAGAGGCTTCACTATCACTCCATTCCCTTTCCCCACGCTCAAACACAAACTGAGTTTATCCATCTGATAAGGTTCGAACGGTTTTCCCAAGGGCACGATCTTCGATTTGGCCGTGGTAGATGGAAGGTTCGCAACTTGTATGCACCGGCTATGATTTTATATGTAAGCGCGTAAATCTACACCGACTATTTCGACAGCTTGTAATATCATCGCTTGCCAACTATCACGTGCACTAAGC AAAGAACCTCGAAAGACAGCCAGCCTGCCATACGTTGGTCTCTCGCGCTCACGCGAAAACGAACTCGCCAGAGCCTATCGTGTCGATTTGTAAATAGTGGGGTATTTAGTCGCGAACCAGACATAGAGTGACCATACGGACACTAGGGGGCACTCGACTACCTCGGGACCATCACATAACCCGGTCTGAATACCTCTGGTCTCGTAGGATTGAGTCGGATACCCTGCGGGTGCAACTGCCACTGTTCGGGCGCACTAACCATAACAGCGTATTGTATGCGTTTCTGCGAATGGCTCGGAAAGTGTCGGCAAGCGC GTGCTTAGTAGGTTTGACCTGAACAGTGAATCCTTATTTGTCGAGGTGCCAACAGACCGTTGTTGAGCACACGCCAACGCCTTCTCGTGTGCACTGTGCTTAGCTATCGCCCTTCAAGCCAATACTAGTCATTTATTAAATGGGCCGACACTAATTCCGTCTTTCGTACCGCCGTGAACGGCTCATAAGATAGCTCCAACTTCCCTGCGTTATTTACATCTTTACCCGCGACACGACTATGTGCCCAGGAACCATGGTCGCGCTGGGCCGACTTGCAGCGCCACCCGTTTTTACACACAGCTAGGAAGACCGGAG TGCTGCATCAACTGAAATGCGTTCCGGGCGCTTTTCGATCGACGGCAGACAGCTTTGAGCCATGTCAATTGAAGGCTCTTATGAGGCCCCATCCAACTTCCTAATTCAGCTATGTGCCTTGGCGTTGTGGCGATTTATCCTAAGTTAATCCAACCTTACGGTGTTATAGCGCATCTATGTGTGATCTGATCTGCTATTTGATGCAGTGGGATATCTCCAGTAGTCATAAGGGTTGGATAGTTTGGCCTGGCGCTTGTGGCGAGAGGATGACGGAATTGTAAACGTGTAGTCGGGGACGTTTTCTGCAACCCTCAT ATCACCCTGAAGTAGTGTTGATCGATGCATAGTCCTCTAAACAGATAACCCGGGCACGTGTATTCAAGGCTGTGTCAAAGTCTGCTGCCGAATGAGGTCTAGTGATACCTGTGGTCTTATATAGAAATGGAGCTGTAGCAGTACAAGATCCTTCTCAATCGTCGCTCTTCTTGGAAAGACGTTTGCAGCGGCAACGGGACACGGCTAACTCATTGGTACGACACTCATGCACCTAGTGCATGCGGGCTACGAACAAGGGGTATGCGTTCCCGACCCCACAACCGATTAAGTTTGATTTAGCGAGGCTCATTGGTT'\n",
    "print(run_multiple_starts(dna_string, k, t, N, runs=20))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "57c4df3d-cab1-4176-a96c-c96c125355d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting scipy\n",
      "  Downloading scipy-1.15.3-cp313-cp313-win_amd64.whl.metadata (60 kB)\n",
      "Requirement already satisfied: numpy<2.5,>=1.23.5 in c:\\users\\user\\appdata\\local\\programs\\python\\python313\\lib\\site-packages (from scipy) (2.2.1)\n",
      "Downloading scipy-1.15.3-cp313-cp313-win_amd64.whl (41.0 MB)\n",
      "   ---------------------------------------- 0.0/41.0 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.3/41.0 MB ? eta -:--:--\n",
      "    --------------------------------------- 0.5/41.0 MB 1.8 MB/s eta 0:00:23\n",
      "   - -------------------------------------- 1.0/41.0 MB 1.9 MB/s eta 0:00:21\n",
      "   - -------------------------------------- 1.6/41.0 MB 2.0 MB/s eta 0:00:20\n",
      "   - -------------------------------------- 1.8/41.0 MB 2.0 MB/s eta 0:00:20\n",
      "   -- ------------------------------------- 2.6/41.0 MB 2.2 MB/s eta 0:00:18\n",
      "   -- ------------------------------------- 2.9/41.0 MB 2.2 MB/s eta 0:00:18\n",
      "   --- ------------------------------------ 3.4/41.0 MB 2.2 MB/s eta 0:00:18\n",
      "   --- ------------------------------------ 3.7/41.0 MB 2.2 MB/s eta 0:00:18\n",
      "   ---- ----------------------------------- 4.2/41.0 MB 2.2 MB/s eta 0:00:18\n",
      "   ---- ----------------------------------- 4.7/41.0 MB 2.2 MB/s eta 0:00:17\n",
      "   ---- ----------------------------------- 5.0/41.0 MB 2.1 MB/s eta 0:00:17\n",
      "   ----- ---------------------------------- 5.8/41.0 MB 2.1 MB/s eta 0:00:17\n",
      "   ------ --------------------------------- 6.6/41.0 MB 2.3 MB/s eta 0:00:15\n",
      "   ------ --------------------------------- 6.8/41.0 MB 2.3 MB/s eta 0:00:16\n",
      "   ------ --------------------------------- 6.8/41.0 MB 2.3 MB/s eta 0:00:16\n",
      "   ------ --------------------------------- 7.1/41.0 MB 2.1 MB/s eta 0:00:17\n",
      "   ------- -------------------------------- 7.3/41.0 MB 2.0 MB/s eta 0:00:17\n",
      "   ------- -------------------------------- 7.6/41.0 MB 2.0 MB/s eta 0:00:18\n",
      "   ------- -------------------------------- 7.9/41.0 MB 2.0 MB/s eta 0:00:17\n",
      "   ------- -------------------------------- 8.1/41.0 MB 1.9 MB/s eta 0:00:18\n",
      "   -------- ------------------------------- 8.4/41.0 MB 1.9 MB/s eta 0:00:18\n",
      "   -------- ------------------------------- 8.4/41.0 MB 1.9 MB/s eta 0:00:18\n",
      "   -------- ------------------------------- 8.7/41.0 MB 1.8 MB/s eta 0:00:19\n",
      "   -------- ------------------------------- 8.7/41.0 MB 1.8 MB/s eta 0:00:19\n",
      "   -------- ------------------------------- 9.2/41.0 MB 1.7 MB/s eta 0:00:19\n",
      "   -------- ------------------------------- 9.2/41.0 MB 1.7 MB/s eta 0:00:19\n",
      "   --------- ------------------------------ 9.4/41.0 MB 1.6 MB/s eta 0:00:20\n",
      "   --------- ------------------------------ 9.7/41.0 MB 1.6 MB/s eta 0:00:20\n",
      "   --------- ------------------------------ 10.0/41.0 MB 1.6 MB/s eta 0:00:20\n",
      "   ---------- ----------------------------- 10.5/41.0 MB 1.6 MB/s eta 0:00:19\n",
      "   ---------- ----------------------------- 10.7/41.0 MB 1.6 MB/s eta 0:00:19\n",
      "   ----------- ---------------------------- 11.3/41.0 MB 1.6 MB/s eta 0:00:19\n",
      "   ----------- ---------------------------- 11.5/41.0 MB 1.6 MB/s eta 0:00:19\n",
      "   ----------- ---------------------------- 11.8/41.0 MB 1.6 MB/s eta 0:00:19\n",
      "   ------------ --------------------------- 12.3/41.0 MB 1.6 MB/s eta 0:00:18\n",
      "   ------------ --------------------------- 12.6/41.0 MB 1.6 MB/s eta 0:00:18\n",
      "   ------------ --------------------------- 12.8/41.0 MB 1.6 MB/s eta 0:00:18\n",
      "   ------------ --------------------------- 13.1/41.0 MB 1.6 MB/s eta 0:00:18\n",
      "   ------------- -------------------------- 13.4/41.0 MB 1.6 MB/s eta 0:00:18\n",
      "   ------------- -------------------------- 13.9/41.0 MB 1.6 MB/s eta 0:00:17\n",
      "   -------------- ------------------------- 14.4/41.0 MB 1.6 MB/s eta 0:00:17\n",
      "   -------------- ------------------------- 14.7/41.0 MB 1.6 MB/s eta 0:00:17\n",
      "   --------------- ------------------------ 15.5/41.0 MB 1.7 MB/s eta 0:00:16\n",
      "   --------------- ------------------------ 16.0/41.0 MB 1.7 MB/s eta 0:00:15\n",
      "   ---------------- ----------------------- 16.5/41.0 MB 1.7 MB/s eta 0:00:15\n",
      "   ---------------- ----------------------- 17.3/41.0 MB 1.8 MB/s eta 0:00:14\n",
      "   ----------------- ---------------------- 17.8/41.0 MB 1.8 MB/s eta 0:00:14\n",
      "   ----------------- ---------------------- 18.4/41.0 MB 1.8 MB/s eta 0:00:13\n",
      "   ------------------ --------------------- 18.6/41.0 MB 1.8 MB/s eta 0:00:13\n",
      "   ------------------ --------------------- 18.9/41.0 MB 1.8 MB/s eta 0:00:13\n",
      "   ------------------ --------------------- 19.4/41.0 MB 1.8 MB/s eta 0:00:13\n",
      "   ------------------- -------------------- 19.7/41.0 MB 1.8 MB/s eta 0:00:13\n",
      "   ------------------- -------------------- 19.9/41.0 MB 1.8 MB/s eta 0:00:12\n",
      "   ------------------- -------------------- 20.4/41.0 MB 1.8 MB/s eta 0:00:12\n",
      "   -------------------- ------------------- 21.0/41.0 MB 1.8 MB/s eta 0:00:12\n",
      "   -------------------- ------------------- 21.2/41.0 MB 1.8 MB/s eta 0:00:12\n",
      "   -------------------- ------------------- 21.5/41.0 MB 1.8 MB/s eta 0:00:11\n",
      "   --------------------- ------------------ 21.8/41.0 MB 1.8 MB/s eta 0:00:11\n",
      "   --------------------- ------------------ 22.3/41.0 MB 1.8 MB/s eta 0:00:11\n",
      "   ---------------------- ----------------- 22.8/41.0 MB 1.8 MB/s eta 0:00:11\n",
      "   ---------------------- ----------------- 23.1/41.0 MB 1.8 MB/s eta 0:00:11\n",
      "   ----------------------- ---------------- 23.9/41.0 MB 1.8 MB/s eta 0:00:10\n",
      "   ----------------------- ---------------- 24.4/41.0 MB 1.8 MB/s eta 0:00:10\n",
      "   ------------------------ --------------- 24.9/41.0 MB 1.8 MB/s eta 0:00:09\n",
      "   ------------------------- -------------- 26.0/41.0 MB 1.9 MB/s eta 0:00:09\n",
      "   -------------------------- ------------- 27.0/41.0 MB 1.9 MB/s eta 0:00:08\n",
      "   -------------------------- ------------- 27.5/41.0 MB 1.9 MB/s eta 0:00:07\n",
      "   --------------------------- ------------ 27.8/41.0 MB 1.9 MB/s eta 0:00:07\n",
      "   --------------------------- ------------ 28.6/41.0 MB 1.9 MB/s eta 0:00:07\n",
      "   ----------------------------- ---------- 29.9/41.0 MB 2.0 MB/s eta 0:00:06\n",
      "   ------------------------------ --------- 31.2/41.0 MB 2.1 MB/s eta 0:00:05\n",
      "   ------------------------------- -------- 32.8/41.0 MB 2.1 MB/s eta 0:00:04\n",
      "   -------------------------------- ------- 33.0/41.0 MB 2.1 MB/s eta 0:00:04\n",
      "   -------------------------------- ------- 33.6/41.0 MB 2.1 MB/s eta 0:00:04\n",
      "   --------------------------------- ------ 34.1/41.0 MB 2.1 MB/s eta 0:00:04\n",
      "   ---------------------------------- ----- 34.9/41.0 MB 2.2 MB/s eta 0:00:03\n",
      "   ---------------------------------- ----- 35.4/41.0 MB 2.2 MB/s eta 0:00:03\n",
      "   ----------------------------------- ---- 35.9/41.0 MB 2.2 MB/s eta 0:00:03\n",
      "   ----------------------------------- ---- 36.2/41.0 MB 2.2 MB/s eta 0:00:03\n",
      "   ----------------------------------- ---- 36.4/41.0 MB 2.2 MB/s eta 0:00:03\n",
      "   ------------------------------------ --- 37.0/41.0 MB 2.2 MB/s eta 0:00:02\n",
      "   ------------------------------------ --- 37.5/41.0 MB 2.2 MB/s eta 0:00:02\n",
      "   ------------------------------------ --- 37.7/41.0 MB 2.1 MB/s eta 0:00:02\n",
      "   ------------------------------------- -- 38.0/41.0 MB 2.1 MB/s eta 0:00:02\n",
      "   ------------------------------------- -- 38.3/41.0 MB 2.1 MB/s eta 0:00:02\n",
      "   ------------------------------------- -- 38.5/41.0 MB 2.1 MB/s eta 0:00:02\n",
      "   ------------------------------------- -- 38.8/41.0 MB 2.1 MB/s eta 0:00:02\n",
      "   -------------------------------------- - 39.1/41.0 MB 2.1 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 39.6/41.0 MB 2.1 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 39.8/41.0 MB 2.1 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 39.8/41.0 MB 2.1 MB/s eta 0:00:01\n",
      "   ---------------------------------------  40.1/41.0 MB 2.1 MB/s eta 0:00:01\n",
      "   ---------------------------------------  40.9/41.0 MB 2.1 MB/s eta 0:00:01\n",
      "   ---------------------------------------  40.9/41.0 MB 2.1 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 41.0/41.0 MB 2.0 MB/s eta 0:00:00\n",
      "Installing collected packages: scipy\n",
      "Successfully installed scipy-1.15.3\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "[notice] A new release of pip is available: 25.0.1 -> 25.1.1\n",
      "[notice] To update, run: python.exe -m pip install --upgrade pip\n"
     ]
    }
   ],
   "source": [
    "pip install scipy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "80e1f88f-8b98-45fd-bc8a-1ce9d04fdab4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy.stats import chi2_contingency\n",
    "\n",
    "def compare_count_matrices_chi2(count_matrix_1, count_matrix_2):\n",
    "    \"\"\"\n",
    "    Compares two count matrices (Position Frequency Matrices) using the Chi-squared test.\n",
    "    It performs the Chi-squared test column by column and returns the average p-value.\n",
    "\n",
    "    Args:\n",
    "        count_matrix1 (numpy.ndarray): Count matrix 1 (shape: length x 4).\n",
    "        count_matrix2 (numpy.ndarray): Count matrix 2 (shape: length x 4).\n",
    "            Rows represent nucleotides (A, C, G, T), columns represent positions.\n",
    "\n",
    "    Returns:\n",
    "        float: The average p-value from the column-wise Chi-squared tests.\n",
    "               A lower p-value indicates a greater difference between the columns.\n",
    "    \"\"\"\n",
    "    if count_matrix_1.shape != count_matrix_2.shape:\n",
    "        raise ValueError(\"Count matrices must have the same shape.\")\n",
    "\n",
    "    p_values = []\n",
    "    for i in range(count_matrix_1.shape[0]):\n",
    "        # Create a contingency table for the current column\n",
    "        contingency_table = np.array([count_matrix_1[i, :], count_matrix_2[i, :]])\n",
    "        # Perform the Chi-squared test\n",
    "        chi2, p, dof, expected = chi2_contingency(contingency_table)\n",
    "        p_values.append(p)\n",
    "\n",
    "    return np.mean(p_values)\n",
    "\n",
    "def correct_for_multiple_comparisons(p_values, method='bonferroni'):\n",
    "    \"\"\"\n",
    "    Corrects p-values for multiple comparisons using Bonferroni or Benjamini-Hochberg.\n",
    "\n",
    "    Args:\n",
    "        p_values (list of float): List of p-values.\n",
    "        method (str): 'bonferroni' or 'fdr' (for Benjamini-Hochberg).\n",
    "            Defaults to 'bonferroni'.\n",
    "\n",
    "    Returns:\n",
    "        list of float: List of corrected p-values.\n",
    "    \"\"\"\n",
    "    if method == 'bonferroni':\n",
    "        corrected_p_values = [min(1, p * len(p_values)) for p in p_values]\n",
    "    elif method == 'fdr':\n",
    "        from scipy.stats import rankdata\n",
    "        ranked_p_values = rankdata(p_values, method='min')\n",
    "        corrected_p_values = [\n",
    "            min(1, p * len(p_values) / r) for p, r in zip(p_values, ranked_p_values)\n",
    "        ]\n",
    "    else:\n",
    "        raise ValueError(\"Invalid method. Choose 'bonferroni' or 'fdr'.\")\n",
    "    return corrected_p_values\n",
    "    \n",
    "def compare_count_matrices_chi2_with_correction(count_matrix_1, count_matrix_2, correction_method='bonferroni'):\n",
    "    \"\"\"\n",
    "    Compares two count matrices using column-wise Chi-squared tests, and corrects\n",
    "    the resulting p-values for multiple comparisons.\n",
    "\n",
    "    Args:\n",
    "        count_matrix1 (numpy.ndarray): Count matrix 1 (shape: length x 4).\n",
    "        count_matrix2 (numpy.ndarray): Count matrix 2 (shape: length x 4).\n",
    "        correction_method (str): 'bonferroni' or 'fdr'.  Defaults to 'bonferroni'.\n",
    "\n",
    "    Returns:\n",
    "        tuple: (average_p_value, corrected_p_values, average_corrected_p_value)\n",
    "            average_p_value (float): The average of the uncorrected p-values.\n",
    "            corrected_p_values (list of float): The list of p-values, corrected for\n",
    "                multiple comparisons.\n",
    "            average_corrected_p_value (float): The average of the corrected p-values.\n",
    "    \"\"\"\n",
    "    if count_matrix_1.shape != count_matrix_2.shape:\n",
    "        raise ValueError(\"Count matrices must have the same shape.\")\n",
    "\n",
    "    p_values = []\n",
    "    for i in range(count_matrix_1.shape[0]):\n",
    "        contingency_table = np.array([count_matrix_1[i, :], count_matrix_2[i, :]])\n",
    "        chi2, p, dof, expected = chi2_contingency(contingency_table)\n",
    "        p_values.append(p)\n",
    "\n",
    "    average_p_value = np.mean(p_values)\n",
    "    corrected_p_values = correct_for_multiple_comparisons(p_values, method=correction_method)\n",
    "    average_corrected_p_value = np.mean(corrected_p_values)\n",
    "\n",
    "    return average_p_value, corrected_p_values, average_corrected_p_value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "62840de4-bd9f-4db5-828b-44bd8f1ce1dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy.stats import chi2_contingency\n",
    "\n",
    "def compare_count_matrices_chi2(count_matrix_1, count_matrix_2):\n",
    "    \"\"\"\n",
    "    Compares two count matrices (Position Frequency Matrices) using the Chi-squared test.\n",
    "    It performs the Chi-squared test column by column and returns the average p-value.\n",
    "\n",
    "    Args:\n",
    "        count_matrix1 (numpy.ndarray): Count matrix 1 (shape: length x 4).\n",
    "        count_matrix2 (numpy.ndarray): Count matrix 2 (shape: length x 4).\n",
    "            Rows represent nucleotides (A, C, G, T), columns represent positions.\n",
    "\n",
    "    Returns:\n",
    "        float: The average p-value from the column-wise Chi-squared tests.\n",
    "               A lower p-value indicates a greater difference between the columns.\n",
    "               Returns 1.0 if a column comparison is invalid.\n",
    "    \"\"\"\n",
    "    if count_matrix_1.shape != count_matrix_2.shape:\n",
    "        raise ValueError(\"Count matrices must have the same shape.\")\n",
    "\n",
    "    p_values = []\n",
    "    for i in range(count_matrix_1.shape[1]):  # Iterate over columns\n",
    "        # Create a contingency table for the current column\n",
    "        contingency_table = np.array([count_matrix_1[:, i], count_matrix_2[:, i]])\n",
    "        # Perform the Chi-squared test\n",
    "        try:\n",
    "            chi2, p, dof, expected = chi2_contingency(contingency_table)\n",
    "            p_values.append(p)\n",
    "        except ValueError as e:\n",
    "            print(f\"Chi-squared test failed for column {i}: {e}\")\n",
    "            # Append a p-value of 1.0 for failed tests.  This will have minimal\n",
    "            # impact on the *average* p-value, and indicates no difference for this comparison.\n",
    "            p_values.append(1.0)\n",
    "\n",
    "    return np.mean(p_values)\n",
    "\n",
    "\n",
    "\n",
    "def correct_for_multiple_comparisons(p_values, method='bonferroni'):\n",
    "    \"\"\"\n",
    "    Corrects p-values for multiple comparisons using Bonferroni or Benjamini-Hochberg.\n",
    "\n",
    "    Args:\n",
    "        p_values (list of float): List of p-values.\n",
    "        method (str): 'bonferroni' or 'fdr' (for Benjamini-Hochberg).\n",
    "            Defaults to 'bonferroni'.\n",
    "\n",
    "    Returns:\n",
    "        list of float: List of corrected p-values.\n",
    "    \"\"\"\n",
    "    if method == 'bonferroni':\n",
    "        corrected_p_values = [min(1, p * len(p_values)) for p in p_values]\n",
    "    elif method == 'fdr':\n",
    "        from scipy.stats import rankdata\n",
    "        ranked_p_values = rankdata(p_values, method='min')\n",
    "        corrected_p_values = [\n",
    "            min(1, p * len(p_values) / r) for p, r in zip(p_values, ranked_p_values)\n",
    "        ]\n",
    "    else:\n",
    "        raise ValueError(\"Invalid method. Choose 'bonferroni' or 'fdr'.\")\n",
    "    return corrected_p_values\n",
    "    \n",
    "def compare_count_matrices_chi2_with_correction(count_matrix_1, count_matrix_2, correction_method='bonferroni'):\n",
    "    \"\"\"\n",
    "    Compares two count matrices using column-wise Chi-squared tests, and corrects\n",
    "    the resulting p-values for multiple comparisons.\n",
    "\n",
    "    Args:\n",
    "        count_matrix1 (numpy.ndarray): Count matrix 1 (shape: length x 4).\n",
    "        count_matrix2 (numpy.ndarray): Count matrix 2 (shape: length x 4).\n",
    "        correction_method (str): 'bonferroni' or 'fdr'.  Defaults to 'bonferroni'.\n",
    "\n",
    "    Returns:\n",
    "        tuple: (average_p_value, corrected_p_values, average_corrected_p_value)\n",
    "            average_p_value (float): The average of the uncorrected p-values.\n",
    "            corrected_p_values (list of float): The list of p-values, corrected for\n",
    "                multiple comparisons.\n",
    "            average_corrected_p_value (float): The average of the corrected p-values.\n",
    "    \"\"\"\n",
    "    if count_matrix_1.shape != count_matrix_2.shape:\n",
    "        raise ValueError(\"Count matrices must have the same shape.\")\n",
    "\n",
    "    p_values = []\n",
    "    for i in range(count_matrix_1.shape[1]):\n",
    "        contingency_table = np.array([count_matrix_1[:, i], count_matrix_2[:, i]])\n",
    "        try:\n",
    "            chi2, p, dof, expected = chi2_contingency(contingency_table)\n",
    "            p_values.append(p)\n",
    "        except ValueError as e:\n",
    "            print(f\"Chi-squared test failed for column {i}: {e}\")\n",
    "            p_values.append(1.0)\n",
    "            \n",
    "    average_p_value = np.mean(p_values)\n",
    "    corrected_p_values = correct_for_multiple_comparisons(p_values, method=correction_method)\n",
    "    average_corrected_p_value = np.mean(corrected_p_values)\n",
    "\n",
    "    return average_p_value, corrected_p_values, average_corrected_p_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "898d23d6-11c3-4dd3-b2c6-7be62e6dc0f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chi-squared test failed for column 9: The internally computed table of expected frequencies has a zero element at (np.int64(0), np.int64(1)).\n",
      "Chi-squared test failed for column 18: The internally computed table of expected frequencies has a zero element at (np.int64(0), np.int64(0)).\n",
      "Average p-value from Chi-squared test: 0.10335288859259402\n",
      "Chi-squared test failed for column 9: The internally computed table of expected frequencies has a zero element at (np.int64(0), np.int64(1)).\n",
      "Chi-squared test failed for column 18: The internally computed table of expected frequencies has a zero element at (np.int64(0), np.int64(0)).\n",
      "\n",
      "Average p-value: 0.10335288859259402\n",
      "Corrected p-values (FDR): [np.float64(2.7293734740555677e-06), np.float64(1.4535514765952907e-05), np.float64(9.440696715003092e-05), np.float64(5.256962676203467e-10), np.float64(0.0005525707830363717), np.float64(1.2756501825974245e-11), np.float64(0.027505348685333837), np.float64(1.1077995571012232e-06), np.float64(1.8183708945690876e-10), 1, np.float64(1.0458614179073322e-10), np.float64(0.000281366925171536), np.float64(3.885380910763746e-12), np.float64(4.8906750099339215e-06), np.float64(0.028007659612003206), np.float64(1.641610589028895e-12), np.float64(0.019800744236281122), np.float64(5.647236166580573e-10), 1, np.float64(0.002656390092077084)]\n",
      "Average corrected p-value (FDR): 0.10394608760294936\n",
      "Chi-squared test failed for column 9: The internally computed table of expected frequencies has a zero element at (np.int64(0), np.int64(1)).\n",
      "Chi-squared test failed for column 18: The internally computed table of expected frequencies has a zero element at (np.int64(0), np.int64(0)).\n",
      "\n",
      "Average p-value: 0.10335288859259402\n",
      "Corrected p-values (Bonferroni): [np.float64(2.4564361266500107e-05), np.float64(0.00015989066242548197), np.float64(0.0011328836058003711), np.float64(3.6798738733424273e-09), np.float64(0.007735990962509203), np.float64(3.826950547792273e-11), np.float64(0.4675909276506752), np.float64(8.862396456809785e-06), np.float64(9.091854472845437e-10), 1, np.float64(4.183445671629329e-10), np.float64(0.003657770027229968), np.float64(7.770761821527491e-12), np.float64(4.890675009933922e-05), np.float64(0.5041378730160577), np.float64(1.641610589028895e-12), np.float64(0.31681190778049795), np.float64(3.3883416999483443e-09), 1, np.float64(0.03984585138115626)]\n",
      "Average corrected p-value (Bonferroni): 0.16705777185188012\n"
     ]
    }
   ],
   "source": [
    "count_matrix_1 = np.array([\n",
    "    [7, 19, 13, 1, 3, 9, 13, 0, 1, 35, 6, 1, 12, 7, 7, 0, 8, 0, 0, 7],\n",
    "    [21, 9, 13, 11, 4, 0, 1, 10, 7, 0, 24, 22, 1, 13, 14, 0, 12, 25, 22, 22],\n",
    "    [8, 5, 10, 5, 7, 27, 18, 20, 28, 1, 0, 1, 3, 0, 8, 33, 14, 7, 1, 4],\n",
    "    [0, 3, 0, 19, 22, 0, 4, 6, 0, 0, 6, 12, 20, 16, 7, 3, 2, 4, 13, 3]\n",
    "])\n",
    "count_matrix_2 = np.array([\n",
    "    [20, 4, 10, 2, 8, 6, 4, 14, 18, 18, 1, 2, 0, 5, 1, 1, 21, 21, 0, 4],\n",
    "    [2, 7, 0, 1, 13, 27, 7, 0, 2, 0, 1, 4, 34, 20, 19, 29, 5, 0, 14, 9],\n",
    "    [6, 25, 21, 33, 10, 0, 21, 8, 3, 16, 28, 1, 0, 11, 3, 0, 8, 13, 6, 13],\n",
    "    [8, 0, 5, 0, 5, 3, 4, 14, 13, 2, 6, 29, 2, 0, 13, 6, 2, 2, 16, 10]\n",
    "])\n",
    "\n",
    "# Compare using Chi-squared test and print the result\n",
    "average_p_value = compare_count_matrices_chi2(count_matrix_1, count_matrix_2)\n",
    "print(f\"Average p-value from Chi-squared test: {average_p_value}\")\n",
    "\n",
    "# Example of using the function with multiple comparisons correction:\n",
    "average_p_value, corrected_p_values, average_corrected_p_value = compare_count_matrices_chi2_with_correction(count_matrix_1, count_matrix_2, correction_method='fdr')\n",
    "print(f\"\\nAverage p-value: {average_p_value}\")\n",
    "print(f\"Corrected p-values (FDR): {corrected_p_values}\")\n",
    "print(f\"Average corrected p-value (FDR): {average_corrected_p_value}\")\n",
    "\n",
    "average_p_value, corrected_p_values, average_corrected_p_value = compare_count_matrices_chi2_with_correction(count_matrix_1, count_matrix_2, correction_method='bonferroni')\n",
    "print(f\"\\nAverage p-value: {average_p_value}\")\n",
    "print(f\"Corrected p-values (Bonferroni): {corrected_p_values}\")\n",
    "print(f\"Average corrected p-value (Bonferroni): {average_corrected_p_value}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
